{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: stats4\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
      "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
      "    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,\n",
      "    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,\n",
      "    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,\n",
      "    table, tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "Loading required package: Biobase\n",
      "\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:utils’:\n",
      "\n",
      "    findMatches\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "\n",
      "Warning message:\n",
      "“replacing previous import ‘utils::findMatches’ by ‘S4Vectors::findMatches’ when loading ‘AnnotationDbi’”\n",
      "\n",
      "\n",
      "Loading required package: EDASeq\n",
      "\n",
      "Loading required package: ShortRead\n",
      "\n",
      "Loading required package: BiocParallel\n",
      "\n",
      "Loading required package: Biostrings\n",
      "\n",
      "Loading required package: XVector\n",
      "\n",
      "Loading required package: GenomeInfoDb\n",
      "\n",
      "\n",
      "Attaching package: ‘Biostrings’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    strsplit\n",
      "\n",
      "\n",
      "Loading required package: Rsamtools\n",
      "\n",
      "Loading required package: GenomicRanges\n",
      "\n",
      "Loading required package: GenomicAlignments\n",
      "\n",
      "Loading required package: SummarizedExperiment\n",
      "\n",
      "Loading required package: MatrixGenerics\n",
      "\n",
      "Loading required package: matrixStats\n",
      "\n",
      "\n",
      "Attaching package: ‘matrixStats’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:Biobase’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘MatrixGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
      "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
      "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
      "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
      "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
      "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
      "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
      "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
      "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
      "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
      "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
      "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
      "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
      "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
      "    rowWeightedSds, rowWeightedVars\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:Biobase’:\n",
      "\n",
      "    rowMedians\n",
      "\n",
      "\n",
      "Warning message:\n",
      "“multiple methods tables found for ‘acbind’”\n",
      "Warning message:\n",
      "“multiple methods tables found for ‘arbind’”\n",
      "Loading required package: edgeR\n",
      "\n",
      "Loading required package: limma\n",
      "\n",
      "\n",
      "Attaching package: ‘limma’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    plotMA\n",
      "\n",
      "\n",
      "Loading required package: ggplot2\n",
      "\n",
      "\n",
      "Attaching package: ‘plotly’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:ggplot2’:\n",
      "\n",
      "    last_plot\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:XVector’:\n",
      "\n",
      "    slice\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:AnnotationDbi’:\n",
      "\n",
      "    select\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    slice\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:S4Vectors’:\n",
      "\n",
      "    rename\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:stats’:\n",
      "\n",
      "    filter\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:graphics’:\n",
      "\n",
      "    layout\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘plyr’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:plotly’:\n",
      "\n",
      "    arrange, mutate, rename, summarise\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:ShortRead’:\n",
      "\n",
      "    id\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:matrixStats’:\n",
      "\n",
      "    count\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:XVector’:\n",
      "\n",
      "    compact\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    desc\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:S4Vectors’:\n",
      "\n",
      "    rename\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:plyr’:\n",
      "\n",
      "    arrange, count, desc, failwith, id, mutate, rename, summarise,\n",
      "    summarize\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:ShortRead’:\n",
      "\n",
      "    id\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:GenomicAlignments’:\n",
      "\n",
      "    first, last\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:matrixStats’:\n",
      "\n",
      "    count\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:GenomicRanges’:\n",
      "\n",
      "    intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:Biostrings’:\n",
      "\n",
      "    collapse, intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomeInfoDb’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:XVector’:\n",
      "\n",
      "    slice\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:AnnotationDbi’:\n",
      "\n",
      "    select\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:IRanges’:\n",
      "\n",
      "    collapse, desc, intersect, setdiff, slice, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:S4Vectors’:\n",
      "\n",
      "    first, intersect, rename, setdiff, setequal, union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:Biobase’:\n",
      "\n",
      "    combine\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "Registered S3 method overwritten by 'GGally':\n",
      "  method from   \n",
      "  +.gg   ggplot2\n",
      "\n",
      "── \u001b[1mAttaching core tidyverse packages\u001b[22m ──────────────────────── tidyverse 2.0.0 ──\n",
      "\u001b[32m✔\u001b[39m \u001b[34mforcats  \u001b[39m 1.0.0     \u001b[32m✔\u001b[39m \u001b[34mstringr  \u001b[39m 1.5.0\n",
      "\u001b[32m✔\u001b[39m \u001b[34mlubridate\u001b[39m 1.9.2     \u001b[32m✔\u001b[39m \u001b[34mtibble   \u001b[39m 3.2.1\n",
      "\u001b[32m✔\u001b[39m \u001b[34mpurrr    \u001b[39m 1.0.1     \u001b[32m✔\u001b[39m \u001b[34mtidyr    \u001b[39m 1.3.0\n",
      "\u001b[32m✔\u001b[39m \u001b[34mreadr    \u001b[39m 2.1.4     \n",
      "── \u001b[1mConflicts\u001b[22m ────────────────────────────────────────── tidyverse_conflicts() ──\n",
      "\u001b[31m✖\u001b[39m \u001b[34mlubridate\u001b[39m::\u001b[32m%within%()\u001b[39m masks \u001b[34mIRanges\u001b[39m::%within%()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32marrange()\u001b[39m      masks \u001b[34mplyr\u001b[39m::arrange(), \u001b[34mplotly\u001b[39m::arrange()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mcollapse()\u001b[39m     masks \u001b[34mBiostrings\u001b[39m::collapse(), \u001b[34mIRanges\u001b[39m::collapse()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mcombine()\u001b[39m      masks \u001b[34mBiobase\u001b[39m::combine(), \u001b[34mBiocGenerics\u001b[39m::combine()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mpurrr\u001b[39m::\u001b[32mcompact()\u001b[39m      masks \u001b[34mplyr\u001b[39m::compact(), \u001b[34mXVector\u001b[39m::compact()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mpurrr\u001b[39m::\u001b[32mcompose()\u001b[39m      masks \u001b[34mShortRead\u001b[39m::compose()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mcount()\u001b[39m        masks \u001b[34mplyr\u001b[39m::count(), \u001b[34mmatrixStats\u001b[39m::count()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mdesc()\u001b[39m         masks \u001b[34mplyr\u001b[39m::desc(), \u001b[34mIRanges\u001b[39m::desc()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mtidyr\u001b[39m::\u001b[32mexpand()\u001b[39m       masks \u001b[34mS4Vectors\u001b[39m::expand()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mfailwith()\u001b[39m     masks \u001b[34mplyr\u001b[39m::failwith()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mfilter()\u001b[39m       masks \u001b[34mplotly\u001b[39m::filter(), \u001b[34mstats\u001b[39m::filter()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mfirst()\u001b[39m        masks \u001b[34mGenomicAlignments\u001b[39m::first(), \u001b[34mS4Vectors\u001b[39m::first()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mid()\u001b[39m           masks \u001b[34mplyr\u001b[39m::id(), \u001b[34mShortRead\u001b[39m::id()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mlag()\u001b[39m          masks \u001b[34mstats\u001b[39m::lag()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mlast()\u001b[39m         masks \u001b[34mGenomicAlignments\u001b[39m::last()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mmutate()\u001b[39m       masks \u001b[34mplyr\u001b[39m::mutate(), \u001b[34mplotly\u001b[39m::mutate()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mggplot2\u001b[39m::\u001b[32mPosition()\u001b[39m   masks \u001b[34mBiocGenerics\u001b[39m::Position(), \u001b[34mbase\u001b[39m::Position()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mpurrr\u001b[39m::\u001b[32mreduce()\u001b[39m       masks \u001b[34mGenomicRanges\u001b[39m::reduce(), \u001b[34mIRanges\u001b[39m::reduce()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mrename()\u001b[39m       masks \u001b[34mplyr\u001b[39m::rename(), \u001b[34mplotly\u001b[39m::rename(), \u001b[34mS4Vectors\u001b[39m::rename()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mlubridate\u001b[39m::\u001b[32msecond()\u001b[39m   masks \u001b[34mGenomicAlignments\u001b[39m::second(), \u001b[34mS4Vectors\u001b[39m::second()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mlubridate\u001b[39m::\u001b[32msecond<-()\u001b[39m masks \u001b[34mS4Vectors\u001b[39m::second<-()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mselect()\u001b[39m       masks \u001b[34mplotly\u001b[39m::select(), \u001b[34mAnnotationDbi\u001b[39m::select()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mslice()\u001b[39m        masks \u001b[34mplotly\u001b[39m::slice(), \u001b[34mXVector\u001b[39m::slice(), \u001b[34mIRanges\u001b[39m::slice()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32msummarise()\u001b[39m    masks \u001b[34mplyr\u001b[39m::summarise(), \u001b[34mplotly\u001b[39m::summarise()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32msummarize()\u001b[39m    masks \u001b[34mplyr\u001b[39m::summarize()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mtibble\u001b[39m::\u001b[32mview()\u001b[39m        masks \u001b[34mShortRead\u001b[39m::view()\n",
      "\u001b[36mℹ\u001b[39m Use the conflicted package (\u001b[3m\u001b[34m<http://conflicted.r-lib.org/>\u001b[39m\u001b[23m) to force all conflicts to become errors\n",
      "\n",
      "Attaching package: ‘scales’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:purrr’:\n",
      "\n",
      "    discard\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:readr’:\n",
      "\n",
      "    col_factor\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘reshape2’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:tidyr’:\n",
      "\n",
      "    smiths\n",
      "\n",
      "\n",
      "========================================\n",
      "circlize version 0.4.15\n",
      "CRAN page: https://cran.r-project.org/package=circlize\n",
      "Github page: https://github.com/jokergoo/circlize\n",
      "Documentation: https://jokergoo.github.io/circlize_book/book/\n",
      "\n",
      "If you use it in published research, please cite:\n",
      "Gu, Z. circlize implements and enhances circular visualization\n",
      "  in R. Bioinformatics 2014.\n",
      "\n",
      "This message can be suppressed by:\n",
      "  suppressPackageStartupMessages(library(circlize))\n",
      "========================================\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘gplots’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    space\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:S4Vectors’:\n",
      "\n",
      "    space\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:stats’:\n",
      "\n",
      "    lowess\n",
      "\n",
      "\n",
      "Loading required package: grid\n",
      "\n",
      "\n",
      "Attaching package: ‘grid’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:Biostrings’:\n",
      "\n",
      "    pattern\n",
      "\n",
      "\n",
      "Loading required package: futile.logger\n",
      "\n",
      "'getOption(\"repos\")' replaces Bioconductor standard repositories, see\n",
      "'help(\"repositories\", package = \"BiocManager\")' for details.\n",
      "Replacement repositories:\n",
      "    CRAN: https://cloud.r-project.org\n",
      "\n",
      "Bioconductor version 3.17 (BiocManager 1.30.20), R 4.3.0 (2023-04-21)\n",
      "\n",
      "Warning message:\n",
      "“package(s) not installed when version(s) same as or greater than current; use\n",
      "  `force = TRUE` to re-install: 'DESeq2'”\n",
      "Installation paths not writeable, unable to update packages\n",
      "  path: /usr/lib/R/library\n",
      "  packages:\n",
      "    boot, foreign, Matrix\n",
      "\n",
      "Old packages: 'AnnotationDbi', 'bench', 'BiocParallel', 'Biostrings', 'cachem',\n",
      "  'deldir', 'evaluate', 'genefilter', 'httpuv', 'httr', 'httr2', 'Lahman',\n",
      "  'later', 'microbenchmark', 'profvis', 'rlang', 'S4Vectors', 'sass',\n",
      "  'SummarizedExperiment', 'testthat', 'tzdb', 'viridisLite', 'vroom', 'waldo'\n",
      "\n",
      "Installing packages into ‘/home/kchen/R/x86_64-pc-linux-gnu-library/4.3’\n",
      "(as ‘lib’ is unspecified)\n",
      "\n",
      "Warning message in install.packages(c(\"ggrepel\", \"ggplot2\", \"plotly\", \"plyr\", \"dplyr\", :\n",
      "“installation of package ‘plyr’ had non-zero exit status”\n",
      "'getOption(\"repos\")' replaces Bioconductor standard repositories, see\n",
      "'help(\"repositories\", package = \"BiocManager\")' for details.\n",
      "Replacement repositories:\n",
      "    CRAN: https://cloud.r-project.org\n",
      "\n",
      "Bioconductor version 3.17 (BiocManager 1.30.20), R 4.3.0 (2023-04-21)\n",
      "\n",
      "Warning message:\n",
      "“package(s) not installed when version(s) same as or greater than current; use\n",
      "  `force = TRUE` to re-install: 'edgeR' 'preprocessCore' 'limma'”\n",
      "Installation paths not writeable, unable to update packages\n",
      "  path: /usr/lib/R/library\n",
      "  packages:\n",
      "    boot, foreign, Matrix\n",
      "\n",
      "Old packages: 'AnnotationDbi', 'bench', 'BiocParallel', 'Biostrings', 'cachem',\n",
      "  'deldir', 'evaluate', 'genefilter', 'httpuv', 'httr', 'httr2', 'Lahman',\n",
      "  'later', 'microbenchmark', 'profvis', 'rlang', 'S4Vectors', 'sass',\n",
      "  'SummarizedExperiment', 'testthat', 'tzdb', 'viridisLite', 'vroom', 'waldo'\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(\"AnnotationDbi\")\n",
    "library(\"org.Hs.eg.db\")\n",
    "library(RUVSeq)\n",
    "source(\"downstreamAnalysis_RNAseqFunctions.R\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>41</li><li>15</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 41\n",
       "\\item 15\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 41\n",
       "2. 15\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] 41 15"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>61114</li><li>41</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 61114\n",
       "\\item 41\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 61114\n",
       "2. 41\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] 61114    41"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Read in data\n",
    "cts = as.matrix(read.csv(\"../../data/cts_c.csv\", sep=',', row.names='gene'))\n",
    "coldata = read.csv('../../data/coldata_c.csv', sep=',', row.names='Sample_ID')\n",
    "\n",
    "cts = cts[,-which(colnames(cts) == 'PB17_236_257_A1')]\n",
    "coldata = coldata[-which(rownames(coldata) == 'PB17_236_257_A1'),]\n",
    "cts = cts[,-which(colnames(cts) == 'PB600052_A7')]\n",
    "coldata = coldata[-which(rownames(coldata) == 'PB600052_A7'),]\n",
    "cts = cts[,-which(colnames(cts) == 'PB600298_A8')]\n",
    "coldata = coldata[-which(rownames(coldata) == 'PB600298_A8'),]\n",
    "\n",
    "dim(coldata)\n",
    "dim(cts)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 15</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>recur</th><th scope=col>subject_id_per_gc</th><th scope=col>disease</th><th scope=col>sex</th><th scope=col>ethnicity</th><th scope=col>race</th><th scope=col>smoking</th><th scope=col>age</th><th scope=col>months_surg_postop_colonoscopy</th><th scope=col>subject_id</th><th scope=col>TIN</th><th scope=col>PC1</th><th scope=col>tissue</th><th scope=col>block_location</th><th scope=col>inflammation_status</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>PB12_770_A1</th><td>Non-recurrent</td><td>12-770                                      </td><td>CD</td><td>F</td><td>NH</td><td>WC     </td><td>0</td><td>47</td><td>6</td><td>12_770    </td><td>76.2</td><td>-24.46333</td><td>colon</td><td>colon</td><td>NI</td></tr>\n",
       "\t<tr><th scope=row>PB138_A8</th><td>Recurrent    </td><td>138                                         </td><td>CD</td><td>F</td><td>NH</td><td>WC     </td><td>0</td><td>22</td><td>6</td><td>138       </td><td>80.4</td><td>-30.36808</td><td>colon</td><td>colon</td><td>NI</td></tr>\n",
       "\t<tr><th scope=row>PB140_COLON</th><td>Recurrent    </td><td>140 (604606)                                </td><td>CD</td><td>M</td><td>NH</td><td>BAA    </td><td>0</td><td>22</td><td>4</td><td>140       </td><td>79.3</td><td>-28.24646</td><td>colon</td><td>     </td><td>  </td></tr>\n",
       "\t<tr><th scope=row>PB146_A2</th><td>Recurrent    </td><td>604935 (146) (17-0236-382)                  </td><td>CD</td><td>F</td><td>NH</td><td>Refused</td><td>0</td><td>21</td><td>9</td><td>146       </td><td>76.8</td><td>-25.71356</td><td>colon</td><td>colon</td><td>NI</td></tr>\n",
       "\t<tr><th scope=row>PB17_236_122_A5</th><td>Recurrent    </td><td>(Hans' bx study) 17-0236-122\n",
       "SHARE - 600045</td><td>CD</td><td>M</td><td>NH</td><td>WC     </td><td>0</td><td>32</td><td>6</td><td>17_236_122</td><td>75.7</td><td>-27.38536</td><td>colon</td><td>colon</td><td>NI</td></tr>\n",
       "\t<tr><th scope=row>PB17_236_177_A6</th><td>Non-recurrent</td><td>17-0236-177                                 </td><td>CD</td><td>F</td><td>NH</td><td>WC     </td><td>0</td><td>18</td><td>6</td><td>17_236_177</td><td>81.5</td><td>-21.39823</td><td>colon</td><td>colon</td><td>NI</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 15\n",
       "\\begin{tabular}{r|lllllllllllllll}\n",
       "  & recur & subject\\_id\\_per\\_gc & disease & sex & ethnicity & race & smoking & age & months\\_surg\\_postop\\_colonoscopy & subject\\_id & TIN & PC1 & tissue & block\\_location & inflammation\\_status\\\\\n",
       "  & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <int> & <int> & <int> & <chr> & <dbl> & <dbl> & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\tPB12\\_770\\_A1 & Non-recurrent & 12-770                                       & CD & F & NH & WC      & 0 & 47 & 6 & 12\\_770     & 76.2 & -24.46333 & colon & colon & NI\\\\\n",
       "\tPB138\\_A8 & Recurrent     & 138                                          & CD & F & NH & WC      & 0 & 22 & 6 & 138        & 80.4 & -30.36808 & colon & colon & NI\\\\\n",
       "\tPB140\\_COLON & Recurrent     & 140 (604606)                                 & CD & M & NH & BAA     & 0 & 22 & 4 & 140        & 79.3 & -28.24646 & colon &       &   \\\\\n",
       "\tPB146\\_A2 & Recurrent     & 604935 (146) (17-0236-382)                   & CD & F & NH & Refused & 0 & 21 & 9 & 146        & 76.8 & -25.71356 & colon & colon & NI\\\\\n",
       "\tPB17\\_236\\_122\\_A5 & Recurrent     & (Hans' bx study) 17-0236-122\n",
       "SHARE - 600045 & CD & M & NH & WC      & 0 & 32 & 6 & 17\\_236\\_122 & 75.7 & -27.38536 & colon & colon & NI\\\\\n",
       "\tPB17\\_236\\_177\\_A6 & Non-recurrent & 17-0236-177                                  & CD & F & NH & WC      & 0 & 18 & 6 & 17\\_236\\_177 & 81.5 & -21.39823 & colon & colon & NI\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 15\n",
       "\n",
       "| <!--/--> | recur &lt;chr&gt; | subject_id_per_gc &lt;chr&gt; | disease &lt;chr&gt; | sex &lt;chr&gt; | ethnicity &lt;chr&gt; | race &lt;chr&gt; | smoking &lt;int&gt; | age &lt;int&gt; | months_surg_postop_colonoscopy &lt;int&gt; | subject_id &lt;chr&gt; | TIN &lt;dbl&gt; | PC1 &lt;dbl&gt; | tissue &lt;chr&gt; | block_location &lt;chr&gt; | inflammation_status &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| PB12_770_A1 | Non-recurrent | 12-770                                       | CD | F | NH | WC      | 0 | 47 | 6 | 12_770     | 76.2 | -24.46333 | colon | colon | NI |\n",
       "| PB138_A8 | Recurrent     | 138                                          | CD | F | NH | WC      | 0 | 22 | 6 | 138        | 80.4 | -30.36808 | colon | colon | NI |\n",
       "| PB140_COLON | Recurrent     | 140 (604606)                                 | CD | M | NH | BAA     | 0 | 22 | 4 | 140        | 79.3 | -28.24646 | colon | <!----> | <!----> |\n",
       "| PB146_A2 | Recurrent     | 604935 (146) (17-0236-382)                   | CD | F | NH | Refused | 0 | 21 | 9 | 146        | 76.8 | -25.71356 | colon | colon | NI |\n",
       "| PB17_236_122_A5 | Recurrent     | (Hans' bx study) 17-0236-122\n",
       "SHARE - 600045 | CD | M | NH | WC      | 0 | 32 | 6 | 17_236_122 | 75.7 | -27.38536 | colon | colon | NI |\n",
       "| PB17_236_177_A6 | Non-recurrent | 17-0236-177                                  | CD | F | NH | WC      | 0 | 18 | 6 | 17_236_177 | 81.5 | -21.39823 | colon | colon | NI |\n",
       "\n"
      ],
      "text/plain": [
       "                recur         subject_id_per_gc                           \n",
       "PB12_770_A1     Non-recurrent 12-770                                      \n",
       "PB138_A8        Recurrent     138                                         \n",
       "PB140_COLON     Recurrent     140 (604606)                                \n",
       "PB146_A2        Recurrent     604935 (146) (17-0236-382)                  \n",
       "PB17_236_122_A5 Recurrent     (Hans' bx study) 17-0236-122\\nSHARE - 600045\n",
       "PB17_236_177_A6 Non-recurrent 17-0236-177                                 \n",
       "                disease sex ethnicity race    smoking age\n",
       "PB12_770_A1     CD      F   NH        WC      0       47 \n",
       "PB138_A8        CD      F   NH        WC      0       22 \n",
       "PB140_COLON     CD      M   NH        BAA     0       22 \n",
       "PB146_A2        CD      F   NH        Refused 0       21 \n",
       "PB17_236_122_A5 CD      M   NH        WC      0       32 \n",
       "PB17_236_177_A6 CD      F   NH        WC      0       18 \n",
       "                months_surg_postop_colonoscopy subject_id TIN  PC1       tissue\n",
       "PB12_770_A1     6                              12_770     76.2 -24.46333 colon \n",
       "PB138_A8        6                              138        80.4 -30.36808 colon \n",
       "PB140_COLON     4                              140        79.3 -28.24646 colon \n",
       "PB146_A2        9                              146        76.8 -25.71356 colon \n",
       "PB17_236_122_A5 6                              17_236_122 75.7 -27.38536 colon \n",
       "PB17_236_177_A6 6                              17_236_177 81.5 -21.39823 colon \n",
       "                block_location inflammation_status\n",
       "PB12_770_A1     colon          NI                 \n",
       "PB138_A8        colon          NI                 \n",
       "PB140_COLON                                       \n",
       "PB146_A2        colon          NI                 \n",
       "PB17_236_122_A5 colon          NI                 \n",
       "PB17_236_177_A6 colon          NI                 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(coldata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "TRUE"
      ],
      "text/latex": [
       "TRUE"
      ],
      "text/markdown": [
       "TRUE"
      ],
      "text/plain": [
       "[1] TRUE"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "converting counts to integer mode\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "factor_cols = c('sex', 'recur', 'inflammation_status')\n",
    "for (col in factor_cols) {\n",
    "  coldata[,col] = as.factor(coldata[,col])\n",
    "}\n",
    "sex = coldata$sex\n",
    "source = coldata$source\n",
    "# tissue = coldata$tissue\n",
    "recur = coldata$recur\n",
    "inflammation_status = coldata$inflammation_status\n",
    "tin = coldata$TIN\n",
    "\n",
    "sample = rownames(coldata)\n",
    "\n",
    "#round counts to integers\n",
    "cts <- round(cts);\n",
    "\n",
    "#check if row names of coldata match colnames of cts\n",
    "all(rownames(coldata) == colnames(cts))\n",
    "\n",
    "#adjust for batch and sexBin\n",
    "design <- as.formula(~ recur)\n",
    "#create the dds object\n",
    "dds = DESeqDataSetFromMatrix(countData = cts, colData = coldata, design = design)\n",
    "\n",
    "vsd = vst(dds)\n",
    "mat = assay(vsd)\n",
    "#PCA of raw data\n",
    "pcaData <- plotPCA(vsd, intgroup=c(\"recur\"), ntop = 500, returnData=TRUE)\n",
    "percentVar <- round(100 * attr(pcaData, \"percentVar\"))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "\n",
    "# #save the plot\n",
    "# png(\"results/i/pca_source_raw.png\", width=1024, height=512)\n",
    "# ggplot(pcaData, aes(PC1, PC2, color=source, shape=recur)) +\n",
    "#     geom_point(size=5) +\n",
    "#     #geom_text(aes(label=sample),hjust=0, vjust=0) +\n",
    "#     xlab(paste0(\"PC1: \",percentVar[1],\"% variance\")) +\n",
    "#     ylab(paste0(\"PC2: \",percentVar[2],\"% variance\")) + \n",
    "#     coord_fixed() +\n",
    "    \n",
    "#     theme(text = element_text(size = 24))\n",
    "# dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Import genomic features from the file as a GRanges object ... \n",
      "OK\n",
      "\n",
      "Prepare the 'metadata' data frame ... \n",
      "OK\n",
      "\n",
      "Make the TxDb object ... \n",
      "Warning message in .get_cds_IDX(mcols0$type, mcols0$phase):\n",
      "“The \"phase\" metadata column contains non-NA values for features of type\n",
      "  stop_codon. This information was ignored.”\n",
      "OK\n",
      "\n",
      "'select()' returned 1:1 mapping between keys and columns\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library('GenomicFeatures')\n",
    "#use the GenomicFeatures package to create a transcript database (txDb) file from our GENCODE annotation file, and use it to pull out the chromosomal coordinates for each of our transcripts (including which chromosome it is on)\n",
    "txdb = makeTxDbFromGFF('../../data/gencode.v39.annotation.gtf', dbxrefTag = \"gene_id\", organism = 'Homo sapiens')\n",
    "k <- keys(txdb, keytype = 'TXNAME')\n",
    "tx2gene <- AnnotationDbi::select(txdb, keys=k, columns=c('GENEID', 'TXCHROM', 'TXSTART', 'TXEND', 'TXSTRAND'), keytype='TXNAME')\n",
    "# make a subset of the list containing only the Ensembl gene IDs and the names of the chromosomes, and create a subset for all the genes on chromosomes X and Y\n",
    "tx2gene_short = tx2gene[,2:3]\n",
    "tx2gene_short = unique(tx2gene_short)\n",
    "tx2gene_short_xychrom = tx2gene_short[tx2gene_short$TXCHROM==\"chrX\" | tx2gene_short$TXCHROM==\"chrY\",]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>58219</li><li>41</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 58219\n",
       "\\item 41\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 58219\n",
       "2. 41\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] 58219    41"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#now remove the rows from cts if they are present in 'GENEID' in tx2gene_short_xychrom\n",
    "cts <- cts[! match(rownames(cts), tx2gene_short_xychrom$GENEID, nomatch=0),]\n",
    "dim(cts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "keep\n",
       "FALSE  TRUE \n",
       "29854 31260 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "estimating size factors\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "-- replacing outliers and refitting for 413 genes\n",
      "-- DESeq argument 'minReplicatesForReplace' = 7 \n",
      "-- original counts are preserved in counts(dds)\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#DESeq on the raw data to identify control genes\n",
    "#filter genes for low expression with edgeR\n",
    "y <- DGEList(counts=counts(dds), group=recur)\n",
    "keep <- filterByExpr(y)\n",
    "table(keep)\n",
    "y <- y[keep,]\n",
    "dds <- dds[keep,]\n",
    "\n",
    "dds <- DESeq(dds)\n",
    "res <- results(dds)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\n",
       "FALSE  TRUE \n",
       "14354 16906 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "table(res$pvalue < 0.5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "set <- newSeqExpressionSet(counts(dds))\n",
    "set <- betweenLaneNormalization(set, which='upper')\n",
    "not_sig <- rownames(res)[which(res$pvalue > 0.5)]\n",
    "empirical <- rownames(set)[ rownames(set) %in% not_sig ]\n",
    "set <- RUVg(set, empirical, k=20)\n",
    "pdat <- pData(set)\n",
    "pdat$recur <- recur\n",
    "pdat$sex <- sex\n",
    "pdat$inf <- inflammation_status\n",
    "pdat$TIN <- tin\n",
    "pdat$source <- source"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "#write out the ruv factors to csv\n",
    "write.csv(pdat, file = \"../../results/recur/c/ruvg_runs/data_ruvg.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "formula_list = list(\n",
    "~ sex + W_1 + recur, \n",
    "~ sex + W_1 + W_2 + recur, \n",
    "~ sex + W_1 + W_2 + W_3 + recur, \n",
    "~ sex + W_1 + W_2 + W_3 + W_4 + recur, \n",
    "~ sex + W_1 + W_2 + W_3 + W_4 + W_5 + recur, \n",
    "~ sex + W_1 + W_2 + W_3 + W_4 + W_5 + W_6 + recur, \n",
    "~ sex + W_1 + W_2 + W_3 + W_4 + W_5 + W_6 + W_7 + recur, \n",
    "~ sex + W_1 + W_2 + W_3 + W_4 + W_5 + W_6 + W_7 + W_8 + recur, \n",
    "~ sex + W_1 + W_2 + W_3 + W_4 + W_5 + W_6 + W_7 + W_8 + W_9 + recur, \n",
    "~ sex + W_1 + W_2 + W_3 + W_4 + W_5 + W_6 + W_7 + W_8 + W_9 + W_10 + recur,\n",
    "~ sex + W_1 + W_2 + W_3 + W_4 + W_5 + W_6 + W_7 + W_8 + W_9 + W_10 + W_11 + recur,\n",
    "~ sex + W_1 + W_2 + W_3 + W_4 + W_5 + W_6 + W_7 + W_8 + W_9 + W_10 + W_11 + W_12 + recur,\n",
    "~ sex + W_1 + W_2 + W_3 + W_4 + W_5 + W_6 + W_7 + W_8 + W_9 + W_10 + W_11 + W_12 + W_13 + recur,\n",
    "~ sex + W_1 + W_2 + W_3 + W_4 + W_5 + W_6 + W_7 + W_8 + W_9 + W_10 + W_11 + W_12 + W_13 + W_14 + recur,\n",
    "~ sex + W_1 + W_2 + W_3 + W_4 + W_5 + W_6 + W_7 + W_8 + W_9 + W_10 + W_11 + W_12 + W_13 + W_14 + + W_15 + recur,\n",
    "~ sex + W_1 + W_2 + W_3 + W_4 + W_5 + W_6 + W_7 + W_8 + W_9 + W_10 + W_11 + W_12 + W_13 + W_14 + + W_15 + W_16 + recur,\n",
    "~ sex + W_1 + W_2 + W_3 + W_4 + W_5 + W_6 + W_7 + W_8 + W_9 + W_10 + W_11 + W_12 + W_13 + W_14 + + W_15 + W_16 + W_17 + recur,\n",
    "~ sex + W_1 + W_2 + W_3 + W_4 + W_5 + W_6 + W_7 + W_8 + W_9 + W_10 + W_11 + W_12 + W_13 + W_14 + + W_15 + W_16 + W_17 + W_18 + recur,\n",
    "~ sex + W_1 + W_2 + W_3 + W_4 + W_5 + W_6 + W_7 + W_8 + W_9 + W_10 + W_11 + W_12 + W_13 + W_14 + + W_15 + W_16 + W_17 + W_18 + W_19 + recur,\n",
    "~ sex + W_1 + W_2 + W_3 + W_4 + W_5 + W_6 + W_7 + W_8 + W_9 + W_10 + W_11 + W_12 + W_13 + W_14 + + W_15 + W_16 + W_17 + W_18 + W_19 + W_20 + recur)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "converting counts to integer mode\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "estimating size factors\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "converting counts to integer mode\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "estimating size factors\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "converting counts to integer mode\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "estimating size factors\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "converting counts to integer mode\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "estimating size factors\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "converting counts to integer mode\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "estimating size factors\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "converting counts to integer mode\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "estimating size factors\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "converting counts to integer mode\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "estimating size factors\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "converting counts to integer mode\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "estimating size factors\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "3 rows did not converge in beta, labelled in mcols(object)$betaConv. Use larger maxit argument with nbinomWaldTest\n",
      "\n",
      "converting counts to integer mode\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "estimating size factors\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "7 rows did not converge in beta, labelled in mcols(object)$betaConv. Use larger maxit argument with nbinomWaldTest\n",
      "\n",
      "converting counts to integer mode\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "estimating size factors\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "7 rows did not converge in beta, labelled in mcols(object)$betaConv. Use larger maxit argument with nbinomWaldTest\n",
      "\n",
      "converting counts to integer mode\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "estimating size factors\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "fitting model and testing\n",
      "\n"
     ]
    }
   ],
   "source": [
    "\n",
    "\n",
    "#create an empty dataframe to store the number of significant genes\n",
    "deg_counts = data.frame(matrix(ncol=20, nrow=0))\n",
    "colnames(deg_counts) = c(\"formula\", \"sig_genes\")\n",
    "\n",
    "for (x in seq_along(formula_list)) {\n",
    "    design_formula = formula(formula_list[[x]])\n",
    "    dds1 <- DESeqDataSetFromMatrix(countData=cts, colData=pdat, design = design_formula)\n",
    "    y <- DGEList(counts=counts(dds1), group=coldata$recur)\n",
    "    keep <- filterByExpr(y)\n",
    "    table(keep)\n",
    "    y <- y[keep,]\n",
    "    dds1 <- dds1[keep,]\n",
    "    dds1 <- DESeq(dds1)\n",
    "    res <- results(dds1)\n",
    "    res <- res[order(res$padj),]\n",
    "    #get the number of significant genes\n",
    "    sig_genes <- res[which(res$padj < 0.05),]\n",
    "    #save the number of significant genes to the dataframe\n",
    "    deg_counts = rbind(deg_counts, c(x, nrow(sig_genes)))\n",
    "    #save the results to a csv file\n",
    "    write.csv(res, paste0(\"../../results/recur/c/ruvg_runs/res5_\", x, \".csv\"))\n",
    "    \n",
    "}\n",
    "\n",
    "head(deg_counts, 10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 10 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>formula</th><th scope=col>sig_genes</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>11</th><td>11</td><td>283</td></tr>\n",
       "\t<tr><th scope=row>12</th><td>12</td><td>318</td></tr>\n",
       "\t<tr><th scope=row>13</th><td>13</td><td>416</td></tr>\n",
       "\t<tr><th scope=row>14</th><td>14</td><td>404</td></tr>\n",
       "\t<tr><th scope=row>15</th><td>15</td><td>519</td></tr>\n",
       "\t<tr><th scope=row>16</th><td>16</td><td>626</td></tr>\n",
       "\t<tr><th scope=row>17</th><td>17</td><td>700</td></tr>\n",
       "\t<tr><th scope=row>18</th><td>18</td><td>744</td></tr>\n",
       "\t<tr><th scope=row>19</th><td>19</td><td>813</td></tr>\n",
       "\t<tr><th scope=row>20</th><td>20</td><td>787</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 10 × 2\n",
       "\\begin{tabular}{r|ll}\n",
       "  & formula & sig\\_genes\\\\\n",
       "  & <int> & <int>\\\\\n",
       "\\hline\n",
       "\t11 & 11 & 283\\\\\n",
       "\t12 & 12 & 318\\\\\n",
       "\t13 & 13 & 416\\\\\n",
       "\t14 & 14 & 404\\\\\n",
       "\t15 & 15 & 519\\\\\n",
       "\t16 & 16 & 626\\\\\n",
       "\t17 & 17 & 700\\\\\n",
       "\t18 & 18 & 744\\\\\n",
       "\t19 & 19 & 813\\\\\n",
       "\t20 & 20 & 787\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 10 × 2\n",
       "\n",
       "| <!--/--> | formula &lt;int&gt; | sig_genes &lt;int&gt; |\n",
       "|---|---|---|\n",
       "| 11 | 11 | 283 |\n",
       "| 12 | 12 | 318 |\n",
       "| 13 | 13 | 416 |\n",
       "| 14 | 14 | 404 |\n",
       "| 15 | 15 | 519 |\n",
       "| 16 | 16 | 626 |\n",
       "| 17 | 17 | 700 |\n",
       "| 18 | 18 | 744 |\n",
       "| 19 | 19 | 813 |\n",
       "| 20 | 20 | 787 |\n",
       "\n"
      ],
      "text/plain": [
       "   formula sig_genes\n",
       "11 11      283      \n",
       "12 12      318      \n",
       "13 13      416      \n",
       "14 14      404      \n",
       "15 15      519      \n",
       "16 16      626      \n",
       "17 17      700      \n",
       "18 18      744      \n",
       "19 19      813      \n",
       "20 20      787      "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "tail(deg_counts, 10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA0gAAANICAMAAADKOT/pAAADAFBMVEUAAAABAQECAgIDAwMEBAQFBQUGBgYHBwcICAgJCQkKCgoLCwsMDAwNDQ0ODg4PDw8QEBARERESEhITExMUFBQVFRUWFhYXFxcYGBgZGRkaGhobGxscHBwdHR0eHh4fHx8gICAhISEiIiIjIyMkJCQlJSUmJiYnJycoKCgpKSkqKiorKyssLCwtLS0uLi4vLy8wMDAxMTEyMjIzMzM0NDQ1NTU2NjY3Nzc4ODg5OTk6Ojo7Ozs8PDw9PT0+Pj4/Pz9AQEBBQUFCQkJDQ0NERERFRUVGRkZHR0dISEhJSUlKSkpLS0tMTExNTU1OTk5PT09QUFBRUVFSUlJTU1NUVFRVVVVWVlZXV1dYWFhZWVlaWlpbW1tcXFxdXV1eXl5fX19gYGBhYWFiYmJjY2NkZGRlZWVmZmZnZ2doaGhpaWlqampra2tsbGxtbW1ubm5vb29wcHBxcXFycnJzc3N0dHR1dXV2dnZ3d3d4eHh5eXl6enp7e3t8fHx9fX1+fn5/f3+AgICBgYGCgoKDg4OEhISFhYWGhoaHh4eIiIiJiYmKioqLi4uMjIyNjY2Ojo6Pj4+QkJCRkZGSkpKTk5OUlJSVlZWWlpaXl5eYmJiZmZmampqbm5ucnJydnZ2enp6fn5+goKChoaGioqKjo6OkpKSlpaWmpqanp6eoqKipqamqqqqrq6usrKytra2urq6vr6+wsLCxsbGysrKzs7O0tLS1tbW2tra3t7e4uLi5ubm6urq7u7u8vLy9vb2+vr6/v7/AwMDBwcHCwsLDw8PExMTFxcXGxsbHx8fIyMjJycnKysrLy8vMzMzNzc3Ozs7Pz8/Q0NDR0dHS0tLT09PU1NTV1dXW1tbX19fY2NjZ2dna2trb29vc3Nzd3d3e3t7f39/g4ODh4eHi4uLj4+Pk5OTl5eXm5ubn5+fo6Ojp6enq6urr6+vs7Ozt7e3u7u7v7+/w8PDx8fHy8vLz8/P09PT19fX29vb39/f4+Pj5+fn6+vr7+/v8/Pz9/f3+/v7////isF19AAAACXBIWXMAABJ0AAASdAHeZh94AAAgAElEQVR4nO3dd2DU5P8H8E9bWsreggqCCgqKggPcA0XF8dRCAVnigMpSQQFFkS+iIiICKiBDWf5QFGUjIFsUEBnK3kN2ewgyChTa5pfkVi7jend5YnOX9/sPLnny9NNcmhd32SQgCGI6VNAzgCCxEEBCEA4BJAThEEBCEA4BJAThEEBCEA4BJAThEEBCEA757yGdPsE9Jy/yr3nixIXcfy2oejbLgqInci1ZAtmnLCh6JteKJXDykgVFT5zPbxX4twAh/evinuOX+Nd0uS4I/1hQ9UyWBUVdgiVLIPukBUVPCWctqHo8x4Ki4ipwIniHE4CUfwAJkACJQwAJkACJQwAJkACJQwAJkACJQwAJkACJQwAJkACJQwAJkACJQwAJkACJQwAJkACJQwAJkACJQwAJkACJQwAJkACJQwAJkACJQwAJkACJQwAJkACJQwAJkACJQwAJkACJQwAJkACJQwAJkACJQwAJkACJQwAJkACJQwAppiEdWTnrp7UZwfsAEo8AUgxDmvV0CRJT4dmVwXoBEo8AUsxC2vE4eZOYfti4HyDxCCDFKqQ/qpEid+0x7AhIPAJIMQpp17UUkIePGfUEJB4BpBiF1IxU6WfUE5B4BJBiE9JStSMqvcugKyDxCCDFJqT2Gkg01KArIPEIIMUmpCpaSI0MugISjwBSTEI6qHVE1xj0/S8gbfm0+zMv9p102t9ydGynpq17/XQpWAsgAVIBQ1qnA6mYQV/rIZ0ZxNxptdDbND/N3dLlmHELIAFSQUPaqgOpgkFf6yG9z1hqv7Gfd2YsZY27ZVkKY73GjWjLWPpZoxZAAqQCh5SRqIV0s0FfyyHNFXEcEF/zZjDWJktqOdOaPb1cfM3ux9gYQb8FkFyAVOCQXPdqIXU06Go5pK6MbXYPDWBsnfT6HWPj5Iastiz1pH4LILkAqeAhDdBCmmXQ1XJITVizPPfQQsa+k167sFRP0W8Zm6vfAkguQCp4SH9XVDtK/sGgq9WQ8lJZC8/gcsa+FF8yGevpadnDWD/dFkCSAkgFDck1SuUoIZ5Stuv2tPwTqTNj+91DI92fNmsY+8ozLTeNtdNtASQpgFTgkFxdAiGNmFuLSg/O1OloOaT5jHU/Iw2se5q1knbJTWNsjndiR5ZyQa8FkKQAUsFDyuylYFTkS5fraP+idNcKbUfrd39/yVirMXN/fJ+xljuk8YmMrfBOe5sxl14LIEkBpIKH5PqCSngdPea+RHZtA0p8VXOJ339wZsMCzwHZzsfl0VGenXdS+jF2UK9FEDIWijl0mn9yLKh5+pJwxoKq5y9YUPS0YM0SOGtB0XOCJUsgN4y+Ry5PWrv4tcduuuWJvn/4GidWoGtnqzpeFPJZAmfMQlrXzgOJDZZPEvqMsY3eaR8ytkuvRRCW3CZmdWS/EUF45TXqo9N68tX4uGddOhOCJNc3FBmk5U+ztEm7slxr3mTsFUnSSMbWeyf2k/dEaFsE4e8JYvae5Z6sXP41z569JGRZUDX7ogVFzwqWLIGccxYUvSBkW1A1nFVgTWKVDN0JC2vSZWN8Yxnfdk57/Nn31wb9teYguRqzNM9euzGMDRRfxjO2yjtV3CLK0GvxBNtIFgTbSKFvI2XeQxMNJh3pU5juWSUP7n29qGcrqt4c41omt5HGMfaDZ/BSOkvJkPfR+Y65dmYsS68FkKQAUkFD+oIaGE/84wFK7nnY5fq1mmLH3ut6u8blmIT0DmM7vcPitpC41bOasQmehrxm7DlBrwWQpABSAUPaVynp92DTx5ajWnOXlww40vSsUWeTkF5TfFWbwNgSQchgrLen4QBjfQW9FkCSAkgFDKkDdQ/eYVfbuLhiqpMfBhj0NQnpQ/lTyJ2+jG0SXzqxNM+1EtMZmyXotgCSC5AKGtKviZX/zq/PrDLqs/GKb9PvaRLSfMZ6eU5a3Z3KnpGugJ3E2FS5IacjS5X3IWpbAMkFSAUMKfMe+jrfTjsKa84Pf0W/p0lIl7oyNlQ+frS5PWPzpIHTzVkz6VhR3mjGhgv6LYDkAqQChjQi2J4Gbz7XOKIq+j3NntlwsCljrfpPGN6TMfahu2khY01HL54ttrQ/ZdQCSIBUsJDy29PgTnMtJFqv29P0KUJ7enpObGjytfd01Gmp7pZOBwXDFkACpAKFlO+eBjn36UDSP5hk/ly7vKXvP9f4mc4jj/ib9g5PT2vVfeaFYC2ABEgFCCmUPQ1i6upAmqzbE/e1CyGAFGuQMu8MYU+DmIY6kBbr9gSkEAJIsQYppD0NYtK1jhL0bw8OSCEEkGIM0t6Q9jSI+UEL6U79noAUQgApxiC9RD1CK3ZEe3vwz/R7AlIIAaTYghTingYpX6gd1TR4FhkghRBAiilImffQ/4VaLTM10FHRpQYdASmEAFJMQRoe4p4GOQcCDiUVmWTUD5BCCCDFEqS9FUPc0+DOkZf8twi/epFhN0AKIYAUS5DSqWd4FZc3Li4zunHAEeNOgBRCACmGIC1PrHwg3JqH502e8OOfQbsAUggBpNiBlHl36HsaFMGjL3kEkGIHUlh7GvwBJB4BpJiBFOaeBl8AiUcAKWYghb2nwRNA4hFAihVIywuFv6dBDiDxCCDFCKTMu8nwiGrwABKPAFKMQBoW2Z4GFyDxCSDFBqRI9zS4AIlPACk2ILWPcE+DC5D4BJBiAlLEexpcgMQngBQLkCLf0+ACJD4BpFiANIweirwqIPEIIMUApL0Vk1ZHXhWQeASQYgBSe3rDRFVA4hFAimZIGTuly4iWF6p6yERVQOIRQIpaSHPbXZVAcZe3mHwXfWOmKiDxCCBFKaQ/G/lvt3C3qaqAxCOAFJ2Q5lZQ3rdkgpmqgMQjgBSVkFaWCriTVtJ0E1UBiUcAKRohHamhurdj2R2RVwUkHgGkaIQ0QHO34Q6RVwUkHgGkaIR0jQZSkYhPtQMkLgGkKIS0QuOIIrp/kDuAxCOAFIWQRulAivzcBkDiEUCKQkj9dCC1jbgqIPEIIEUhpPd1ID0XcVVA4hFAikJIY3Qg9Yq4KiDxCCBFIaQ1OpCmRFwVkHgEkKIQkquWxlGJwxFXBSQeAaRohPS5BlK3yKsCEo8AUjRCOnazytHleyOvCkg8AkjRCMm1KjHAUfLPJqoCEo8AUjRCOvIIFVE4KjPVTFVA4hFAikJIxxrTfRtb+hw9ttZUVUDiEUCKPkiZz9Jt+1yuFd3rlKHitTrNN1kVkHgEkKIPUheqtdPTksGhKiDxCCBFHaSedPVmnlUBiUcAKdogDaAr1nOtCkg8AkhRBmlYXLmVfKsCEo8AUnRBmlCo5GLOVQGJRwApqiD9kFRkDu+qgMQjgBRNkH4tmvQ996qAxCOAFEWQlpVOGMe/KiDxCCBFD6TfK8R9xr8qIHEJIEUNpLWVaDD3oi5A4hNAihZIG6tSH+OHMZsIIPEIIEUJpB3XU7cgTzU3EUDiEUCKDkh769KLQZ5qbiaAxCOAFBWQDt5FzTIAiUsAybmQjjSkRkeDPNXcVACJRwApCiAdS6X7pKfE2h3SiVUz1+YAEscAkmlIGfP+91Lb7l9IFx5ltqHb9ktt9oS0f8Zi+fX4c9LluhUHc6UESBbEQZCOfHy5+zLyQk+vdnX2XshnR0i76xN1kwayrvdc+d7kEiDxCiCZg7Shjv/OJkmP0NVb3M02hLS4BHkgveyb488BiVcAyRSkdRUD76LqvZDPfpBybyQPpOOJRJf938ax4ifplRcAiVMAyQykv9V3JP7KM8F+kCaKc9fw1yxxaLQ4tFB83St+RI0CJE4BJDOQeqrvR1zOcx9V+0FqQVQ/Vx56mOhmeeA1oucAiVMAyQSk3cU0d/Z+yz3FfpDqE30nD1xIJuorDy0kug+QOAWQTEAaqXFEtdxT7AepPNFGeWCZOJe/yEO7xI0kQOIUQDIBqYUWErnvv2U/SCWJ9sgD7xIlu/cxbCdKBCROASQTkO7TgeS+m6r9IN1CtEweqEf0kLtpBlFZfpAu5vBPngU1c/IEK6rm5lpRVbBmCVgxr7lC5FXr6kCa455kyR8r31XAf4BVA6mpuP0mve6JI/rY3fQm0Y38IOETyYI45BPpQR1Ii+Qp9vtEGkhUeocg5ImgaIfcsieZ6FlA4hRAMgGprdZR/A55iv0gnSonfpF7Y9QT4jzeK41nz6kpDk4HJE4BJBOQvtZCus09xX6QpI8k74emOLZUfv5ZLY4n2wGSBXEIpAPlNZAGuKfYEFLu23HuWewljc2Xhq78k58jQLKgqFMguQaoHVX1PLLchpAEYWZ9ouS6PwleSM+c0OkESJEFkMxAOqraAZ400zPBlpAE4cLRXM/Qpndn7OPJCJAAyczZ3zsDz1r93NtuU0hWBpAsiGMgufbV9zMqO8XXDEg8AkjOgXSsZtxN8TKj8t32+JttC+nEz+MGfTBBEPIAiW8AyRykgdTCtXXch72HzQ14SKw9IZ18o6a8566hICyq9+kRQOIYQDIFaV+F5L/02m0JaURZz3dQCRJR4leAxC+AZArSq9RDt92OkPz76t2QiLrnansBUmQBJDOQ/kqusE93gg0hyadhVPtg6mYZ0tHm0pbdeEDiFUAyA6kxGTwGyX6QLlYmKjwkW5okQRKEdZWIqpwHJE4BJBOQ5sfdcEx/iv0gjRc/gL50T3JDEnYmEw0DJE4BJBOQ7qApBlPsBymVqJkQAEnoStQGkDgFkCKHNJYeMZpkP0g3Ec1XQVpEVB+QOAWQIoZ0+OqEX42m2Q9SMaK9Kkj7iEoBEqcAUsSQ3qUXDafZD1IZohUqSFsAiV8AKVJIu8oU32o40X6QbiAaqoL0LdF1gMQpgBQppPbUx3ii/SC9RFTnQiCkhkQtAIlTAClCSKsSrzxgPNV+kBYTUYcASGPFlsmAxCmAFCGkRjQmyFT7QZI+fyhllw9S7tAEourZgMQpgBQZpBl0a2aQyTaEtEu6wURS04+XEN1/YPWQa8SxhPn8HAGSBUVjH1JGHZoTbLoNIQkbLlPdYCKO6+nfgGRBYh7SMEoJOt2OkARXywBHVy/j6QiQLCga85AOVk76I2gHW0IShG2v1fAoKtJgEsftI0ACpIggvUldgnewKSQx//wybdzEWWsv8lUESIAUAaStxUvvDN7DvpAsCyBZkBiH1Np7Q1XDABKPAFJsQ/olofqRfLoAEo8AUmxDakD/l18XW0I6t3rkR6oAEq8AUtiQvqW78+1jR0iTS2pu+c/zcwuQLEgsQzpWM35hvp1sCKm/lhEg8QsghQvpY2qRfyf7QTqQLMEpdn3NgAASrwBSmJD2XZb8Z/697AfpVZHRw5v436kYkNwBpDAhGd0SMjD2g/QA0b0WMgIkQAoL0oYiBreEDIz9IF1O9JuVjgDJgqIxDKkJfRpKN/tBKkdxWYAkBZBsAOlnw1tCBsZ+kB6nopY6AiQLisYuJONbQgbGfpCGEh0GJCmAVPCQxlLD0DraD9KZMtQPkKQAUoFDOnx1wvLQetoPkjCBEuYBkguQbACpH70QYk8bQhL6UMnPzwESIBU4pF1lim8JsasdIQnfFKeKPUYv2LLdH0DiFUAKHVI6vRNqVxtCqlatmvasVUDiFUAKGdKapCuC3BIyMDaEpHfOKiBxCyCFDOlxGh1yVUDiEUCKRUgzg98SMjA2hLRWL4DEK4AUIqSMusFvCRkYG0KyOoBkQWIJ0sFfp876/ZhrOLEwqgISjwBS7ECa+lRRaVui3DMVk1aHURWQeASQYgXS9ob+7fKbDodR1baQTvw8btAHEwSB/6VJgGRBYgTSqirKPVz1d4de1Z6QTr5RM056Jw0FYVG9T48AEscAkjGkHdUC9xU/FNIVFHJsCWlEWc8baSg/0TyR68MoAMmCojECqbH6qEu/kKvaEdIA3/twQyLqngtIvAJIhpAWaQ5flgn5y50NIX0tvYNqH0zdLEM62jxeHB0PSLwCSIaQXtCeCBDSZeZS7AfpYmWiwkOkR7l4niG7rhJRlfOAxCmAZAjpCi2kx0Otaj9I48XZ/9I9yfMw5p3JRMMAiVMAyQjSfq0jqh5qVftBSiVqJgRAEroStQEkTgEkI0hrdCCVCLWq/SDdRDRfBUncCKwPSJwCSEaQtuhAqhhqVftBKka0VwVpH1EpQOIUQDKCdCxRC6luqFXtB6kM0QoVpC2AxC+AZLiz4R4tpC6hVrUfpBuIhqogfUt0HSBxCiAZQhqghTQ31Kr2g/QSUZ0LgZAaErUAJE4BJENIf1dSO7o/5Kr2g7RYnP8OAZDGii2TAYlTAMn4FKHRKkdFfgm5qv0gSZ8/lLLLByl3aAJR9WxA4hRACnL298uBkMaEXtWGkHaVF99CUtOPl4ifrAdWD7lGHEuYz88RIFlQNFYgZb6tYFR0fBhVbQhJ2HCZ6hM2juvp34BkQWIEkss123dF0lPhXCBrS0iCq2WAo6uX8XQESBYUjR1IrtoJ6Y/Vrfdk37AY2RSSIGx7rYZ3e6/BJP3to7yV77dLe3mQ4h6sR8d2atq610+XgrUAEiAFhbSQHo2oqk0hifnnl2njJs5ae9FgsustJidlhPdi9Plp7pYuxwTDFkACpOCQnqeJEVW1L6TgOZXO2HPDx/ZNYcyzZ3yZONhr3Ii2jKWfNWoBJEAKDulgqQpHIqoarZD6MTYiR3zd1oyl7JMazrRmTy8XX7PFKWME/RZAcgFScEgj6JXIqkYppI2M9XF/pZvH2CTp9TvGxskNWW1Z6kn9FkByAVJwSHfTisiq2hXSse8Ckxk4+UPGdruHznd9+WPptQtL9RT9lrG5+i2A5AKkoJDWxt0RYVW7QTo9Mu1t6VV9H4rAPeDZTVnnwJ/LZKynZ3APY/10WwBJCiAFgdSNhkVY1V6QLvQuSdQxf0jbxC2kQEhrGPMetc1NY+10WwBJCiAZQzpaqdj+CKvaCtLO2pIZH6RS8W5EFZp+ui7wflwLGftOEHYtmLpgv6dlGmNzvFM7spQLei2AJAWQjCFNomcjrWonSPvdZ2e86oUk5Bz8+knxtba/tyeTGPt5axf5GFGPg3LLRMZWeKe+zZhLrwWQpACSMaTHaX6kVW0EKedW6fTUJnNzfZCkrKxOdFeWCtKXjI1MdR9sZWmLpZZRjK3zTu3H2EG9FkHY0V/MzvP8k2tBzfM5wgULql68ZEHR84IlSyA324KiFwWjJfB34g2Rl82L/EeNk/8qoANpiGgnebpnxAdJ+Kce0bMqSMPlz6I1x09vFIk0PSq2fMbYRu/UDxnbpdciCEtuE7NaQBC9DKQhBT0L4ca/yeODlHOl6Mi3T8EPSThUkWhp4I9/Ljoa7j6ONIaxPuLLSMbWe6eKuPbrtQjC6a1iMk5yz785/GuePHlROGVB1azzFhQ9KVyyourF0xYUPSucM5hyXdKuiKv+mxvxjwZJdn6rwCktpJ+8m0dqSMJocTMpENJoxl7wfKidbytv/4xnbJV3qrhFlKHX4gm2kSxILGwjzaGUyKvaZxupu/iBdFgX0kXxI8m3uSNnImOfe4c/Yux3eR+d75hrZ8ay9FoASQogGUFqQVMir2ofSPcQNfZTUUISniN6KwDSfMameYe/ZmyhIKxmbIKnIa8Ze07QawEkKYBkAGl/sStCfxySJvaBVJmolwGkL4juC4C0hbEvvcPD5a9wGYz19jQcYKyvoNcCSFIAyQDSYOphoqp9IBUnUlxR/nvdunV9I1OJqgdAutSC9fAOd2VMeqRfJ5bmuVZiOmOzBN0WQHIBkiGkW+PWmKhqH0iJRDME/cwkKhrY8jljv7mHfmOsq/Q6ibGpckNOR5bq0m8BJBcgGUFaTg+YqWofSBWIRhtAGkNUOrAlM421kCWtasVSdkgDp5uzZtKxorzRjA0X9FsAyQVIRpA6hHPzLW3sA+l6ojcNIPUgukHVNJcx9trIUd2Z96Ij6fy7pqMXz+7JWPtTRi2ABEgGkA6XK3PITFX7QGqq1eJNbaJH1W0/ek4RajbXe8+GaZ6WTgcFwxZAAiR9SGMp3VRV+0AaSkRrdR39JU75WNO6f3THZs06j/YfaBX2Dk9Pa9V95oVgLYAESLqQGtBSU1XtA2l3PNHdeYJOnhAh7dIlFlEAyYJEO6QNCSE/CUk/9oEkSFdM9NNZ8aU7m9/PzxEgWVA06iG9QYPMVbURpLWFRDEfatb7SWJz/DpNMyBFGEDSgZRZNXm3uao2giT8T7qs79G/A9Z61/NS48scHQGSBUWjHdIP1NxkVTtByu0goUlqtcx7d9VLv79URGpigfccBiQzASQdSKk002RVO0HyfCYRJd/Xrsf7b7ZrWNw9+uQ5no4AyYKiUQ5pZ+GqmSar2guSMLU6qVP4U91deYAUYQBJC+lDesdsVZtBEi6OuCKAUdHOu/kyAiRA0kCqXWiT2ap2gyQIeX+8e0dRt6IrW395nDcjQAIkNaSF9JjpqvaDJOfsnlVr95/hjwiQAEkL6Xn62nRVm0KyMoBkQaIZUsSPclEGkHgEkKIZ0vBIH+WiDCDxCCBFM6S7In2UizKAxCOAFMWQIn+UizKAxCOAFMWQukb8KBdlAIlHACl6IZl4lIsygMQjgBS9kCZRWx5VAYlHACl6IT1OP/OoCkg8AkhRC2lrYi0uVQGJRwApaiH1oQ+4VAUkHgGkqIVUI2k7l6qAxCOAFK2QZpt5lIsyNoS0ffsBVUuutgmQIg0gKSGZepSLMjaERNRQ1XIpnm4CJE4BJAUkc49yUSYqIAmlqSQgcQogKSANpp6cqkYFpL/jKBGQOAWQFJBujVvLqaqtIL3RSApRuUYBebA00VWAxCmA5Ie0nB7kVdVWkB7U3PnEl2aAxCmA5IfUgb7kVTVKIJXZBEicAkg+SGYf5aKMrSANai+F6Ir2gXlp4N9aD4AUWQDJB+krk49yUcZWkDwt2r12XANIFiQqIZl9lIsygMQjgBSNkDYk3MKvqg0hffLJFECSAkjWQnqDPuFX1YaQrA4gWZAohGT+US7KABKPAFIUQvqBnuFY1ZaQzq0e+ZEqgMQrgOSBlEqzOFa1I6TJJbUHkgCJVwDJDWlHkulHuShjQ0j99Y7IAhKvAJIbUn/zj3JRxn6QDiRLcIpdXzMggMQrgOSGdKP5R7koYz9Ir4qMHt7E+eFigOQLIMmQFnB4lIsy9oP0ANG9FjICJECSID3H4VEuytgP0uVEv1npCJAsKBp1kLg8ykUZ+0EqR3FZgCQFkKyDNJxe5VvVfpAep6KWOgIkC4pGHaS7aCXfqvaDNJToMCBJASTLIP0edyfnqvaDdKYM9QMkKYBkAaQNQ3q+9PpHbbk8ykUZ+0ESJlDCPEByAZIFkObeE+c+wB8/nnNlG0IS+lDJz88BEiDxhnSsk+JcmdQDXGvbEZLwTXGq2GP0gi3b/QEkXnEupIynAs46u/0gz+I2hFStWjXtWauAxCvOhdRNtU415lnchpB0byMESLziWEjLE9Ur1WSO1QGJRwApCiA9pVmpbuJY3YaQ1uoFkHjFqZD2F9b+97yKX3kbQrI6gGRB7A9pss73nAH8ygMSjwCS/SEN0oHUhV/56IA05auFgMQpToXUWwdSK37lowNSVaoFSJziVEhDdCBxPAM8GiDlTSdKBiROcSqkH3QgxfgNInO/eLRmNUUqiG+5PCBxilMhHSqmcRT3J7/yNoR0WrvDn4gBEqc4FZKrhWal4nkphQ0hNdZxdMNOQOIUx0Jan6Req37iWN1+kHbGExV/dmCfGkR3f/LJJ/0ax1Pcj7n8HAGSBUWjAZJLfcPEl3gWtx+k14hKrBNfM4pQNblhQWm6meddHADJgkQDJFf7AEeNjvKsbT9IjxB1lQfSiPbJA+OJXgMkXnEwJFd/xbe7l49xLW0/SDcQuY++fkg0w930MCXtBSROcTIk17qqbkVJjZZxrmw/SCWJtskD04gGCt4hjrdxACQLEiWQDhSr9P3gfp9M2su5rh0hJRPtkQf+ImrnbtpDdAcgcYqjIY2kropnyPKM/SBdSfSLPHCGqI67Ka8YVQQkTnE0pIdouWMg3UfU3z1UheJPyQMXE3CKELc4GdK2QrVdjoH0DlFFlzz0BNFX8sBawkmr3OJkSB9QP+dA2hRHdNWUfwR5t13Vk+Jr3lNEqfwgnfmXe07l8K/5778XhdMWVD133oKi/wqWLIFLnP9Wt8Rv+/ffLMGKJXAq14Ki+a8Cp40hCU2k/ZMVxIHdCUTXTlwzraE4/ik/SBcsSJ4VRXOFbAuqXrpkQdELQq4VVXP5LoAN9LD470XBkiVgySqQk+8qEATS8Wu9Z3u/4jt6Vj2bHyR8tbMgUfHV7jX5RsVO+WonCCe6FnJDOnWXx1HppfwcAZIFRaMCUuZVRfa5nARJpLTgO/n1XI/SIqPEJw5ydARIFhSNCkizKU16cRIkf8799dNfHL/WAZKDIT3nviOkMyFZEECyIFEA6XDZcvLTLp0JieeFSIAkxbGQxlO6/OokSIc/S093Dy0qUf/dfYDEMY6F9BQtkF+dA+lsh8JEt3sgEVGx4XmAxC1OhbQr6Rr3gGMgnb5H2uOtgET0MM/njgGSBbE/pMHUyz3gGEidJTqV33GPZPSpJ42+CEi84lRId8atdQ84BdKWeKJKsxUNi64mitsMSJziUEjr4+p7hpwC6R1xo2h/QMuhkkSvAxKnOBTS2zTIM+QUSPW015UPJLoNkDjFoZBqJu3wDDkFUiWipaqm33DLYn5xJqRF9IR30CmQkrx34fLnIFEiIHGKMyF1oHHeQadAupJotarpD6LLAYlTHAnp2GUlD3qHnQLpPqJ3VE19iO4CJE5xJKTv6VnfsFMgfUJUYldAy/6SRB8AEqc4ElJTmukbdgqko8WJKq9QNPx5DVHhfYDEKU6E9Hexyhm+EadAEj6VTmW465uT8kjWz42k0d78HAGSBUVtDmkEdfOPOAaS0MV9ffXaYagAACAASURBVHmx6x+8pZx7MPUSIPGKEyE1oOX+EedAEgYXDXj+RtwrXK+RBSQLYmtIWwvdrBhzECRh78ulfYwKNVnJkxEgOQ/Se/SeYsxJkATh/LKBzz/9YKOWb07L5MsIkJwH6eb4jYoxZ0GyMIBkQewMaQU1UI4CEiDxiPMgdaMRylFAAiQecRykzKuK7FeOxzyk4WIyPa+aABKvOA7STGoaMB7zkKRddGs9r5oAEq84DtKz9H3AOCABEo84DdLhMuWPBjTEPKTKYjZ4XjUBJF5xGqTx1CGwIeYh/VcBJAtiX0hP0MLABkACJB5xGCTffSF9ASRA4hGHQRpEb6taYh7SyePH5dfjx08BkhRA4gDpjrh1qpaYh3QtlXa3UENAkgJI5iGtj7tT3RTzkIpQch4gKQJI5iH1ok/UTTEP6UqihYCkCCCZh1TDd19IX2Ie0sNERZp37NiR6MqOmgASrzgK0gJ6StMW85Cm657SgDMbOMdRkNJpvKYt5iEJrwBSQADJLKSjFUof1jTGPiThlzdYw4YNico21ASQeMVJkCZTW22jAyB5WrCzwR1AMgspjWZrGwEJkHjEQZD+LlYlU9vqGEi9en0FSFIAySSk4fSaTqtjIFkdQLIgtoT0IP2q0wpIgMQjzoG0OaGOXjMgARKPOAdSP/pArznmIRUSs9bzqgkg8YpzINVO2KLXHPOQcM8GVQDJFKTf6CHddkACJB5xDKSu9IVue8xDmiBGurTvR70AEq84BVJmlaL7dSfEPKT/KoBkQewHaQY1158ASIDEI06B1Jqm6E8AJEDiEYdAOlRKdV9IXwAJkHjEIZDGUkeDKQ6CdG71yI9UASRecQikx2mxwRTnQJpcEru/pQBS5JB2Jl1nNMkxkPrjOJI7gKSCdPTPuYs2hfaTA6m30SSnQDqQLMEpdn3NgAASr0QtpHnNy0hrRpV09R0f9VJPc19IX5wC6VVxYT28KY+jHEBSJkoh7UrxfTtJ6mqwP86f9XF3GU5zCqQHiO61kBEgRSWkddcqv+g30D9nwZ83aYjhNKdAupzoNysdAZIFRa2GtOe6wE3mp3QuIVemetJOw2lOgVSO4rIASQog+SA9o975NDDoz83XuS+kL06B9DgVtdQRIFlQ1GJIyzR7ccvvC/Zz7Wmi8USnQBpKdBiQpACSF1J77fGQEUF+TPe+kL44BdKZMtQPkKQAkhdSZS0kFuTHvqXng0x1CiRhAiXMAyQXIPkg7dc5Qn99kB9rQnOCTHUMJKEPlfz8HCABkhfSGh1IpXV/IPOP2d8v21TkqmA79ZwDSfimOFXsMXrBlu3+ABKvRCGkjTqQKul0X9/+cmlSQvC9446BVK1aNe1Zq4DEK1EI6WiCFlLV3erOGW8k+abestq4qGMg6fz3YwhpT2P2l2/k6NhOTVv3+umSEKQFkKIQkqu+zhqRUKfnIuUHz+EnlVPLGG8lAZI22Z2ZH9L8NCanyzHjFkCKSkj9tCtEyzpxRBWaj93r7dsmcHqZNUZFHQNprV70IY1kfkjLUhjrNW5EW8bSzxq1AFJ0QtpbTu3oLpdr2/AUcROgUP0+v0ldxql73G60neQYSKFnDfNDOtOaPb1cfM3ux9gYgxZAckUnJNdnKiWF3Ze/Hpvzah1xrGrbSfuvVUPSeeilO4Ckzr/Psqd9kL5jbJw8kNWWpZ7UbwEkV5RCcr0YiGSYv8/6wSnFiZI0jgzuswpI2rzHnv/MB6kLS/UU/ZaxufotgOSKVkgZryqIJKtuoXrwx1dLayElGVxrAUiqzGUpf47zQspkrKenfQ9j/XRbAElKdEJyub4v5hXSYLm250NaSKTTTYpjIP2mybqt+85ruh1qKm75+CCJm0ve5/zlprF2ui2AJCVaIf0WV/Ole2rUatBzqV7Pm3Ug/aBf1DGQdBYJUXy1J8eeUPbK6cY6ZfshTWNsjndSR5ZyQa8FkKREK6QXaXSQnnforDNz9bs6G5KUksNz/b0msNRdgh/SRMZWeCe9zZhLrwWQpEQppH0lKga5NMKVqrO+rNfv6hhINWvWMKLU3tdpUwr7XlBAGsXYOu+0fowd1GsRv+61EbPhEv/kWVDzUp5gRdXcXCuqCtYsgRzf4BDqE6znMO3KUtWga45gyRKwaBXICd7hYhBIwr57iYqkvPfV3LULJ/R/Kp7ontWrfvosrai4cEZ7upx9kfXIVUL6jLGN3p//kLFdei2C8GsDMWvz+EewoGaeYElVa2LNvPqL5l6feDBYzwOFNJBey78qzxTMKpATBNKGUpTwlv+4z/6eCXSb9Al29EmiKzw7HQaxZkcEJaSRjK1XfP7s12vxBF/tLIjVX+2+p8bBu7ZROypidCtJp3y1O34lxU0PaJkZRw9L551erEPkvuRvKWPzhQBI4xlb5e0vbhFl6LUAkpTohPQo/RS86+YKKkj/M+rpFEjvE7VTNbUnmiS9jiXqLr1mPeM9LqTca+c75tqZsSy9FkCSEpWQ1ifUzq/vT4UDHKUaXpLkFEg3Ey1VNf1C9KD0+ru4fKTX4ywwnwrCasYmeHrnNWPPCXotgCQlKiF1pk/z7TxX+Zn0ovHNWJ0CqRTRflXTQaJK0utJorsNIGUw1tvT+wBjfQW9FkCSEo2QDpYpfSD/3lvaevc4XD8pSDenQCpJ9KeqaaO46Si9niBqKL1mTfSmO2MfT5wo3Zm1E0vzXCsxnbFZgm4LILmiE9IQeiWk/psGNbnzlke7zDoWrJNTIN1ENEDVNIjoOun1T6KOgVN820jCJMamygM5HVmqS78FkFzRCenm+FCeQBFinALpTaKyfwe0HCpP9Io08DrRYCNIp5uzZtKxorzRjA03aAEkV1RCmk2NOBZ1CqTdhYmuWqpo+PVqokKbxIHpRSh+ixEkYSFjTUcvnt2TsfanjFoAKSohPW30gPKI4hRIwifS5mLtYRuO5wnCPxu/kC6BpN6CMP/eOKI2qr4KSMK0VPeuh04HjVsAKQohbU68JoNjUcdAkr7cyUm63Hto4HnR1Dvia4U9QSAJe4enp7XqPtN/nrdOCyBFH6QeNIBnUedAEr6pFHBsregw6cFjIqSb1PvFIwogWRALIR2pVGwPz6IOgiSc+rSuj1G5Nw7IbUNbf31Wry8ghZeogzSK2nEt6iRIYo7OGvhmx259x2zIMegASJEl6iDVM7pmPMI4DJJ1ASQLYh2kJfQA36KABEg8Em2QWgZ7+l4kASRA4pEog7QjuXLQE37CDyABEo9EGaR36B3ORWMeUiExaz2vmgASr0QXpGNXJW3jXDTmIUn7utcK4T3WBZDCTnRBmkgteRcFJEDikeiC9AAt5F005iFNEHNcfP1RL4DEK1EFaXVcPe5FYx7SfxVAsiAWQWpPo7gXBSRA4pFognS8RPlgt1eNLIAESDwSTZCGUg/+RQEJkHgkiiBlXl9oA/+qzoOU/VXnB+9t9eU/gMQzUQRpSn63V40ojoC08c0H6u71DM+8xr3nu3Bf7cORACniRBGkRkZPZjEVB0A6ky7B2e4emR7vO4h02wmBYwDJglgBaX3CjSfz7xV2Yh/SzqvJD2mD+0mHZeR/b78ESLwSPZBeoS8AKQJIF+SrYgvVcd+ppIk4XGTESeHEZ0XEoaGAxCtRA+lg2VInACkCSK+JYC6f49keOpBAFLdYHlxzpdjO8TpZQLIgFkD6lDpnA1L4kM4VJ6rne/hKP/I/n28meR/nAkjmEzWQ6sStBqQIIH0vcvE90Ei4Sxzb5hnOFbedegISp0QLpJ/oURcgRQCpJdHjvtaT4je7W3xj4pe+hwGJU6IFUmP6HpAigSR+Bn3qa50qfiB94BsbTFQVkDglSiBtTro6A5AigVSNaKavtYMI6Q/f2LdEpQGJU6IEUk/q7wKkSCAlE/nvPixuFZXN9Y3NJeJ4rTkgWRDekI5cLt1eFZAigFSUyPcU8l3iB1Iz/6o/mag8IHFKdEAaQy+4ACkiSFcS+a6DHS5CGuNf9T8lugGQOCU6INWXb68KSBFAuofofe/wAyKkff5V/xmiBoDEKVEBaSndK70AUgSQuhNdle0ePBRPVN2/5p8rT9QXkDglKiC1pvHSCyBFAGmR+Cn0mXuwPXmedenOQHF0NSBxSjRA2lXkyqPSKyBFACnvOqL4gdLQ+DhSPtl8YWGia3MFbgEkC8IXUl/qLb8CUiQnrc6STv6u8kwX6ewgqu9tPf2W9Ni+mQK/AJIF4Qop46qkrfIAIEV0PdKr/vtBJruPxu7q0biCNJrC0REgWVCUL6T/oxbuAUCK7ArZtxM8jhInuxtmuEef4HqtOSBZEK6QGtAC9wAgRXjPhg1ty4puSrbcLSgglXjrIk9HgGRBUa6QVsff5hkCpIjvIpR7bFOGf2wGFb3jPa43bAAk+0NqTyM8Q4DE6b522VmcEQGS/SHtL1nukGcQkHCDSA5xKKSP6HXvICABEoc4FFIt/+1VAQmQOMSZkKYS8w0DEiBxiDMhPUEzfcOABEgc4khIfybUzPSNABIgcYgjIXWlT/wjgARIHOI4SOJH0eHypf72NwASIHGIoyD9+tpNZahErQbUUdEISIDEIQ6CtLWZ73zle9f6mwEJkDjEOZCWVfaf+E9lpvnaAQmQOMQxkNaVJWWKLPROACRA4hCnQDp2IwXmin2eKYAUCaQlM2bLrzNmrAAkKU6BNJTU6e6ZAkiRQKpCZdwt1BCQpDgFUg0NpJKH3VMAKRJIhSnhIiAp4hBIKzWOiKa4JwFSJJCqE30CSIo4BNJIHUhvuycBUiSQ0qTNzNq1axMVq60JIPGK7SC9qwOpnXsSIEUCaUOCzhL1BpB4xXaQ/qfz537BPQmQItr9/V05QFLGIZBG6Py533JPAqTIjiOdmDv4gw8+ILr2A00AiVdsB2m5DqRv3ZMAycwBWexs8MQhkFzVNI6Kee5+AkiAxCFOgTRAA+llzxRAMgNp/vw1gCTFKZAOV1I5KrfLMwWQcK4dhzgFUv/4QEdJs7xTAIkDpP2Lp02YsnB7HiDxjd0gHW5BFT8tr9xA+to3DZBMQsqZ3cS7K7zM07NzAIljbAZp5z10w5+u9Y/5HN253D8RkMxBWlEr4KP+hj8AiV/sBemP6vTwXmlg7otXJVDc5S2nKKcCkilIY+NUG58J3wISt9gK0o+lKD3DO3JsxxHVZEAyA2mOvO358Cffr9y3durgJyRVhRYDEq/YCdLQxEIDg00HJBOQzkvH51pv9I1vbi2OX8/xEUmAZEEigXTsVSo9LWgPQDIB6f9ENwMCWvqLLd8BEqfYBtL+RlRtZfAugGQCUjOiB1VNDxK15Acp6yz3ZOXyr3n27CUrZvXshWwLip4Vwl4CO+vSnfvz6ZNzLsLZCZYLgiVLwKJVIJ8l4H+AmAZSdaLvVU1TxO92HCGd4Z6zufxrnjlzSThrQdUL2RYUPSPkhPkDyyrSs//k1ynHgr/VmfOCJUvAolUgnyVw1hhSCaLtqqadRKX4QcJXOwsS7le7cUXieubfC1/tTHy1SybarWraT1QUkDjFFpD6xxedGEI3QDIBqRrRXFXTfKJrAIlTbADp8DNUaWH+3QDJFKRGRC1UTa2IngAkTikISJnftr6mSELFe9/fIo3tuJtu/CukqoBkAtIXRDQ9oGWm2DISkDilACDNudl7jkrRHoddq6+lpw6EVhWQTEA6UY4oceAZ3/jZTxKJyp9QdwOkCPPfQ/o4SXG6153jlGcF5RNAMnOK0FfS8i79xpLdF4TsPUt7lZFGv+TnCJAsKBoM0ueBJ07GJQ4LuSogmTpptbd3kZf2nr3ai6MjQLKgaBBIywqrTkF+JPSqgGTuMooJZQKWfKmxPB0BkgVFg0BqoHJEib+EXBWQTF7Y5/rIf2P1a/tncHUESBYUNYakc7OtFiFXBSTzl5ofmTH03e59h047xFcRIP3HkHppIZU9GmpVQMLNTzgkJiA10kKi1aFWBSRA4pCYgHSLDqSZoVYFJEDikJiAVEcH0vRQqwISIHFITEB6VAfSqlCrAhIgcUhMQOqudVRSfY8TwwASIHFITEBaqIXUJOSqgARIHBITkFx3aCDNDbkqIAESh8QGpLlqR41DrwpIgMQhsQFJfUj2ur2hVwUkQOKQ2IB0oF6Aoxv/DKMqIAESh8QEpMMNKG1SNS+jpHZ/h1MVkACJQ2IB0rEUevCw6/CwRqWIEmt1C/nkIHcAyQSk9Nq16/J/KBIgefOfQsp8lm73fAYd2JoZdlVAMgGpiPgN4BgguWICUjeqtdNEVUAyAekGEdJqQHLFAqQBVHWzmaqAZAKS9IzrPoDkigFIX8SXC/m0Ot0AkpnHutxKVHgHIEU/pEmFSiwxVxWQzOy1O3E70Q08nywGSMr8Z5DmFk2epdMznACSqd3f2T3jiR778vddBw/5Aki88l9BWl46YbzZqoBkBlLt2rWLa050BCRe+Y8grbks7jPTVQHJDCSdK8EAiV/+G0ibrqJ+5qsCEiBxSBRD2lObXuNQFZDMQFqmF0Dilf8C0sE76DkeVQEJ59pxSNRCOtKQnjjGoyogARKH/BeQDk1+9ZmU9h+Fc22DbpSQMlvQfYfNFpQDSIDEIdZD+rtXSfcmaNxDS81VVULqSLfsM1fNG0AyD+nEz+MGfTBBEPifCA5I3qy8RrE7x9xONgWkN+iabaZq+QNIJiGdfKOm/ESXhoKwqN6nRwCJY/yQlpcK2DH6spmqfkgf0xXrzVRSBpDMQRpR1vO3lSARJX4FSPzig7TzKtUhhuEmqvogjU8ou8LkHPoDSKYgDfD9ad2QiLrnAlIk+fOj5xux9DF7lG0+SB3Vx+rK7o7w1/w18PmUtPTR0o9PTSoS+t228g0gmYH0tfRHrfbB1M0ypKPN48XR8YAUfn59zPPEw6T0Hf5WL6StSWpI9KZRpaA3/vmtkffXtN/+c7GkKabmOTCAZALSxcpEhYdkS5MkSIKwrhJRlfOAFG4GK6RU8H9MeCEN0Z4+UkuvzIKONQpTUrXnjR4hMUTxa8qUSBhrZpbVASQTkMaT99nLHkjCzmSiYYAUZt4JMFLY9wQIL6TGWki0RVNlw1O+iQ+s1Ps1/wus0NHEHGsDSCYgpRI1EwIgCV2J2gBSeJmk3gDy7krzQqqvA2mhusr8yxRTS36v/TXfqiqUWRf5LGsDSCYg3UQ0XwVpEVF9QAorh6qokaR4pnggZVTXgfSjqsqqwB3khTX7EQ5rfs1TEc+yTgDJBKRiRHtVkPYRlQKksDJIq2S5e8qZndMHv/zE9dpdDWKeGfmb8hy5I9epplfco/o1g7U1Qn9mef4BJBOQyhCtUEHaAkjh5j7tGt5z16KxPVPquC+aTKpaWU+SOOH65v3nHHQXGaCZ2k31ax7QFugR+RvWBJDM3Y5rqAqS+EX8OkAKJ0cTtWt4ghtKtQZt+/+4NsP1tbbH3XP6N68jfVQVEjX9uNNVTdOjRODJqMd0Ptfqm37f/gCSCUgvEdW5EAipIVELQAonG3U+axIf6zTox/UZ3p0NBy/T9BgmtR+YO7CNrClefeaDlMCjRKt1elQx8Y7VASQTkBaLf4wOAZDGii2TASmcrNBZw8u4J/nObNBsRtXybx8dWfp5+/qFdYqwj/t06/BsSsO761xdsYTOdPFLuJm3rAogmTmzQfz8oZRdPki5Q8XvJNWzASmc7NRZw693T/JBOvZw4PQi6pvQvaNTxJek0lVr3XaPzoQakc6zTgDJDKRd5aW/U9OPlxDdf2D1EOlU/4T5/Bw5AlJmGe0a3sg9yX/2955blZOTv1YX6aPD5N6RE39c8Nv6XUfdXcpqezwa6TzrBJBMnbS6Qf3tPY7r6d9OgORqrl3DB7unKK5HOtTCP7XKz5oaw3Ug9Q7s0kLbY1DE86wNIJm7jMLVMuBPczXPW584BNJUzQpe3HPiasCl5vMede93q/6ezrXhv+pA+iGwy3Ttr9ke8TxrA0hmr5Dd9loNzx+mSINJHLePHAPJ9aB6DX/LM0F185N9Pw4f9H8Gt7q/WsOkpNpbA3UPwzPIIwkgcbhnwz+/TBs3cdbai3wVOQbS+vKBK3h9L4Ew7iI0UANJc7O6PysEdqjH564nngASbn7CIebO/p4bcN/nWlu97WFAOnK9ylEl7YVJ8wJ2a9TUnj9uJoAESBxiDtLGUnH+FTzVf1ufcO5r93vpAEfJ83X6rLnZ3yGF092DvAEks5Byl36QWq9GlZvu7zD2b0CKJBn3U/9ht8iWkhrOVkwI6waRCyopHJXWv/g1Y/it7l/zsNmnuGgCSOYgXRpc1f/ni398OSCFnz7UINPl2jxl2MhZgR8T4d1pddPTvj/EQ8aPJN88ZcLEWZw/jaQAkilIe+oEfjOPa3cBkMLM4qRyBpsr4d6yeHGXmsmUdE27OUF7GTyM2WQAyQyk/e6LxYrUbNQupXZRefhRjrdscASkv6vHTTKYFMm9v/fn2wOQ7AdJPgXs/tmeW6wulI9UvAZIYaU1tTea9F8+1dxsAMkEpHmim3JLFA1LyoobSpsBKYxMoOsPGk0DJIdAep6o0B8BLb8XInoTkELPxrJJywwnApJDINUgSlc1pRPVBaSQk3E/DTCeCkgOgVTCdxchXxaIX/YAKeS493wbBZAcAqkY0TZV006iIoAUaoz3fMsBJIdAuo5olarpD6IagBRiguz5lgNIDoHUlqi/qmkAUZrGw1+fdXymZY+hW/wtR8d2atq610+XgrXEPqQge77lAJJDIC0huiwjoCWjItEsVa+zHzJPhpz2NM1Pczd0OSYYtsQ+pGB7vuUAkkMgCW2Ibj+qGD96O9FDqgck5bzDWErvCWN6iUz6ug/dLkthrNe4EW0ZSz8rGLTEPqSge77lAJJTIJ1pSFS86x7P2J6uxYmqH1X1mc9Y4zXSwF8ikznyT7VmT0tnt2b3Y2yMoN8S+5CC7/mWA0ixDukjbz6QrrqMr/1k53c7P1lbeszYjau2qyC9zJjnvsbrU1gr6ePqO8bGyQ1ZbVnqSUG3JfYhBd/zLQeQYh0SBUugo0OMveIdfp2xneJLF5bqKfotY3MF3ZaYh5TPnm85gARIvqxmbJR3eDhjvwtCJmM9PQ17GOsn6LXEPKT89nzLAaRYh1QzWAIhLUlLm+4dHsLYNkFYw5j35ne5aaydoNcS85Dy2/MtB5BiHVJkyXmBpZwWhGmefQ5SOrKUC3otsQ4p3z3fcgAJkPTyA2MDxJeJvr0PgvA2Yy69lhiHlP+ebzmABEg6mZPCmuwTX0cxts7b1o+xg3otgnDplJgTx7nnn0v8ax4/ni2EM6uZ99OAUPqdPRfh7ASNYM0S+NeCoqeFLAuq/pNjQdHjF4STwTv490abgnRUFJLymzT0GWMbva0fMrZLr0XctrpNzGozv9G++YgeyyvoeUD+6/hPVNCDdPyX0R+polfk/KQmjLVaKQ+PZGy9t13UtV+vRRD+6iRm00X+ybOg5sVc4VLonVcnVTgQUsecnAhnJ2gEa5ZAGAsg5OQIliyBglkF/Lfz1kI6/47Ocxt1HK18gTHW3/MlcTxjvnPGxS2iDL0WT2JyGymkPd9ysI3klG2kJ/I/jiQma6DIqJvvy9s0xTHXzoxl6bXEMqSQ9nzLASSHQJohwyl0ReWAqHudfVn8VrfYv1GwmrEJnsG8Zuw53ZYYhhTanm85gOQQSKkio6YbLqmbA5Pdi7G+JxQNGYz19gweECfptsQupBD3fMsBJIdAuiGUR5iLG0Cf5AS0dGJpnmslpjM2S78lViGFcM63P4DkEEjJROpzvTW50JKlqx4/NomxqfJATkeW6tJviTFIqz/s0Pyl95aHdM63P4DkEEjlqHCuuk2dhYxNUzWdbs6aSceK8kYzNtygJaYg/XiLZz9MzX4hnPPtDyA5BJK4fhxTt6nzBWMvv+3PUQ+upqMXz+7JWPtTPm6qltiBdLiNcp/m2DCqApJDIPUlmpwfpL4sIHvlxmmp7rFOB73dtC2xAulI4LNbbz0QelVAcgikw6Xpivw+kjroQRL2Dk9Pa9V9puIZMNqWGIHUXnWY7enQqwKSQyAJk+Oo7pp8JJlJDEBaoDlgPTnkqoDkFEjCxASKZ/2+n6EMICnzkAbSzSFXBSSnQMobkhTSuXbOhbSzkHYB/R5qVUByCqSXQzrXzsmQxussoI9DrQpIDoE0T88RICnTX2cBdQ21KiA5BJJ0rt2tP2w/fTYggKTIGzqQngu1KiA5BFIVooctvdQz+iEN1IHUI9SqgOQQSMlEf1npKAYgfa8DaVioVQHJIZBupsR8rqFwPKSDRTWO4kM+2w6QHALpZYr/F5CkGB9HekYDqUHIVQHJIZA2Ef0ESFKMIa3THGlbHHJVQHIIJKEN3XgGkFxBz7X7UOXo9dCrApJTIJ29mR49AEjBL6PoEOCoaUboVQHJKZCEk49SkZcnrt6uDCCp8lGyj1HSW6FfHwtIzoFUrVpl7d5dQFLnr+buJVM0bXVYVQHJKZB0DpIAkk7GUIth//vsx1Bvw+UNIAESICnTjBZEUhWQnAJprV4ASZ1jZcuHsYvBH0ByCiSrEyOQ5lKLiKoCEiABkiLdaFxEVQEJkABJkRsTd0dUFZAACZD82RR3X2RVAckpkBbpBZBU+YTei6wqIDkFEnZ/exIU0mO0KrKqgARIgOTL4WJVI6wKSM6FdFvDhoAUmO/ppQirApJTIClueXJm78KRdxFdt5Ojo9iA1J5+iLAqIDkFkjrzrqJy+wApMNWKHo6wKiA5FZLwZ1Gqz/E2DrEA6Td6ItKqgORYSMLnRBxv/h0LkPrS0EirApJzIR0nagJIytwdtzHSqoDkXEhCSboekBTZkxj60yfUASTnQjoXT0UASZEvqXvEVQHJuZB+I6oCSIo0p/kRVwUkx0LKvYfoYUDyJ6N8uYiu6ZMDSA6FlP3HfUTUE5D8s523JAAAHatJREFUmRfhNX1yAMkpkCoFpGKidJJQsaOA5M/rNDbyqoDkFEh656wO5ecoBiDVLhTZNX1yAMm5kMrP4ugo+iFtirvHRFVAcgqkUap8szWXp6PohzSE3jVRFZCcAsnqRD2kx2mliaqABEiAJOVw8avMVAUkQAIkKVOovZmqgARIgCSlPU0xUxWQAAmQpFxd9JCZqoAU65BqBwsgebOSHjdVFZBiHZLu3YNwFyF13qUhpqoCEiABkph74jaYqgpIsQ4pVT81AEmZvYm1zVUFpFiHpJvzbxWSHDFA8mRsOE8w1wsgORHSMvnzqOL3HB1FOaQWNM9cVUByHqR/X4qTHD1/Ikgfh0HKqFD2mLmqgOQ4SNMulxhdw/NJFFEPaT41N1kVkBwG6WiaxCihRxZnR9ENqTt9abIqIDkL0lelJUd1eT6FORYg3Wzmmj45gOQkSLsaSIySB3C8U3FMQNocd7fZqoDkHEiXBhaRHD3I9SEUMQFpKPU1WxWQHANp/S0So9Jj8ixxFNWQnqDfzFYFJIdAOvemfAi2yRFrGEU1pMPFq5iuCkjOgLSkusTo8qlWMYpqSD+Yu6ZPDiA5AdLJ9hKjuPR/rXMUzZBeou9NVwUkB0CaWklyVGOZhYyiGtLVyQdNVwWkmId0uLHEqFCv85Y6imJIq+gx81UBKdYhjSklP8H8L2sZRTOk9+gT81UBKdYhuW+q2q2HbgBJzH30p/mqgOQISLhC1h81pL1JN3KoCkiA5HBI46gbh6qAFOuQkoMFkKRr+uZyqApIsQ7pv0q0Qsq4rIzJa/rkABIgORvSAmrGoyogAZKzIfWkMTyqAhIg8cmFi/yTZ0HNi7nCJeXorQlHeVTNyeFRRR3BkiWQdyn/PmEnR7BkCfwXq4A22QUI6fQJ7jmZw7/miRPZwr+Kse1xd3GpmnWOSxlVhEtWVL14yoKiZwQrlsDJXAuKqlYBnfhPSsVXO8MEfrX7jPpwqYqvdvhq52hIT9GvXKoCEiA5GdLh4pX5VAUkQHIypB/pRT5VAQmQnAypA03mUxWQAMnJkK5JPsCnKiABkoMhraFHOVUFJEByMKT3aRCnqoAESA6GdD+t51QVkADJuZD2J93AqyogAZJzIY3nck2fHEACJOdCakU/8aoKSIDkWEiZlcoc5VUVkADJsZAWUhq3qoAESI6F9AaN4lYVkADJsZBuSdjBrSogAZJTIW2Lv4NfVUACJKdC+pze4VcVkADJqZCeouX8qgISIDkU0pGSV2byqwpIgORQSNPoBY5VAQmQHAqpE33LsSogAZJDIV3L65o+OYAESM6EtJYa8qwKSIDkTEj9aSDPqoAESM6E9CC3a/rkABIgORLS/qRaXKsCEiA5EtIEepVrVUACJEdCak2zuVYFJEByIqTMy0tzu6ZPDiABkhMhLaImfKsCEiA5EdKbNJJvVUACJCdCupXjNX1yAAmQHAhpe3x9zlUBCZAcCGkY9eZcFZAAyYGQUugXzlUBCZCcB+lIqSs4XtMnB5AAyXmQptNzvKsCEiA5D1JnmsS7KiABkpMgbR3a4uG7nyqb9DfvwoAESM6BtOX5RHLnph84lwYkQHIMpLkVyJ8ux7jWBiRAcgqkOYVJmZZciwMSIDkE0sbyFJgPeVYHJECKEkgHJjzboM4DLb/aG2G1lipHVHwbh3n0BpAAKSogHf6grGf9L9n7YCTF1hdSQ+L33EsXIAFSdEDafqcCwM0bIijWX+OIruY3q4AESNEAafd1AQIqbw2/2FNaSLSZ37wCEiBFAaSGKgH1M8IudpsOpAX85hWQAMn+kP5PQ2BY2MVq6kCazm9eAQmQ7A/pFg2BamEXu18H0kp+8wpIgGR7SH/qGFgabrEXtDUKc7yLPiABku0hjdSBFPbR1O+0NR7hOK+ABEi2h9RXB9Ir4RY7rD6xgWgCx3kFJECyPaSeOpDahVlr45OaEvV4XiULSIBke0iDdCC9HValjAEl6FbVPvSya3jOKyABku0hTdeBND6cQr/cTiX6Hzv4kLJAGdz7m2sAyf6QjpTWOEoI44TTgz2T6NG/xIFjXZN8Bequ5juvgARItofkaqT9RCrfc0+IVSZXoUrez68/2pSTfrjQvaNxFyHOASTbQ1r9kIZRUocSVLbnrhBqbGlO8W33+ceP/fbTjAW7+c8rIAGSvSEdEL+Y3f2RCtIw146epal4uvqs06Mb5i1TntCaObws1f45sI/3qeZ8A0iAZGtIk6pQxeGZqh1370hT9/W/jIqmK6+omNu8jDS1aqe/PA2/30tF+qjvzQBIgOQ4SGsaUqF0+ZLYHyr7GF3mPZD6d//LKam5d6/BTubrUbiHpEfeyaB94jIgAZLDIB3qU5juXO4deb9unITkht77/T0OD7+a4lNWSIN/XKP8zHr0oGtmDfGzTKcqIAFSbEM6uDNwfFJVumy4cu/a9p+/m7tF9UNHhteg+EcXunZdG7gV9WTbuLi2urd3ACRAil1Ihz9/rBRR0g2vr/G2bEih+DahPA0sY9LNRA88oNmzd8M8/f6ABEgxC2liFd+O7XT52oYj/YtR3YUhrkaZk3Qufy1mtI8bkAApViF1Vwqos9HlmlaDyvTPCOO+djqHbMcYdAUkQIpRSO+ovpP93pzim+8I6waRFbWQjB5bDkiAFJuQZqkJJNCti6QJoUPapXVENxr0BSRAsgmko2M7NW3d66dLYUA68OvU2av1b/2TebvGwOvunqFDWqEDqbxBX0ACJHtAmp/G5HQ5FiqkKY8nS6t2hWdX6UycqzXQwj0ldEjrdCBVMegLSIBkC0jLUhjrNW5EW8bSz4YEabN/13RSxyOaya9qDZR1n9YTOqRDcdoi9QyXIiBZUBSQwsyZ1uzp5eJrdj/GxoQCafkVytX73n3q6Q10Pk3WyVPC2NmgvWEXvWq4FAHJgqKAFGa+Y2ycPJDVlqWezB/S1sqB6/fj6i2lWjqQfpKnhAGpt7bGXMOlCEgWFAWkMNOFpXp+zbeMzc0f0uPqFXxgwOTMdVfpQHIjCAPSLs1FtPcZL0VAsqAoIIWXTMZ6egb3MNYvX0g/aZCU9550umtO/7b1i+swInKfsx3Og8Y+VlVIXma8FAHJgqKAFF7WMPaVZzA3jbXLF1ILrZJRIqHB6Q+4L/qu+mh9bY8K4e7+FtMqsMQXQZYiIFlQFJDCyzTG5niHO7KUC/lAyiynZXKZ+5nIFR9IH77ogMu1SNujrfuHw4J0rIPy82hUsKUISBYUBaTwMpGxFd7htxlz5QNpu94Xt6sefXXEYv9j9+5RT09c4Z4Q5jNkJ9XwFnhoeZBugARINoA0irF13uF+jB2UXv9ZLebIv3pZpeOogqrP0sKqDumeCadydGsa5vj0l+6uUbNBr1+D9roonA6vbEg5d8GCov8KYS6B0HLpjAVFs4TzFlQ9lWtB0fxXgdPWQ/qMsY3e4Q8Z2yW9LrlNzGrd7nt1IF2j7qS6Sf5dF/QqIch/llzfkGWQRjK23jssfiLtl17/niBm71m9uOK1kO7W9Aq4tckDB7zNWbm6NU3mkpBlQdXsixYUPStYsgRyzllQ9IKQbUFVq1aBfJZAlvWQxjO2yjssbiNl+CYY7LXTueSuu7bXrNreicXfOuprDXMbKcRgGwnbSDbYRpqmOArbmTE/XQNI/9NCWqLTLWNSy6uT4yve1195K2JAAqSYhbSasQmewbxm7Dn/BANIe8uqHT1qOPvq2wgDEiDFLKQMxnp7Bg8w1jdfSK7PVI6Khf7gVkACpJiFJHRiaZ6rJ6YzNit/SK72gZAmhv5+AQmQYhfSJMamygM5HVmq/3isMaRM5a1NSnwTxvsFJECKXUinm7Nm0tGjvNGMDVe0B7lCdnodr6PUdeG8X0ACpNiFJCxkrOnoxbN7Mtb+VGiQXJnzuj5ap95T760N7/0CEiDFMCRhWqr7ng2dDipbrXzQGNcAEiDZA5Kwd3h6WqvuMwPP5AEkCwJIMQ1JN4BkQQAJkDgEkAAJkDgEkAAJkDgEkAAJkDgEkAAJkDgEkAAJkDgEkAAJkDgEkAAJkDgEkAAJkDgEkAAJkDgEkAAJkDgEkAAJkDgEkAAJkDgEkAAJkDgEkAAJkDgEkAAJkDgEkAAJkDgEkAAJkDgEkAAJkDgEkAAJkDgEkADJeZCiJlP7n86/k03Sf3xBz0HI2d5/ZUHPQsiZ2f+fULsCkmF63paRfyeb5LbnC3oOQs7i2yYW9CyEnD63HQi1KyAZBpAsCSA5LYBkSQDJaQEkSwJICIIYBZAQhEMACUE4BJCiOicy8gp6FkJONM1r+AEkvXzF/NlR0DMTJDkt2EnF6NGxnZq27vXTpQKbn2AJmFcbL+C/Puv4TMseQ7f4W0JarICkl/ft+3cOyB9MCWl+mnuOuxwruDkyTsC82nYBn/3QO1tDvOe1hLZYAUkvnVmz17zZX9AzY5zznZUr57IUxnqNG9GWsfSzBThTBgmcV7su4Jx3GEvpPWFMLxFOX/c30RAXKyDpJK8JG1DQ8xBCNkp/bt/KeaY1e3q5+Jrdj7ExBThX+gmcV9su4PmMNV4jDfwlwpkjDYS6WAFJJ5mM2f6g4ez3npO/cPhWzu8YGycPZLVlqScNf64goplX2y7glxlb4R5an8Ja5QqhL1ZA0skGxhYW9Dzkl7c9X+V9f9suLNVzUv+3jM0tqNnSjWZe7bqADzH2inf4dcZ2CqEvVkDSifgBvyX/XgWb+RPFtPevnOJ/8j09g3sY61dQs6Ub9bzadgGvZmyUd3g4Y7+HsVgBSSfjxb/5+U3zZyw/kX/fAk1f/8q5hrGvPIO5aaxdQc2RcRTzatsFvCQtbbp3eAhj28JYrICkkw9Z2qw20neRlMH2vrZPsXJO82wcS+nIUi4U0BwZRwkpChZwzgss5XQYixWQdPKy/yDHs5sKemaCRbFyTvRtJsvbJK4CmiPjKCFFwQL+gck7FkNerICkTV5TxlrNcWUfWPg8Y8+fK+jZCRLFyjmKsXXe5n6MHSygOTKOYl6jYAHPSWFN9glhLFZA0uY4Y+nu/3rO9lBsfdowipXzM8Y2eps/ZGxXAc2RcRTzavsFfFQ0k/KbNBTyYgUkbXJPn872DO5JYY3PF+jMBI1i5RzJ2Hpvs7ga2Ol0AXcU82rzBXx+UhPxE9N9i5aQFysgBU8vxjYU9DwYR7FyjmdslbdZ/DJvv4t7+wacF+iL/RbwyhfETbf+nv2JIS9WQAqe4YwtLeh5ME7gXjvf4cLOjGUV0BwZxwCS3RZw1kCRUTff17mQFysgBc8km/2dA6NYOVczNsEzmNeMPVdAMxQkBpBstoDPvix+q1vsv3Aq5MUKSNrs2rrNNzyIsb8KcFbyiWLlzGCst2fwAGN9C2iGgkQxr/ZdwNniN82+yqPEIS9WQNKmr/8ambwOLNWWu2fdUf4v34mleU7zn87YrIKaI+Mo5tW+C1jcJPokJ6Al1MUKSNrMYuwDz+BcccEW6LwEjxKS+B1pqjyQ05Gl2u94rHJebbuAL7Rk6dmBTaEuVkDSJqsNY19JyzNvfjPW0l5XJARGCel0c9ZMOsyRN5qx4QU4T0ZRzKttF/BCxqapmkJdrICkk1WpjD0/9OvBnRhjiwp6ZoIlYANeXAuajl48uydj7U8V4DwZRTmvdl3AXzD28tv+HJXaQlysgKSX1W09Z4I9s6SgZyVoAveETUt1z3Qn+50fJKjm1aYLuC8LyF65MbTFCki6OTfnnecbt+rxnV3PTfZEtUt57/D0tFbdZ9rvzG8pgfNqzwXcQQ9SaIsVkBCEQwAJQTgEkBCEQwAJQTgEkBCEQwAJQTgEkBCEQwAJQTgEkBCEQwAJQTgEkBCEQwAJQTgEkJyQ3JV9H6hROrHCjc99daag5yVGA0jGSSYi1bOlUolqmylZk+gjMz8fNBubXZd81VKdCbNuIl9K9rloxe+eQFTIirpRE0AyjgSp1JGAJjtD+j5BgjJD036hOQXklkwLfjkgFfQM2DgSJGoc0GRjSCfKiXObUHq2uv3MvdLbKPfS8Glj32xcQZZkwde7/CANevDBN/j/VhsFkIwjQ6KpyiYbQ5orzusgna9tz4vtRYd6JlwYXlQc7cr/t+cHqT1RI/6/1UYBJOO4IV2uvAbVxpAGE12h0/yd+BbKb/aPrygkfm5t5f7bj69duzbYdEBycERI5cXVsL2iycaQPtCdtfOXESWtUbZ8KL6lVy2aB+MAkoMjQup2m7jaKe7PEXWQxojz/1ZAy0VxO6lsrkUzYRhAcnBESD3+FL8JVfffCjQ0SJtfqVukauN50uDhfreWL3JjY+8tOt2Qlj5/ffGS17/4c8APLel4Y9nidZpMVG3mnB94Y9F0zW84OeqJa4qUuq75JP8dOXQh5dUiKn02sK2daOvPIIX0s+l/91cpUurGx15XPSJIb/50+qog6b3Z0x83v7VUhTtaff2fK+cRQDKOBEnoJa52b/qafJBGEVXytZ4V+8i3Z1tLVE3IejbOvZ/5lTzhu9Kefc5PuldUCdKZtt4d0Y/87Sux+xFvY+WvPU3LpN91QvpI1AAZV97b+xr3sxLm+3ZuB+7+3ii2qFfzxW3atFlqVOhTcVBxj8S+4qj06Md/msZ5O8Y3Pq0zfz/6dzZo+37im7vtRm9WyB1UzttY21a36AoxgGQcGdL5GuIq4nvWVAiQ/q1DFJcorxFDJ0qrVGF5uI/cU4T0gW8tIrpqj6fCxgr+Ru+XP2lFza5HOpD6KDrHj5NajCAN886ZfrSFjsYTtfJ3uI7oevHlxLWKjvRUrnb+/JB0+qog6b1ZobeircQeIeoCSMaRIQlLRQy3XPI05Q+p8mN0w5yTF9c/JLYlJdDl3x7L3SYdEU06LnUQIYlrUYX3V26b10780kh13IVdFcXhmwf/tmduM2lF+kJulFbUnvKKVTdwviZKbc3n7N46+QEJgPRouTUNG4rrb7GGDRuuCOjaVERtfH9QnUJCQ3E99n3NWyc2DxRfu4qvl32wbPfWBa+U0J8/PySdvlMaNryCqJw4dweM3uwscSmX6rVk94bJ9cW2O0L6+9gqgGQcNyQh3bM2SckfEtGj8vHOvFvlVeyY3CFNHFwuDdSUGm9zN64uKw4PkQdbiUMvuZ8AOUb8ACsr73EXV9Ti8ZTUe9UR/+N6pLjEn0v0fP0SN4yoVo53ULuNJH6cVjd8f7qFxooDvocu9CBKOCJ+yIgfsFd5zodYV1x8h4Jm/nyQ9Psqt5F036zIp6z7e1/uw+J0+z25M78AknE8kE5eTlTEs9kcAqRkz4bPD+Jw3B/u4Q3i8GhpQIJU3HvW0RJxpJqEZLs4kOItNkQckU8CWCahq7VdM1sfKWDLa+UceUgXUhmiBobvT7fQSXHdbutpy6tCxMTXBeK0Kd6OzxLVFDTz54Ok31cBSffNXojzvGkxv5LeiU52DyAZxwNJmEq+1TEESC97GqUVhnmGLxXybAxIkPznFdwvji0VX18Xv/n5dnBlV/OsffKKqrOBI263lPKd5CMRbSIP6UHKEdfPVMP3p1+oMVFpz6NNlott08XX78qXr/CPt+NAaYeKZv58kPT7KiDpvtmj3v9pxGR169ZtseFM2zWAZBwvJKGJ+Gf+Sh4KAdJET+Npcbift8cVfkjx/i3p2WIX6UFB4hewpv5fK4qIkx4aJ62oj2nn6gj5//MWI26Llff+mAbSP2LfNkZvz6DQFLHZ89jUTuLmnOakoz5KSL75+1H3FKE+OpB032yuuKyfzNMWiJ4AknF8kI6UEv+Xlp/xEQKk1YrG7709KvshKdb2vCLyB4a0un/sb5U+AKVDTNKK+qV2rmZQwCHiT7xbFHqQLvg+r3RiUOhcCaIX5ZZLFYi6q38qs4YSkm/+dCH5+/ohGbzZuuJr071GsxoFASTj+CDJ5wekSQMhQNquaPR91VdAelLxG8TReoLwS+AavVkclXZFSyvq79q5ko707PaPfi+Oyl+EdLeRSgTZRjIq1Fbc7pd3Js4TWxRn6QlZG2cMaCqd9OqH5Js/NSRVXz8kgzc7R/qeGFf/3V/s+SiN/ANIxvFDynvQs7XAAVJnxW9oRFRDEKaRJoME94qqOo1AyrviCqdY21aS5wiqLiRxE+Q6o7dnVEg6JrVAGmir2A+956Mnq/qOs/oh+eZPCUmnrx+SwZsV3vSMJDcc49vEiqYAknH8kISd4vAV/3KB9K7iN7QjqiII47TrlrTlJK2oOs+26i5+zVSM7hN7yQ+w14XUUMTyr7qx41NPpeQaF5K+0b0kvp4vSTTKPe3Uqwk+GHcpIfnmzw9Jt68fksGbFYQFd3vHE1+10eOZQw0gGUcBSRhA8tqlD+lMpJ9IKUQ3uY+L3t8wINIaLa2oh7Rz1Ve0oXhgsPQr5d3NupCkw0M/q9rOFXJ/TBkVEroQVciRbRRxIzwnr+PxN784YNoeaVvKD8k3fz5I+n39kAzerJRtHzVwX7lC1x/Xvm+bB5CMo4R0STrxZ5kBpL3hQHpK8RtqiyuVIMwSex7Q/noDSEPFZsUpNNLW+kJpQBfSb+Q/KuSNdJymSZBCwgqSN2PSfHv8eogNV33p+WTLB5J+Xz8kgzfryYVl/6stSXrOsIddA0jGUUIS1opfWGqc14e0MBxINyl+QzGiDoH7+hQxgCRtZCz1jw72/kpdSNI1E0VPBLa9Tu4TKowKCXlVpc/N08nevQLnShNd67u8MTgkg75+SAZvVpEFl4ldTgTtYsMAknECIMn/1b6lhFTRN2lkOJAS/Dt5F4ldxopre7L3VCE5RxctWiQdUjGAdJACrjB6QtzSkQ/A6F+P9L77K6kiki15P7dRIUE6471S7tdEV7vH/yL/0TFB6BcUkkFfPyT9N7tp/vylvrb/E2us0r4VeweQjBMIKesacV25TgEp0beBcXc4kKibr+RD4saEdEJRA6Jr/BfhNJf3iRtCEsT5KOO7xGhznPeEG31I/4gfenFzlS3/E8veE6yQ++KL5Y8Tvece/Ukc9V9k2zgoJIO+igOyum/2XeXilH79TJ03busAknECIclf4Mi7tn4jDnlOpJOPt4QBqcQxT+NS8qxdk8SBod6u6+M9I0aQpB0In3hHWpP3HFN9SPJekiKKE43mSfvUFgcrJMjbbi0LuZEL7k8Z3zVKC+Pz/UTS6auApPtmp5P3lEHBvWMv6A0g7BhAMo4Kknw/Hu/aulXa+eT55lM1PEhUz31y9B/SlWzLpCHpjLNC4909d1YS/3OWT2s1gnSsFFGS565bEpMbgpz9LW7wPCZ2SXjdczFe1vuSo6eDFhLTX36nnhO3hexEokc83/omSlcqVsjTzJ8XkkFfCdKDnmp6b1ba917Ts+dj7+VEl/lmJVoCSMZRQ/qnoh9SrnRp6b0rzgv/DC9KNW4NA9KN4gbIwDW7Fr4kXY/kPhdHvuaJHvlu2/4lr0k7gN0n3hhBEr4SJ8S1WbB/9w8NJCSeTXcDSIKrujTXRZp/9uOMYemlpOGap4IWEty7IYm+846Km0/E1l3M2zfjEfEHyH1s2mCvnX5f4SVxC+yAcCnX6M1Kl2wlvzlz++7Fr5dQflBGTQDJOGpI8lk03rV1pryuxUtrZrmdDcOAtKo6+fLIec/0L+P9jfS6u80QkvCGonOhbzyNRpCEfxpQQO46FLyQlLvEhjK+0x4OyhfMJxST/r35R+nfqlOMIOn3lc4Dp/gy8duN3uw/VyhnMTXqPpAAKUg0kATmhyQMKeb5q9fbJ4QDafsR74od39V/BH+e7wLtymM9m+LGkITRvvsbXOfb/DGEJFwaVcW/jpZ/X3k+t14hKdIl6l38o0u8+BO7ns2tJQ18ZQRJv6+wv4zctt3ozQrra/pm8YpPLwlRF0CKOEe7N6hcpEbjaWH/1XNnNLk2uWydnpuUjdnfPFO9RMnbW39+3ujHlDkxolG1IiWqPzM5tDviZ8/uenuVoslX1Hnpm6yICmUPTamZXKFeX2kzZseTZQvXWhhu341NKxcqWdtzKFbvzV6c2qZW6UJX3PvsyKg8bxWQEIRDAAlBOASQEIRDAAlBOASQEIRDAAlBOASQEIRDAAlBOASQEIRDAAlBOASQEIRDAAlBOASQEIRDAAlBOASQEIRDAAlBOOT/AeYgHS6LUwLvAAAAAElFTkSuQmCC",
      "text/plain": [
       "plot without title"
      ]
     },
     "metadata": {
      "image/png": {
       "height": 420,
       "width": 420
      }
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "colnames(deg_counts) = c(\"formula\", \"sig_genes\")\n",
    "#plot the number of significant genes vs the number of covariates\n",
    "ggplot(deg_counts, aes(x=formula, y=sig_genes)) +\n",
    "    geom_point(size=5) +\n",
    "    geom_line() +\n",
    "    xlab(\"Number of Covariates\") +\n",
    "    ylab(\"Number of Significant Genes\") +\n",
    "    theme(text = element_text(size = 24))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "~sex + W_1 + W_2 + W_3 + W_4 + W_5 + W_6 + W_7 + W_8 + W_9 + \n",
       "    W_10 + W_11 + W_12 + W_13 + W_14 + +W_15 + W_16 + W_17 + \n",
       "    W_18 + W_19 + recur"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#use 18 covariates\n",
    "design_formula = formula(formula_list[[19]])\n",
    "design_formula"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "converting counts to integer mode\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "keep\n",
       "FALSE  TRUE \n",
       "27975 30244 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "estimating size factors\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "60 rows did not converge in beta, labelled in mcols(object)$betaConv. Use larger maxit argument with nbinomWaldTest\n",
      "\n"
     ]
    }
   ],
   "source": [
    "dds = DESeqDataSetFromMatrix(countData=cts, colData=pdat, design = design_formula)\n",
    "#filter genes for low expression with edgeR\n",
    "y <- DGEList(counts=counts(dds), group=recur)\n",
    "keep <- filterByExpr(y)\n",
    "table(keep)\n",
    "y <- y[keep,]\n",
    "dds <- dds[keep,]\n",
    "#run DESeq\n",
    "dds <- DESeq(dds)\n",
    "res <- results(dds)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Attaching package: ‘genefilter’\n",
      "\n",
      "\n",
      "The following object is masked _by_ ‘.GlobalEnv’:\n",
      "\n",
      "    varFilter\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:readr’:\n",
      "\n",
      "    spec\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:MatrixGenerics’:\n",
      "\n",
      "    rowSds, rowVars\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    rowSds, rowVars\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library('pheatmap')\n",
    "library('genefilter')\n",
    "#use RemoveBatchEffect to adjust for the identified covariates\n",
    "#first create variables for each of the RUV factors\n",
    "ruv1 = pdat$W_1\n",
    "ruv2 = pdat$W_2\n",
    "ruv3 = pdat$W_3\n",
    "ruv4 = pdat$W_4\n",
    "ruv5 = pdat$W_5\n",
    "ruv6 = pdat$W_6\n",
    "ruv7 = pdat$W_7\n",
    "ruv8 = pdat$W_8\n",
    "ruv9 = pdat$W_9\n",
    "ruv10 = pdat$W_10\n",
    "ruv11 = pdat$W_11\n",
    "ruv12 = pdat$W_12\n",
    "ruv13 = pdat$W_13\n",
    "ruv14 = pdat$W_14\n",
    "ruv15 = pdat$W_15\n",
    "ruv16 = pdat$W_16\n",
    "ruv17 = pdat$W_17\n",
    "ruv18 = pdat$W_18\n",
    "ruv19 = pdat$W_19\n",
    "\n",
    "#run RemoveBatchEffect using the RUV factors and sex\n",
    "vsd = vst(dds)\n",
    "mat= assay(vsd)\n",
    "assay(vsd) <- removeBatchEffect(assay(vsd), batch=sex, covariates=cbind(ruv1, ruv2, ruv3, ruv4, ruv5, ruv6, ruv7, ruv8, ruv9, ruv10, ruv11, ruv12, ruv13, ruv14, ruv15, ruv16, ruv17, ruv18, ruv19))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong>png:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{png:} 2"
      ],
      "text/markdown": [
       "**png:** 2"
      ],
      "text/plain": [
       "png \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#plot a PCA of the adjusted data\n",
    "pcaData <- plotPCA(vsd, intgroup=c(\"recur\"), ntop = 500, returnData=TRUE)\n",
    "percentVar <- round(100 * attr(pcaData, \"percentVar\"))\n",
    "\n",
    "#save the plot\n",
    "png(\"results/c/pca_ruvg_corrected.png\", width=1024, height=1024)\n",
    "ggplot(pcaData, aes(PC1, PC2, color=inflammation_status, shape=recur)) +\n",
    "    geom_point(size=5) +\n",
    "    #geom_text(aes(label=sample),hjust=0, vjust=0) +\n",
    "    xlab(paste0(\"PC1: \",percentVar[1],\"% variance\")) +\n",
    "    ylab(paste0(\"PC2: \",percentVar[2],\"% variance\")) + \n",
    "    coord_fixed() +\n",
    "    \n",
    "    theme(text = element_text(size = 24))\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "log2 fold change (MLE): recur Recurrent vs Non.recurrent \n",
       "Wald test p-value: recur Recurrent vs Non.recurrent \n",
       "DataFrame with 6 rows and 7 columns\n",
       "                                 gene  baseMean log2FoldChange     lfcSE\n",
       "                          <character> <numeric>      <numeric> <numeric>\n",
       "ENSG00000149781.13 ENSG00000149781.13  748.2077       0.402831 0.0550133\n",
       "ENSG00000283537.2   ENSG00000283537.2   53.6663      11.216101 1.6699201\n",
       "ENSG00000026751.17 ENSG00000026751.17 1805.7035       0.494217 0.0860967\n",
       "ENSG00000122188.13 ENSG00000122188.13  899.5152       0.398828 0.0694072\n",
       "ENSG00000136143.16 ENSG00000136143.16  746.5610      -0.253959 0.0438884\n",
       "ENSG00000187664.9   ENSG00000187664.9  124.4072      -4.925317 0.8708563\n",
       "                        stat      pvalue        padj\n",
       "                   <numeric>   <numeric>   <numeric>\n",
       "ENSG00000149781.13   7.32243 2.43513e-13 7.07918e-09\n",
       "ENSG00000283537.2    6.71655 1.86078e-11 2.70474e-07\n",
       "ENSG00000026751.17   5.74026 9.45338e-09 5.49638e-05\n",
       "ENSG00000122188.13   5.74621 9.12680e-09 5.49638e-05\n",
       "ENSG00000136143.16  -5.78647 7.18811e-09 5.49638e-05\n",
       "ENSG00000187664.9   -5.65572 1.55197e-08 7.51954e-05"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "resOrdered <- res[order(res$padj),]\n",
    "resOrdered$gene <- rownames(resOrdered)\n",
    "resOrdered = resOrdered[,c(7, 1, 2, 3, 4, 5, 6)]\n",
    "head(resOrdered)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "#make a list of the top 15 genes\n",
    "top15 <- resOrdered[1:15,]\n",
    "recur <- data.frame(recur=coldata$recur)\n",
    "rownames(recur) <- colnames(mat)\n",
    "mat15 <- mat[top15$gene,]\n",
    "#center the data\n",
    "mat15 <- mat15 - rowMeans(mat15)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "'select()' returned 1:1 mapping between keys and columns\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A matrix: 6 × 41 of type dbl</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>PB12_770_A1</th><th scope=col>PB138_A8</th><th scope=col>PB140_COLON</th><th scope=col>PB146_A2</th><th scope=col>PB17_236_122_A5</th><th scope=col>PB17_236_177_A6</th><th scope=col>PB17_236_292_B2</th><th scope=col>PB17_236_387_A10</th><th scope=col>PB17_236_398_A2</th><th scope=col>PB17_236_520_A2</th><th scope=col>⋯</th><th scope=col>PB600444_A2</th><th scope=col>PB601102_A3</th><th scope=col>PB601990_A8</th><th scope=col>PB602717_A9</th><th scope=col>PB603387_A2</th><th scope=col>PB603967_A7</th><th scope=col>PB604520_A3</th><th scope=col>PB605312_A5</th><th scope=col>PB75_A10</th><th scope=col>PBCHAMP7_A12</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>FERMT3</th><td>-0.7510937</td><td> 0.3249750</td><td> 0.3277472</td><td> 0.3821316</td><td>-0.62266378</td><td>-0.2517798</td><td> 0.2314318</td><td>-0.4673792</td><td>-0.51300306</td><td> 0.16861463</td><td>⋯</td><td>-0.01658237</td><td>-0.3668822</td><td>-0.2437442</td><td>-0.1456490</td><td>-0.02950445</td><td> 0.2709670</td><td>-0.422292903</td><td>-0.7089087</td><td>-0.5738292</td><td> 0.5743894</td></tr>\n",
       "\t<tr><th scope=row>NA</th><td>-1.0111410</td><td>-1.0111410</td><td> 1.4590577</td><td>-1.0111410</td><td>-1.01114102</td><td>-1.0111410</td><td>-1.0111410</td><td> 1.0562515</td><td>-1.01114102</td><td>-1.01114102</td><td>⋯</td><td> 1.14711129</td><td> 1.1604861</td><td> 1.2882035</td><td> 1.2433581</td><td>-1.01114102</td><td>-1.0111410</td><td> 0.819564235</td><td>-0.5062618</td><td>-1.0111410</td><td>-1.0111410</td></tr>\n",
       "\t<tr><th scope=row>SLAMF7</th><td>-1.5319203</td><td> 0.4619804</td><td> 0.6576220</td><td> 0.1394215</td><td>-0.07947141</td><td>-0.7006812</td><td> 0.5012091</td><td>-0.6652370</td><td>-0.84395437</td><td>-0.63449251</td><td>⋯</td><td> 0.02523538</td><td>-0.6620506</td><td>-0.6479114</td><td>-1.0108806</td><td>-0.51213335</td><td> 1.1175265</td><td> 0.006795738</td><td>-0.7720085</td><td>-0.8488824</td><td> 0.4339631</td></tr>\n",
       "\t<tr><th scope=row>LAX1</th><td>-0.7384398</td><td> 0.1949079</td><td> 0.7841796</td><td> 0.2600180</td><td>-0.58402782</td><td>-0.5547324</td><td>-0.1605173</td><td>-0.6875512</td><td>-0.69550777</td><td>-0.38477126</td><td>⋯</td><td>-0.25998249</td><td>-0.1684362</td><td>-0.6411026</td><td>-0.5767867</td><td>-0.13229522</td><td> 1.0361546</td><td>-0.221319885</td><td>-0.7557723</td><td>-0.5795373</td><td> 0.5443262</td></tr>\n",
       "\t<tr><th scope=row>SUCLA2</th><td> 0.2027654</td><td>-0.1768142</td><td>-0.4738033</td><td>-0.2167849</td><td> 0.09179106</td><td> 0.3543726</td><td> 0.2749289</td><td> 0.4106331</td><td>-0.09847058</td><td> 0.06509136</td><td>⋯</td><td> 0.01240000</td><td>-0.4122660</td><td> 0.2877900</td><td> 0.1509653</td><td>-0.06936091</td><td>-0.3355511</td><td>-0.162408832</td><td> 0.2202813</td><td>-0.2045929</td><td>-0.4229938</td></tr>\n",
       "\t<tr><th scope=row>HAPLN4</th><td> 1.5516064</td><td>-1.9812658</td><td>-1.9812658</td><td>-1.9812658</td><td>-1.98126582</td><td> 0.5855538</td><td> 0.4066118</td><td>-1.9812658</td><td> 1.29276046</td><td> 0.87676949</td><td>⋯</td><td>-0.54964236</td><td> 1.9007916</td><td>-1.9812658</td><td> 1.0444594</td><td>-1.98126582</td><td> 0.4971795</td><td>-1.981265819</td><td>-1.9812658</td><td>-1.9812658</td><td> 0.6304947</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A matrix: 6 × 41 of type dbl\n",
       "\\begin{tabular}{r|lllllllllllllllllllll}\n",
       "  & PB12\\_770\\_A1 & PB138\\_A8 & PB140\\_COLON & PB146\\_A2 & PB17\\_236\\_122\\_A5 & PB17\\_236\\_177\\_A6 & PB17\\_236\\_292\\_B2 & PB17\\_236\\_387\\_A10 & PB17\\_236\\_398\\_A2 & PB17\\_236\\_520\\_A2 & ⋯ & PB600444\\_A2 & PB601102\\_A3 & PB601990\\_A8 & PB602717\\_A9 & PB603387\\_A2 & PB603967\\_A7 & PB604520\\_A3 & PB605312\\_A5 & PB75\\_A10 & PBCHAMP7\\_A12\\\\\n",
       "\\hline\n",
       "\tFERMT3 & -0.7510937 &  0.3249750 &  0.3277472 &  0.3821316 & -0.62266378 & -0.2517798 &  0.2314318 & -0.4673792 & -0.51300306 &  0.16861463 & ⋯ & -0.01658237 & -0.3668822 & -0.2437442 & -0.1456490 & -0.02950445 &  0.2709670 & -0.422292903 & -0.7089087 & -0.5738292 &  0.5743894\\\\\n",
       "\tNA & -1.0111410 & -1.0111410 &  1.4590577 & -1.0111410 & -1.01114102 & -1.0111410 & -1.0111410 &  1.0562515 & -1.01114102 & -1.01114102 & ⋯ &  1.14711129 &  1.1604861 &  1.2882035 &  1.2433581 & -1.01114102 & -1.0111410 &  0.819564235 & -0.5062618 & -1.0111410 & -1.0111410\\\\\n",
       "\tSLAMF7 & -1.5319203 &  0.4619804 &  0.6576220 &  0.1394215 & -0.07947141 & -0.7006812 &  0.5012091 & -0.6652370 & -0.84395437 & -0.63449251 & ⋯ &  0.02523538 & -0.6620506 & -0.6479114 & -1.0108806 & -0.51213335 &  1.1175265 &  0.006795738 & -0.7720085 & -0.8488824 &  0.4339631\\\\\n",
       "\tLAX1 & -0.7384398 &  0.1949079 &  0.7841796 &  0.2600180 & -0.58402782 & -0.5547324 & -0.1605173 & -0.6875512 & -0.69550777 & -0.38477126 & ⋯ & -0.25998249 & -0.1684362 & -0.6411026 & -0.5767867 & -0.13229522 &  1.0361546 & -0.221319885 & -0.7557723 & -0.5795373 &  0.5443262\\\\\n",
       "\tSUCLA2 &  0.2027654 & -0.1768142 & -0.4738033 & -0.2167849 &  0.09179106 &  0.3543726 &  0.2749289 &  0.4106331 & -0.09847058 &  0.06509136 & ⋯ &  0.01240000 & -0.4122660 &  0.2877900 &  0.1509653 & -0.06936091 & -0.3355511 & -0.162408832 &  0.2202813 & -0.2045929 & -0.4229938\\\\\n",
       "\tHAPLN4 &  1.5516064 & -1.9812658 & -1.9812658 & -1.9812658 & -1.98126582 &  0.5855538 &  0.4066118 & -1.9812658 &  1.29276046 &  0.87676949 & ⋯ & -0.54964236 &  1.9007916 & -1.9812658 &  1.0444594 & -1.98126582 &  0.4971795 & -1.981265819 & -1.9812658 & -1.9812658 &  0.6304947\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A matrix: 6 × 41 of type dbl\n",
       "\n",
       "| <!--/--> | PB12_770_A1 | PB138_A8 | PB140_COLON | PB146_A2 | PB17_236_122_A5 | PB17_236_177_A6 | PB17_236_292_B2 | PB17_236_387_A10 | PB17_236_398_A2 | PB17_236_520_A2 | ⋯ | PB600444_A2 | PB601102_A3 | PB601990_A8 | PB602717_A9 | PB603387_A2 | PB603967_A7 | PB604520_A3 | PB605312_A5 | PB75_A10 | PBCHAMP7_A12 |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| FERMT3 | -0.7510937 |  0.3249750 |  0.3277472 |  0.3821316 | -0.62266378 | -0.2517798 |  0.2314318 | -0.4673792 | -0.51300306 |  0.16861463 | ⋯ | -0.01658237 | -0.3668822 | -0.2437442 | -0.1456490 | -0.02950445 |  0.2709670 | -0.422292903 | -0.7089087 | -0.5738292 |  0.5743894 |\n",
       "| NA | -1.0111410 | -1.0111410 |  1.4590577 | -1.0111410 | -1.01114102 | -1.0111410 | -1.0111410 |  1.0562515 | -1.01114102 | -1.01114102 | ⋯ |  1.14711129 |  1.1604861 |  1.2882035 |  1.2433581 | -1.01114102 | -1.0111410 |  0.819564235 | -0.5062618 | -1.0111410 | -1.0111410 |\n",
       "| SLAMF7 | -1.5319203 |  0.4619804 |  0.6576220 |  0.1394215 | -0.07947141 | -0.7006812 |  0.5012091 | -0.6652370 | -0.84395437 | -0.63449251 | ⋯ |  0.02523538 | -0.6620506 | -0.6479114 | -1.0108806 | -0.51213335 |  1.1175265 |  0.006795738 | -0.7720085 | -0.8488824 |  0.4339631 |\n",
       "| LAX1 | -0.7384398 |  0.1949079 |  0.7841796 |  0.2600180 | -0.58402782 | -0.5547324 | -0.1605173 | -0.6875512 | -0.69550777 | -0.38477126 | ⋯ | -0.25998249 | -0.1684362 | -0.6411026 | -0.5767867 | -0.13229522 |  1.0361546 | -0.221319885 | -0.7557723 | -0.5795373 |  0.5443262 |\n",
       "| SUCLA2 |  0.2027654 | -0.1768142 | -0.4738033 | -0.2167849 |  0.09179106 |  0.3543726 |  0.2749289 |  0.4106331 | -0.09847058 |  0.06509136 | ⋯ |  0.01240000 | -0.4122660 |  0.2877900 |  0.1509653 | -0.06936091 | -0.3355511 | -0.162408832 |  0.2202813 | -0.2045929 | -0.4229938 |\n",
       "| HAPLN4 |  1.5516064 | -1.9812658 | -1.9812658 | -1.9812658 | -1.98126582 |  0.5855538 |  0.4066118 | -1.9812658 |  1.29276046 |  0.87676949 | ⋯ | -0.54964236 |  1.9007916 | -1.9812658 |  1.0444594 | -1.98126582 |  0.4971795 | -1.981265819 | -1.9812658 | -1.9812658 |  0.6304947 |\n",
       "\n"
      ],
      "text/plain": [
       "       PB12_770_A1 PB138_A8   PB140_COLON PB146_A2   PB17_236_122_A5\n",
       "FERMT3 -0.7510937   0.3249750  0.3277472   0.3821316 -0.62266378    \n",
       "<NA>   -1.0111410  -1.0111410  1.4590577  -1.0111410 -1.01114102    \n",
       "SLAMF7 -1.5319203   0.4619804  0.6576220   0.1394215 -0.07947141    \n",
       "LAX1   -0.7384398   0.1949079  0.7841796   0.2600180 -0.58402782    \n",
       "SUCLA2  0.2027654  -0.1768142 -0.4738033  -0.2167849  0.09179106    \n",
       "HAPLN4  1.5516064  -1.9812658 -1.9812658  -1.9812658 -1.98126582    \n",
       "       PB17_236_177_A6 PB17_236_292_B2 PB17_236_387_A10 PB17_236_398_A2\n",
       "FERMT3 -0.2517798       0.2314318      -0.4673792       -0.51300306    \n",
       "<NA>   -1.0111410      -1.0111410       1.0562515       -1.01114102    \n",
       "SLAMF7 -0.7006812       0.5012091      -0.6652370       -0.84395437    \n",
       "LAX1   -0.5547324      -0.1605173      -0.6875512       -0.69550777    \n",
       "SUCLA2  0.3543726       0.2749289       0.4106331       -0.09847058    \n",
       "HAPLN4  0.5855538       0.4066118      -1.9812658        1.29276046    \n",
       "       PB17_236_520_A2 ⋯ PB600444_A2 PB601102_A3 PB601990_A8 PB602717_A9\n",
       "FERMT3  0.16861463     ⋯ -0.01658237 -0.3668822  -0.2437442  -0.1456490 \n",
       "<NA>   -1.01114102     ⋯  1.14711129  1.1604861   1.2882035   1.2433581 \n",
       "SLAMF7 -0.63449251     ⋯  0.02523538 -0.6620506  -0.6479114  -1.0108806 \n",
       "LAX1   -0.38477126     ⋯ -0.25998249 -0.1684362  -0.6411026  -0.5767867 \n",
       "SUCLA2  0.06509136     ⋯  0.01240000 -0.4122660   0.2877900   0.1509653 \n",
       "HAPLN4  0.87676949     ⋯ -0.54964236  1.9007916  -1.9812658   1.0444594 \n",
       "       PB603387_A2 PB603967_A7 PB604520_A3  PB605312_A5 PB75_A10   PBCHAMP7_A12\n",
       "FERMT3 -0.02950445  0.2709670  -0.422292903 -0.7089087  -0.5738292  0.5743894  \n",
       "<NA>   -1.01114102 -1.0111410   0.819564235 -0.5062618  -1.0111410 -1.0111410  \n",
       "SLAMF7 -0.51213335  1.1175265   0.006795738 -0.7720085  -0.8488824  0.4339631  \n",
       "LAX1   -0.13229522  1.0361546  -0.221319885 -0.7557723  -0.5795373  0.5443262  \n",
       "SUCLA2 -0.06936091 -0.3355511  -0.162408832  0.2202813  -0.2045929 -0.4229938  \n",
       "HAPLN4 -1.98126582  0.4971795  -1.981265819 -1.9812658  -1.9812658  0.6304947  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#rename the rownames of mat with gene symbols\n",
    "library(\"AnnotationDbi\")\n",
    "library(\"org.Hs.eg.db\")\n",
    "\n",
    "#remove the ENSEMBL version number\n",
    "rownames(mat15) <- gsub(\"\\\\..*\", \"\", rownames(mat15))\n",
    "rownames(mat15) <- mapIds(org.Hs.eg.db, keys=rownames(mat15), column=\"SYMBOL\", keytype=\"ENSEMBL\", multiVals=\"first\")\n",
    "head(mat15)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "#sort each dataframe by recurrence\n",
    "#first sort the annotation data by recurrence while keeping it as a dataframe\n",
    "recur <- recur[order(recur$recur), , drop=FALSE]\n",
    "#then sort the matrix by the order of the annotation data by getting the rownames of the annotation data and reordering the columns of the matrix\n",
    "col_order = rownames(recur)\n",
    "mat15 <- mat15[,col_order]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "#remove the sample names from the heatmap\n",
    "pheatmap(mat15, annotation_col=recur, cluster_cols=F, show_colnames=F, filename='../../results/recur/c/ruvg_runs/heatmap.png', height=10, width=10)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.0"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "916dbcbb3f70747c44a77c7bcd40155683ae19c65e1c03b4aa3499c5328201f1"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
