{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: ggplot2\n",
      "\n",
      "\n",
      "Attaching package: ‘plotly’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:ggplot2’:\n",
      "\n",
      "    last_plot\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:XVector’:\n",
      "\n",
      "    slice\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:AnnotationDbi’:\n",
      "\n",
      "    select\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    slice\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:S4Vectors’:\n",
      "\n",
      "    rename\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:stats’:\n",
      "\n",
      "    filter\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:graphics’:\n",
      "\n",
      "    layout\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘plyr’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:plotly’:\n",
      "\n",
      "    arrange, mutate, rename, summarise\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:ShortRead’:\n",
      "\n",
      "    id\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:matrixStats’:\n",
      "\n",
      "    count\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:XVector’:\n",
      "\n",
      "    compact\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    desc\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:S4Vectors’:\n",
      "\n",
      "    rename\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:plyr’:\n",
      "\n",
      "    arrange, count, desc, failwith, id, mutate, rename, summarise,\n",
      "    summarize\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:ShortRead’:\n",
      "\n",
      "    id\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:GenomicAlignments’:\n",
      "\n",
      "    first, last\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:matrixStats’:\n",
      "\n",
      "    count\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:GenomicRanges’:\n",
      "\n",
      "    intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:Biostrings’:\n",
      "\n",
      "    collapse, intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomeInfoDb’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:XVector’:\n",
      "\n",
      "    slice\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:AnnotationDbi’:\n",
      "\n",
      "    select\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:IRanges’:\n",
      "\n",
      "    collapse, desc, intersect, setdiff, slice, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:S4Vectors’:\n",
      "\n",
      "    first, intersect, rename, setdiff, setequal, union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:Biobase’:\n",
      "\n",
      "    combine\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "Registered S3 method overwritten by 'GGally':\n",
      "  method from   \n",
      "  +.gg   ggplot2\n",
      "\n",
      "── \u001b[1mAttaching core tidyverse packages\u001b[22m ──────────────────────── tidyverse 2.0.0 ──\n",
      "\u001b[32m✔\u001b[39m \u001b[34mforcats  \u001b[39m 1.0.0     \u001b[32m✔\u001b[39m \u001b[34mstringr  \u001b[39m 1.5.0\n",
      "\u001b[32m✔\u001b[39m \u001b[34mlubridate\u001b[39m 1.9.2     \u001b[32m✔\u001b[39m \u001b[34mtibble   \u001b[39m 3.2.1\n",
      "\u001b[32m✔\u001b[39m \u001b[34mpurrr    \u001b[39m 1.0.1     \u001b[32m✔\u001b[39m \u001b[34mtidyr    \u001b[39m 1.3.0\n",
      "\u001b[32m✔\u001b[39m \u001b[34mreadr    \u001b[39m 2.1.4     \n",
      "── \u001b[1mConflicts\u001b[22m ────────────────────────────────────────── tidyverse_conflicts() ──\n",
      "\u001b[31m✖\u001b[39m \u001b[34mlubridate\u001b[39m::\u001b[32m%within%()\u001b[39m masks \u001b[34mIRanges\u001b[39m::%within%()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32marrange()\u001b[39m      masks \u001b[34mplyr\u001b[39m::arrange(), \u001b[34mplotly\u001b[39m::arrange()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mcollapse()\u001b[39m     masks \u001b[34mBiostrings\u001b[39m::collapse(), \u001b[34mIRanges\u001b[39m::collapse()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mcombine()\u001b[39m      masks \u001b[34mBiobase\u001b[39m::combine(), \u001b[34mBiocGenerics\u001b[39m::combine()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mpurrr\u001b[39m::\u001b[32mcompact()\u001b[39m      masks \u001b[34mplyr\u001b[39m::compact(), \u001b[34mXVector\u001b[39m::compact()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mpurrr\u001b[39m::\u001b[32mcompose()\u001b[39m      masks \u001b[34mShortRead\u001b[39m::compose()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mcount()\u001b[39m        masks \u001b[34mplyr\u001b[39m::count(), \u001b[34mmatrixStats\u001b[39m::count()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mdesc()\u001b[39m         masks \u001b[34mplyr\u001b[39m::desc(), \u001b[34mIRanges\u001b[39m::desc()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mtidyr\u001b[39m::\u001b[32mexpand()\u001b[39m       masks \u001b[34mS4Vectors\u001b[39m::expand()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mfailwith()\u001b[39m     masks \u001b[34mplyr\u001b[39m::failwith()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mfilter()\u001b[39m       masks \u001b[34mplotly\u001b[39m::filter(), \u001b[34mstats\u001b[39m::filter()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mfirst()\u001b[39m        masks \u001b[34mGenomicAlignments\u001b[39m::first(), \u001b[34mS4Vectors\u001b[39m::first()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mid()\u001b[39m           masks \u001b[34mplyr\u001b[39m::id(), \u001b[34mShortRead\u001b[39m::id()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mlag()\u001b[39m          masks \u001b[34mstats\u001b[39m::lag()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mlast()\u001b[39m         masks \u001b[34mGenomicAlignments\u001b[39m::last()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mmutate()\u001b[39m       masks \u001b[34mplyr\u001b[39m::mutate(), \u001b[34mplotly\u001b[39m::mutate()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mggplot2\u001b[39m::\u001b[32mPosition()\u001b[39m   masks \u001b[34mBiocGenerics\u001b[39m::Position(), \u001b[34mbase\u001b[39m::Position()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mpurrr\u001b[39m::\u001b[32mreduce()\u001b[39m       masks \u001b[34mGenomicRanges\u001b[39m::reduce(), \u001b[34mIRanges\u001b[39m::reduce()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mrename()\u001b[39m       masks \u001b[34mplyr\u001b[39m::rename(), \u001b[34mplotly\u001b[39m::rename(), \u001b[34mS4Vectors\u001b[39m::rename()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mlubridate\u001b[39m::\u001b[32msecond()\u001b[39m   masks \u001b[34mGenomicAlignments\u001b[39m::second(), \u001b[34mS4Vectors\u001b[39m::second()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mlubridate\u001b[39m::\u001b[32msecond<-()\u001b[39m masks \u001b[34mS4Vectors\u001b[39m::second<-()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mselect()\u001b[39m       masks \u001b[34mplotly\u001b[39m::select(), \u001b[34mAnnotationDbi\u001b[39m::select()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mslice()\u001b[39m        masks \u001b[34mplotly\u001b[39m::slice(), \u001b[34mXVector\u001b[39m::slice(), \u001b[34mIRanges\u001b[39m::slice()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32msummarise()\u001b[39m    masks \u001b[34mplyr\u001b[39m::summarise(), \u001b[34mplotly\u001b[39m::summarise()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32msummarize()\u001b[39m    masks \u001b[34mplyr\u001b[39m::summarize()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mtibble\u001b[39m::\u001b[32mview()\u001b[39m        masks \u001b[34mShortRead\u001b[39m::view()\n",
      "\u001b[36mℹ\u001b[39m Use the conflicted package (\u001b[3m\u001b[34m<http://conflicted.r-lib.org/>\u001b[39m\u001b[23m) to force all conflicts to become errors\n",
      "\n",
      "Attaching package: ‘scales’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:purrr’:\n",
      "\n",
      "    discard\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:readr’:\n",
      "\n",
      "    col_factor\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘reshape2’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:tidyr’:\n",
      "\n",
      "    smiths\n",
      "\n",
      "\n",
      "========================================\n",
      "circlize version 0.4.15\n",
      "CRAN page: https://cran.r-project.org/package=circlize\n",
      "Github page: https://github.com/jokergoo/circlize\n",
      "Documentation: https://jokergoo.github.io/circlize_book/book/\n",
      "\n",
      "If you use it in published research, please cite:\n",
      "Gu, Z. circlize implements and enhances circular visualization\n",
      "  in R. Bioinformatics 2014.\n",
      "\n",
      "This message can be suppressed by:\n",
      "  suppressPackageStartupMessages(library(circlize))\n",
      "========================================\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘gplots’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    space\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:S4Vectors’:\n",
      "\n",
      "    space\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:stats’:\n",
      "\n",
      "    lowess\n",
      "\n",
      "\n",
      "Loading required package: grid\n",
      "\n",
      "\n",
      "Attaching package: ‘grid’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:Biostrings’:\n",
      "\n",
      "    pattern\n",
      "\n",
      "\n",
      "Loading required package: futile.logger\n",
      "\n",
      "'getOption(\"repos\")' replaces Bioconductor standard repositories, see\n",
      "'help(\"repositories\", package = \"BiocManager\")' for details.\n",
      "Replacement repositories:\n",
      "    CRAN: https://cloud.r-project.org\n",
      "\n",
      "Bioconductor version 3.17 (BiocManager 1.30.20), R 4.3.0 (2023-04-21)\n",
      "\n",
      "Warning message:\n",
      "“package(s) not installed when version(s) same as or greater than current; use\n",
      "  `force = TRUE` to re-install: 'DESeq2'”\n",
      "Installation paths not writeable, unable to update packages\n",
      "  path: /usr/lib/R/library\n",
      "  packages:\n",
      "    boot, foreign, Matrix\n",
      "  path: /usr/lib/R/site-library\n",
      "  packages:\n",
      "    RcppArmadillo\n",
      "\n",
      "Installing packages into ‘/home/kchen/R/x86_64-pc-linux-gnu-library/4.3’\n",
      "(as ‘lib’ is unspecified)\n",
      "\n",
      "'getOption(\"repos\")' replaces Bioconductor standard repositories, see\n",
      "'help(\"repositories\", package = \"BiocManager\")' for details.\n",
      "Replacement repositories:\n",
      "    CRAN: https://cloud.r-project.org\n",
      "\n",
      "Bioconductor version 3.17 (BiocManager 1.30.20), R 4.3.0 (2023-04-21)\n",
      "\n",
      "Warning message:\n",
      "“package(s) not installed when version(s) same as or greater than current; use\n",
      "  `force = TRUE` to re-install: 'edgeR' 'preprocessCore' 'limma'”\n",
      "Installation paths not writeable, unable to update packages\n",
      "  path: /usr/lib/R/library\n",
      "  packages:\n",
      "    boot, foreign, Matrix\n",
      "  path: /usr/lib/R/site-library\n",
      "  packages:\n",
      "    RcppArmadillo\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(\"AnnotationDbi\")\n",
    "library(\"org.Hs.eg.db\")\n",
    "library(RUVSeq)\n",
    "source(\"../downstreamAnalysis_RNAseqFunctions.R\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>42</li><li>16</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 42\n",
       "\\item 16\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 42\n",
       "2. 16\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] 42 16"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>61114</li><li>42</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 61114\n",
       "\\item 42\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 61114\n",
       "2. 42\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] 61114    42"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Read in data\n",
    "cts = as.matrix(read.csv(\"../../data/cts_i.csv\", sep=',', row.names='gene'))\n",
    "coldata = read.csv('../../data/coldata_i.csv', sep=',', row.names='Sample_ID')\n",
    "\n",
    "cts = cts[,-which(colnames(cts) == 'PB17_236_257_A5')]\n",
    "coldata = coldata[-which(rownames(coldata) == 'PB17_236_257_A5'),]\n",
    "cts = cts[,-which(colnames(cts) == 'PB600298_A4')]\n",
    "coldata = coldata[-which(rownames(coldata) == 'PB600298_A4'),]\n",
    "cts = cts[,-which(colnames(cts) == 'PB43300_A1')]\n",
    "coldata = coldata[-which(rownames(coldata) == 'PB43300_A1'),]\n",
    "\n",
    "\n",
    "dim(coldata)\n",
    "dim(cts)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "TRUE"
      ],
      "text/latex": [
       "TRUE"
      ],
      "text/markdown": [
       "TRUE"
      ],
      "text/plain": [
       "[1] TRUE"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "converting counts to integer mode\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "factor_cols = c('sex', 'recur', 'inflammation_status', 'source')\n",
    "for (col in factor_cols) {\n",
    "  coldata[,col] = as.factor(coldata[,col])\n",
    "}\n",
    "sex = coldata$sex\n",
    "source = coldata$source\n",
    "# tissue = coldata$tissue\n",
    "recur = coldata$recur\n",
    "inflammation_status = coldata$inflammation_status\n",
    "tin = coldata$TIN\n",
    "\n",
    "sample = rownames(coldata)\n",
    "\n",
    "#round counts to integers\n",
    "cts <- round(cts);\n",
    "\n",
    "#check if row names of coldata match colnames of cts\n",
    "all(rownames(coldata) == colnames(cts))\n",
    "\n",
    "#adjust for batch and sexBin\n",
    "design <- as.formula(~ recur)\n",
    "#create the dds object\n",
    "dds = DESeqDataSetFromMatrix(countData = cts, colData = coldata, design = design)\n",
    "\n",
    "vsd = vst(dds)\n",
    "mat = assay(vsd)\n",
    "#PCA of raw data\n",
    "pcaData <- plotPCA(vsd, intgroup=c(\"recur\"), ntop = 500, returnData=TRUE)\n",
    "percentVar <- round(100 * attr(pcaData, \"percentVar\"))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "\n",
    "# #save the plot\n",
    "# png(\"results/i/pca_source_raw.png\", width=1024, height=512)\n",
    "# ggplot(pcaData, aes(PC1, PC2, color=source, shape=recur)) +\n",
    "#     geom_point(size=5) +\n",
    "#     #geom_text(aes(label=sample),hjust=0, vjust=0) +\n",
    "#     xlab(paste0(\"PC1: \",percentVar[1],\"% variance\")) +\n",
    "#     ylab(paste0(\"PC2: \",percentVar[2],\"% variance\")) + \n",
    "#     coord_fixed() +\n",
    "    \n",
    "#     theme(text = element_text(size = 24))\n",
    "# dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Import genomic features from the file as a GRanges object ... \n",
      "OK\n",
      "\n",
      "Prepare the 'metadata' data frame ... \n",
      "OK\n",
      "\n",
      "Make the TxDb object ... \n",
      "Warning message in .get_cds_IDX(mcols0$type, mcols0$phase):\n",
      "“The \"phase\" metadata column contains non-NA values for features of type\n",
      "  stop_codon. This information was ignored.”\n",
      "OK\n",
      "\n",
      "'select()' returned 1:1 mapping between keys and columns\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library('GenomicFeatures')\n",
    "#use the GenomicFeatures package to create a transcript database (txDb) file from our GENCODE annotation file, and use it to pull out the chromosomal coordinates for each of our transcripts (including which chromosome it is on)\n",
    "txdb = makeTxDbFromGFF('../../data/gencode.v39.annotation.gtf', dbxrefTag = \"gene_id\", organism = 'Homo sapiens')\n",
    "k <- keys(txdb, keytype = 'TXNAME')\n",
    "tx2gene <- AnnotationDbi::select(txdb, keys=k, columns=c('GENEID', 'TXCHROM', 'TXSTART', 'TXEND', 'TXSTRAND'), keytype='TXNAME')\n",
    "# make a subset of the list containing only the Ensembl gene IDs and the names of the chromosomes, and create a subset for all the genes on chromosomes X and Y\n",
    "tx2gene_short = tx2gene[,2:3]\n",
    "tx2gene_short = unique(tx2gene_short)\n",
    "tx2gene_short_xychrom = tx2gene_short[tx2gene_short$TXCHROM==\"chrX\" | tx2gene_short$TXCHROM==\"chrY\",]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>58219</li><li>42</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 58219\n",
       "\\item 42\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 58219\n",
       "2. 42\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] 58219    42"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#now remove the rows from cts if they are present in 'GENEID' in tx2gene_short_xychrom\n",
    "cts <- cts[! match(rownames(cts), tx2gene_short_xychrom$GENEID, nomatch=0),]\n",
    "dim(cts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "keep\n",
       "FALSE  TRUE \n",
       "30087 31027 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "estimating size factors\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "-- replacing outliers and refitting for 151 genes\n",
      "-- DESeq argument 'minReplicatesForReplace' = 7 \n",
      "-- original counts are preserved in counts(dds)\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#DESeq on the raw data to identify control genes\n",
    "#filter genes for low expression with edgeR\n",
    "y <- DGEList(counts=counts(dds), group=recur)\n",
    "keep <- filterByExpr(y)\n",
    "table(keep)\n",
    "y <- y[keep,]\n",
    "dds <- dds[keep,]\n",
    "\n",
    "dds <- DESeq(dds)\n",
    "res <- results(dds)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\n",
       "FALSE  TRUE \n",
       "14351 16676 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "table(res$pvalue < 0.5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "set <- newSeqExpressionSet(counts(dds))\n",
    "set <- betweenLaneNormalization(set, which='upper')\n",
    "not_sig <- rownames(res)[which(res$pvalue > 0.5)]\n",
    "empirical <- rownames(set)[ rownames(set) %in% not_sig ]\n",
    "set <- RUVg(set, empirical, k=20)\n",
    "pdat <- pData(set)\n",
    "pdat$recur <- recur\n",
    "pdat$sex <- sex\n",
    "pdat$inf <- inflammation_status\n",
    "pdat$source <- source"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "#write out the ruv factors to csv\n",
    "write.csv(pdat, file = \"../../results/recur/i/data_ruvg.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "formula_list = list(\n",
    "~ sex + W_1 + recur, \n",
    "~ sex + W_1 + W_2 + recur, \n",
    "~ sex + W_1 + W_2 + W_3 + recur, \n",
    "~ sex + W_1 + W_2 + W_3 + W_4 + recur, \n",
    "~ sex + W_1 + W_2 + W_3 + W_4 + W_5 + recur, \n",
    "~ sex + W_1 + W_2 + W_3 + W_4 + W_5 + W_6 + recur, \n",
    "~ sex + W_1 + W_2 + W_3 + W_4 + W_5 + W_6 + W_7 + recur, \n",
    "~ sex + W_1 + W_2 + W_3 + W_4 + W_5 + W_6 + W_7 + W_8 + recur, \n",
    "~ sex + W_1 + W_2 + W_3 + W_4 + W_5 + W_6 + W_7 + W_8 + W_9 + recur, \n",
    "~ sex + W_1 + W_2 + W_3 + W_4 + W_5 + W_6 + W_7 + W_8 + W_9 + W_10 + recur,\n",
    "~ sex + W_1 + W_2 + W_3 + W_4 + W_5 + W_6 + W_7 + W_8 + W_9 + W_10 + W_11 + recur,\n",
    "~ sex + W_1 + W_2 + W_3 + W_4 + W_5 + W_6 + W_7 + W_8 + W_9 + W_10 + W_11 + W_12 + recur,\n",
    "~ sex + W_1 + W_2 + W_3 + W_4 + W_5 + W_6 + W_7 + W_8 + W_9 + W_10 + W_11 + W_12 + W_13 + recur,\n",
    "~ sex + W_1 + W_2 + W_3 + W_4 + W_5 + W_6 + W_7 + W_8 + W_9 + W_10 + W_11 + W_12 + W_13 + W_14 + recur,\n",
    "~ sex + W_1 + W_2 + W_3 + W_4 + W_5 + W_6 + W_7 + W_8 + W_9 + W_10 + W_11 + W_12 + W_13 + W_14 + + W_15 + recur,\n",
    "~ sex + W_1 + W_2 + W_3 + W_4 + W_5 + W_6 + W_7 + W_8 + W_9 + W_10 + W_11 + W_12 + W_13 + W_14 + + W_15 + W_16 + recur,\n",
    "~ sex + W_1 + W_2 + W_3 + W_4 + W_5 + W_6 + W_7 + W_8 + W_9 + W_10 + W_11 + W_12 + W_13 + W_14 + + W_15 + W_16 + W_17 + recur,\n",
    "~ sex + W_1 + W_2 + W_3 + W_4 + W_5 + W_6 + W_7 + W_8 + W_9 + W_10 + W_11 + W_12 + W_13 + W_14 + + W_15 + W_16 + W_17 + W_18 + recur,\n",
    "~ sex + W_1 + W_2 + W_3 + W_4 + W_5 + W_6 + W_7 + W_8 + W_9 + W_10 + W_11 + W_12 + W_13 + W_14 + + W_15 + W_16 + W_17 + W_18 + W_19 + recur,\n",
    "~ sex + W_1 + W_2 + W_3 + W_4 + W_5 + W_6 + W_7 + W_8 + W_9 + W_10 + W_11 + W_12 + W_13 + W_14 + + W_15 + W_16 + W_17 + W_18 + W_19 + W_20 + recur)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "converting counts to integer mode\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "estimating size factors\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "converting counts to integer mode\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "estimating size factors\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "converting counts to integer mode\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "estimating size factors\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "converting counts to integer mode\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "estimating size factors\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "converting counts to integer mode\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "estimating size factors\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "converting counts to integer mode\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "estimating size factors\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "converting counts to integer mode\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "estimating size factors\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "converting counts to integer mode\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "estimating size factors\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "1 rows did not converge in beta, labelled in mcols(object)$betaConv. Use larger maxit argument with nbinomWaldTest\n",
      "\n",
      "converting counts to integer mode\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "estimating size factors\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "2 rows did not converge in beta, labelled in mcols(object)$betaConv. Use larger maxit argument with nbinomWaldTest\n",
      "\n",
      "converting counts to integer mode\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "estimating size factors\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "7 rows did not converge in beta, labelled in mcols(object)$betaConv. Use larger maxit argument with nbinomWaldTest\n",
      "\n",
      "converting counts to integer mode\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "estimating size factors\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n"
     ]
    }
   ],
   "source": [
    "\n",
    "\n",
    "#create an empty dataframe to store the number of significant genes\n",
    "deg_counts = data.frame(matrix(ncol=20, nrow=0))\n",
    "colnames(deg_counts) = c(\"formula\", \"sig_genes\")\n",
    "\n",
    "for (x in seq_along(formula_list)) {\n",
    "    design_formula = formula(formula_list[[x]])\n",
    "    dds1 <- DESeqDataSetFromMatrix(countData=cts, colData=pdat, design = design_formula)\n",
    "    y <- DGEList(counts=counts(dds1), group=coldata$recur)\n",
    "    keep <- filterByExpr(y)\n",
    "    table(keep)\n",
    "    y <- y[keep,]\n",
    "    dds1 <- dds1[keep,]\n",
    "    dds1 <- DESeq(dds1)\n",
    "    res <- results(dds1)\n",
    "    res <- res[order(res$padj),]\n",
    "    #get the number of significant genes\n",
    "    sig_genes <- res[which(res$padj < 0.05),]\n",
    "    #save the number of significant genes to the dataframe\n",
    "    deg_counts = rbind(deg_counts, c(x, nrow(sig_genes)))\n",
    "    #save the results to a csv file\n",
    "    write.csv(res, paste0(\"../../results/recur/i/ruvg_runs/res5_\", x, \".csv\"))\n",
    "    \n",
    "}\n",
    "\n",
    "head(deg_counts, 10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA0gAAANICAMAAADKOT/pAAADAFBMVEUAAAABAQECAgIDAwMEBAQFBQUGBgYHBwcICAgJCQkKCgoLCwsMDAwNDQ0ODg4PDw8QEBARERESEhITExMUFBQVFRUWFhYXFxcYGBgZGRkaGhobGxscHBwdHR0eHh4fHx8gICAhISEiIiIjIyMkJCQlJSUmJiYnJycoKCgpKSkqKiorKyssLCwtLS0uLi4vLy8wMDAxMTEyMjIzMzM0NDQ1NTU2NjY3Nzc4ODg5OTk6Ojo7Ozs8PDw9PT0+Pj4/Pz9AQEBBQUFCQkJDQ0NERERFRUVGRkZHR0dISEhJSUlKSkpLS0tMTExNTU1OTk5PT09QUFBRUVFSUlJTU1NUVFRVVVVWVlZXV1dYWFhZWVlaWlpbW1tcXFxdXV1eXl5fX19gYGBhYWFiYmJjY2NkZGRlZWVmZmZnZ2doaGhpaWlqampra2tsbGxtbW1ubm5vb29wcHBxcXFycnJzc3N0dHR1dXV2dnZ3d3d4eHh5eXl6enp7e3t8fHx9fX1+fn5/f3+AgICBgYGCgoKDg4OEhISFhYWGhoaHh4eIiIiJiYmKioqLi4uMjIyNjY2Ojo6Pj4+QkJCRkZGSkpKTk5OUlJSVlZWWlpaXl5eYmJiZmZmampqbm5ucnJydnZ2enp6fn5+goKChoaGioqKjo6OkpKSlpaWmpqanp6eoqKipqamqqqqrq6usrKytra2urq6vr6+wsLCxsbGysrKzs7O0tLS1tbW2tra3t7e4uLi5ubm6urq7u7u8vLy9vb2+vr6/v7/AwMDBwcHCwsLDw8PExMTFxcXGxsbHx8fIyMjJycnKysrLy8vMzMzNzc3Ozs7Pz8/Q0NDR0dHS0tLT09PU1NTV1dXW1tbX19fY2NjZ2dna2trb29vc3Nzd3d3e3t7f39/g4ODh4eHi4uLj4+Pk5OTl5eXm5ubn5+fo6Ojp6enq6urr6+vs7Ozt7e3u7u7v7+/w8PDx8fHy8vLz8/P09PT19fX29vb39/f4+Pj5+fn6+vr7+/v8/Pz9/f3+/v7////isF19AAAACXBIWXMAABJ0AAASdAHeZh94AAAgAElEQVR4nO3dd2AU1doG8DeFJPQugoCodBFUxPupWFDkqngiVWkiV0GqggKKIiIKIiJFwQJSvajIlY6IdFFEpEjvoUNIgqAECC2Zb2Z3dnd2ymZ250yyu/M8f7CzZ07ezE7mx04fEhAEsRzK7wlAkGgIICEIhwASgnAIICEIhwASgnAIICEIhwASgnAIICEIh+Q9pHNnuOfsFf41z5y5lP23DVXPX7Ch6JlsW+bA5X9sKJqZbcccOHvVhqJnsnJbBP7OR0h/Z3DP6av8a2ZkXBL+sqFq5gUbimYItsyBy2dtKPqPcN6Gqqev2VBUXATOBO5wBpByDyABEiBxCCABEiBxCCABEiBxCCABEiBxCCABEiBxCCABEiBxCCABEiBxCCABEiBxCCABEiBxCCABEiBxCCABEiBxCCABEiBxCCABEiBxCCABEiBxCCABEiBxCCABEiBxCCABEiBxCCABEiBxCCABEiBxCCABEiBxCCABEiBxCCBFMKT9a75fsdtyVUDiEUCKVEgnP7w7lsTUGXjYSs31YwYNGPbN8UBdAMlEAClCIS2sQp6UnRJqwfRJtdwlCj233bgXIJkIIEUmpE8TSZE3Qqu39wFfiUJfGnYDJBMBpIiE9H0C+WVMKOV23uRX4z2jfoBkIoAUiZBSSvs7ooQ/gq924i5Vka8NOgKSiQBSJEJ6hdR5Kvhq76hrXH9UvyMgmQggRSCktOs0kOL3BVvsSElNkbf1ewKSiQBSBEJaoiFA9FmwxaZqa9TR7wlIJgJIEQhpnA6kV4It9qxOkZ26PQHJRAApAiG9rWOgXbDFGukU+VG3JyCZCCBFIKThOgY6B1usnk6Rb3R7ApKJAFIEQtLZvKFBwRbDNxLXAFIEQtodpzWwONhi2EbiGkCKQEgZ/6chcH1asMWw145rACkSIc3SGPgg6GKHS5hdPQQkEwGkSISU8YSKQN2TwVd7R/OtdkS/IyCZCCBFJKSDtfwIlP0zhGo4145nACkiIWXsVe69rrwhpHI4+5tjACkyIaXeSkVlAElUPYQVOyl7Fefs4XokawGkyIQ0hJ4+MrX9fbX+1WrcvjbGXyaBsyCmYk2ZUccdxt0AyUQAKSIhbSlczLfk7y5eJMCF4sY5UT1mccbvo97qO/TrY4H6AZKJAFJEQmpKIxXt71OrUMoNpA7SC+4ixCOAFImQvqM7lAdgT9WhucFX+7NQyT3SKyDxCCBFIKRjN8at9BuxOKZm8PsbHqfRrldA4hFAyhdIKRNfafNs/+9OBFfVA6kPdVWNCWF/g/dbDZB4BJDyAdLWdvJNgIr2PRRMVRnS7wnlUlRjgt/f4PtWAyQeAaS8hzS1kO/wTcVVQVR1Q0q/n7R3hAx6f0Nf7xVMgMQjgJTnkMb4nU9QKIjrH9yQxlEj7ahg9zf8kVj2gDwISDwCSHkNaYHq3o5lt5qu6oK0r0zSBp1xQe5vaEyfewYBiUcAKY8hnaqhPum6pemqLkjP0pu6I4Pa3zCZ/i/dMwxIPAJIeQzpM7Ujol/MVpUg/RR7i/6+vmD2NxyqEL/G+waQeASQ8hhSEy2k18xWFSGl1qE5BmOD2N/Qg172vQEkHgGkvIWUXlAL6W6zVUVI79DTRmPN72/4pcANimcqARKPAFLeQtqrdUQVzVY9fW1L4RK7DEf/EFMr1UyZ9P+jqYq3gMQjgJS3kP7QgVTEbNXT1/xPVlXH5P6GcfSI8i0g8Qgg5S2kFB1IVcxWPT2f6ge6W9Cu4kUDXFbkyf4yiRuU7wGJRwApj3c2aJ8BQQ+ZrXr8ZtXJquqY2t/wH9XeDUDiEUDKY0ittZCGmq3am7oF7mBmf8PyuJv8H70MSDwCSHkMSXtLusRtJouuSyivPllVnR9yPb8hrT59698CSDwCSHl9itB9akg9TdZMb0izcu2U6/6GD4mpWgCJRwApryGtV20lJa4zWXMcPXwt1065nd+wt1TBzaomQOIRQMrzs7/nFvHb903XrzHsqsy+MkmbcoeU2/6GNjRY3QRIPAJIeX890pqbfY4a7hlEJX4wU7IDvXnaBKTA+xsWxdTQbEMBEo8AUj5cIbsiJtGlKOZO6QyDT+IL6j/gyy/SyapmIAXc35B6a4xWGSDxCCDlA6Q2NPW3iUOHT5N3101PjBuTW0HXyaqmIAXa3zCEntE2AhKPAFLeQ9qeUPmUX8O8YjH9cyn4jkTAHCTj/Q07iumd+QBIPAJIeQ+pt+ZpRmvKU5eATwrbUrjEbrOQjPc3JOs+RgmQeASQ8hzSkVIlD6vbNt9CrQIdSX2CPsowDclof8P/qO4pnWZA4hFAynNI79Or2sa9d9EDxrfmmum+B51JSAb7G05UjdV92DIg8Qgg5TWkU1US9E7RPtyI7thj8CPHKrtPVjULKeMZvf0NA+g53c6AxCOAlNeQJlF73fYTzanKBv0f8ZysahrSLp39DX8WKrVXtzMg8Qgg5TWku2IMbnaS3oPKrdYb8UsB+c6qpiGJq4+t1U3/pk/0+wISjwBSHkNaSI0Nf2pQTPGF2tb0hp4rw81D0u5v+IruTtfvC0g8Akh5DOlxw7sAiRkXn6C9G/En3jurmoek2d9wrHK80c2RAYlHAClvIa2PrW3wxeDKjKS4UaqmfaWTNsqDQUBS72/oY3xNICDxCCDlLaROvlsF6+bHUp6THHZMGvbmx4tOZXSggZ6RwUDy39+wPuE6w2sCwxpSFv9cyrGhaNY14ZINVa9ctaFolpBtR9VsW2aAoD8HThau8E/gn9x0A3W7kJU17/9iXee1lm4XW+1v78hgFoGPqK308nem9G8j+q9hx9wXgXyEdP4c92Rm86957txVIdOGqlmXbCh6TrhmR9WrNvytzl0U9OfAmzQ0tx/dWY3YwScUVywN9o0LZhE4W5e+HfFAaYot1/TzL6iRcccruS2tmfkICat2NiTyV+1OXFckt7suiBtFDUjxACUxYz1jglm1k/YPxnoqxMX/ZtwvrFftAMmGRD6kUdTDxA8fUF2MnuDZkx0UpEN+dxnvpneWnTuAxCOAlIeQ0qvH/2nih4epb49SSX7+RDCQTj3iX6OXYU9A4hFAykNIM0w9C+loaTUket89JhhIr6lrfGXUE5B4BJDyENK99JOJn52ucUR3uscEAWm75qkXNxtdpgFIPAJIeQdpOTU087MdtZBi3KebBgGpv7bIVIOugMQjgJR3kFrQ12Z+9kGtAVrmGhMEpNraGjq3a3AFkHgEkPIM0p/x1QOdHeTNbTqQZrrGmId0MlZbo6ZBX0DiEUDKM0jdabSpn71fB9IS1xjzkLbq1Chh0BeQeASQ8gpSStEyx/W6atJGB4H7mlrzkLYAkmEAKbIhvU0DzP3sBK2B2u4xWLXjEUCKaEgnbyiof6W3JilFNAbedI8JYmfDrVpIRk9xBiQeAaQ8gvQpPW/2h19XEygr318oCEj9tJCw+9sdQIpoSLfFBjhv1D9H1V8nHgLBHJAtpHZ04wmDroDEI4CUN5C+p6bmf3pzeT8Cb3jagzlFSPOVNN2oJyDxCCDlDaSHydSzW+Rsv9MHIMF7FUVwJ6028ndkfNo5IPEIIOUJpDUxdwT186mjK7qX/wItNvpag7qM4mBjpaMXcRmFJ4AUwZDakvbuQIGTvnRI1079JxxQtgV3Yd+pQUU9jCpMCNAPkHgEkPIC0q7EysbfCOYTHKSMjL1D7i1GVLrJ2GOBegESjwBSXkDqQ8N5VA0WkpSjRjvrvAEkHgGkPIB0tFQJzYNcQkkokHIPIPEIIOUBpOHUh0tVQOIRQIpUSAYPcgk+gMQjgBSpkKZQWz5VAYlHAClSITUg3ae1BB9A4hFAilBIP9DDnKoCEo8AUoRCakrfc6oKSDwCSJEJaWNcLVO3ajARQOIRQIpMSM/Tp7yqAhKPAFJEQtpXqHyu5xaYDSDxCCBFJKQB9Da3qoDEI4AUiZBOXFfkQOCuQQSQeASQIhHSGONntwYfQOIRQIpASOk14jbm0jWIABKPAFIEQvqGmnOsGu6Qzqybv/EaIHEMIMmQGso3G+aT8IR0eN4K1+vp56SLccuN4koJkGxI5EFaHXMvz6rhCOnA3UR9pIELNeTr2ltcBSReASQ3pJY0g2fVMIS0QrovhAtSL++NVj4BJF4BJBekrQVuSeNZNfwgZbvuZSlBOl2A6Lr/bptcnuiGS4DEKYDkgtSDRnGtGn6QpMd0Nv7lgjgk3fp/mfh6UPyK+gKQOAWQJEiHipUOePueoBN+kNoQ3Z3tGnqEqK5r4BWi5wCJUwBJgvQOvca3avhBuptopmvgUhLRYNfQMqL7AYlTHA5p96dv9npj7IaKibv41g0/SGWItrkGVotrdj+7hvaLG0mAxCmOhrS8kfspXzHUhHPl8INUjCjFNfAOUZJ7H8MeogKAxCkOhpTWW3HH7WfMPezSbMIP0h1Eq10DDYgedjfNIyoFSJziXEjpLfyeAfF/XCWFH6RWRG9IrykxRB+6m14nuhWQOMW5kNSPJTJ63mRICT9II4hK7BWEHBEU7XW1pCQRPQtInOJYSGsLqJ/vNYtj9fCD9E9pcUXutS+eED9oQ+n95UU1xcG5gMQpjoWUrHZEt3OsHn6QpK8kOcvFd6tc/4/U4niyHSDZkPCHdCRRA4l+51c+DCFlvxnj/pgDpHdLpKEb/uTnCJBsKBoBkGZqHdEH/MqHISRBmH83UdLtPwgeSM+c0ekESKHFqZA+0oHUi1/5sIQkCJdSs+Wh7e/MO8STESA5FNJAHUjt+JUPU0h2BpBsSPhDGqUD6SV+5QGJRwAp/CHN0oE0kl/5sIV05qcpI4dOE4QcQOIbp0I6VkgLaTO/8uEJ6exrNV177hoLwvIGY08CEsc4FVJGa42jBhyrhyWkT0vJn1SCRFRgEiDxi2MhbUhQQ1rAsXo4Qhru/aRuSER9s7W9ACm0OBZSxmCVo+d5Fg9DSF9JH7LK0Nk7XJBSn5auIJkKSLziXEgZHf0cPXySZ+3wg3SlIlHi6MvSKAmSIGy6nqhSFiBxioMhpVVWOOqSyrV2+EGaKn7IL92j3JCEfUlE4wCJUxwMaSg1fMp9xl3cw0u5Vg5HSM2IWgt+kITeRB0AiVOcC2lT4WLbMg7/94M3hk/ey7OuK+EH6TaiJSpIy4nuBiROcSyk9AdpnPSqfBgzv4QfpMJEB1WQDhEVByROcSykD+lB15OXnQKpJNFaFaSdgMQvToX0Z5Gif7oGnAKpNtEYFaRviKoDEqc4FdLDnnsUOwXSi0T1LvlDakzUBpA4xaGQxlLDdPeQUyCtIKKufpAmiy3fAhKnOBPS9hIFN8iDToEkff9Q8n4vpOwxcURVLwMSpzgT0hM03DPoGEj7y4iSElp9uJLogaPrR98sHT9bws8RINlQNNwhfUoNvE9DcgwkYet1qrMLY7ie/g1INiS8Ie0pnbTe+8Y5kISMtn6OblrN0xEg2VA0zCE9Se/63jgIkiDsfqWarKhgoxkct48AyYmQJlH9U753joIk5q+f50yZvmDjFb6KAMl5kPaWTfhF8dZpkGwLINmQcIbUnAYq3wISIPGI4yB9TXX8LuEDJEDiEadBSikfv9KvwUGQLq7//ANV/DtMZ34ZK7VNUjTslfulTu7eqv2AH/zvwA9INiR8IbVRP3bZOZC+Laa5cZKqlx6k97SQlrR0v+15CpC8cRikmVT7hH+LYyAN0zJSQ1r0ii+dGXOdiNeDtfa2HXb1Wp3M2IApn3ZkrMt5QPLEWZAO3hC/TNXkFEhHkyQ4hWvU9Iv2e0tOzpush3SgKacFG+4/JrM9e2qN+Hp5CGMTAckTZ0F6lvqom5wC6WWR0SPbTd+peBF7yrUil87YdP8xMxmb4hq40JE1OwtIchwFaW5MNc0jl50C6UGihuZv+J3aSvazlbFl/qN6smbyr/mGscWAJMdJkI5Uif1B0+gUSOWJfjXtSHib9XCf+7CEsZ1+Y8SvqP7yYApjQwBJjpMgvUA9tI1OgVSaYi6YdvQzS5b30E1l7GzW9iXz1nhKb2DMc9J4dkv2AiDJcRCkxbGVj2hbnQLpcSpk2tGljsxzhOl91nJBB2lfd/Koc66GOYwt8vTrxpIvAZI7zoF09OZYvdvkOwXSGKITZiH9jz11VB7s5TuK9Ox2qWE6Y2s9/d5kLAOQ3HEOpO7UWa/ZKZAyS9IQdZtBMtu4j8WKyWnFWLtFGZePLuvEWKeLYssXjG3ydBzC2DHp9cg0MQfPc8+FbP41z5+/KlywoerlKzYUPS/YMgeuXbTww6vibkzTa78kXLZQ1Sh2LQK5zAHfVpD2gOw0ivvRHKRprFmaPHiasS7uL53z/Rj7Qnz9mLFtno7vMybdBEJYWV/MepNMkYjOpVtjfsrvabA/vucd6ZwiNIiKfXLRRJEzLdl4b8Fz5zzX/6Uks+ZZgvA5Y5s9Y8VvJNfJDmnLxBw/xz/XbKh57qqQaUPVrEs2FD0n2DMHzof+s6/S8/ojLgq2zIFsO4peEXKZA5mBIAlfF6Fy/SYs3bnHFz1I0+XvGXUGMLbVtRtvnadF3EZK847GNpINCbttpFUFyh/QH+OUbSShSpUq2rNWdbxkP8d66jkSxjO2yrXXznsUtgdjvpVJQLIh4QYp9TaaaTDKMZD0zlnVg7Sesdm6kGa4IImjp8kNOa3Zc77RgGRDwg1Sf2pjNAqQVHmPPeUrs3/Xbu/wSMa2iNtDjA2UG44yNhiQ5EQ1pEPff/rRV2szMn5OKLffqI9jIG3Ui9bRmafY2753g31X8+V0Zc2kfRXdWUv56om5jC0AJDlRDOnHJu4Hl988qC5NM+zlGEgms8xvzW4BY0PlwcWMfSS9zvB0uNaNNfMdjwUkG4qGA6TjbRRrMA8Z9wMk/3zImG9tTrjQgbFJruuSlrRmbV0XTZx7mrWW9urlTGC+3eSAFK2QDt2l3BRImGrYEZD8ktOBtVDe825dM8Y6jflqVHfG2HJ3k/id1WrCioX9Gev8DyB5EqWQTjXx36guqL4w1huHQTrz05SRQ6eJYAzGpzD2ml/D+o7yqXbPrPQ0zWnmbul+TNkRkGxIvkMard49VT3VoKeTIJ19rWaMNDMaC8LyBmNP6nWZzdhU/5aLi97q1Lxdv5nnfE0Hx3dp2a7v/Et+/QDJhuQ3pOPXa3b0jjXo6iBIn5aS50Vj1xPNC3B9GAUg2VA03yF9rT1ico9BV+dAGu6dF25IRH2ztb0AKbREJ6T/aCHFOf0Uoa+kuVBl6OwdLkipT8eKb6cCEq9EJ6SHtJBolX5Xp0C6UpEocbS0J1t+huwmcfW3UhYgcUp0QqqjA2mWflenQJoqzoIv3aPkhzHvSyIaB0icEp2Q7tOBtES/q1MgNSNqLfhBEnoTdQAkTolOSK11IG3X7+oUSLeJ/5WoIC0nuhuQOCU6IY3TOqph0NUpkAoTHVRBOkRUHJA4JToh7U3SQOpv0NUpkEoSrVVB2glI/BKdkDJeUjsqaXQdhVMg1SYao4L0DVF1QOKUKIWUcpMK0jijnk6B9CJRvUv+kBoTtQEkTolSSBlrS/o56m7Y0SmQVohzoasfpMliy7eAxCnRCiljfSWFo8HG/ZwCSfr+oeT9XkjZY+KIql7WdAOk0BK1kNLrU2GZUSODkxpccQyk/WXEWZHQ6sOVRA8cXT/6ZvFd3BJ+jgDJhqLhAOlzanriu97PPNl5+J8B+zkGkrD1OtV2YwzX078ByYbkP6RjlRJ+N9XROZCEjLZ+jm5azdMRINlQNAwgvRFg/4JfHARJEHa/Uk1WVLDRDI7bR4AUrZB2FS2x11xPR0ES89fPc6ZMX7DxitF4QAotUQqpHQ032dNpkGwLINmQ/Ia0Oq7qSZNdAQmQeCQ6IT1EX5vtCkiAxCNRCem/1NB0UWdBOjXTP+mAxCnRCOnkLbErTBd1BqRzn7d8U3pdrjqUxHEPOCDZkPyFNIyeNV/UCZAuDSxG1A2Q3AEkk5D2lyq8w3xRB0Da57qFhRdS8Vg3orKtxm7ieD8uQLIh+QrpRRoYRNHoh3TYffruyx5IwrVjXzUVX+uc0WAApJATfZA2JFQ4GkTRqId07U7p9NQWi7O9kKT8VpXongsCxwCSDclPSI/RhGCKRj0k6TboSXPlN15Iwl8NiJ4FJG6JOkjzqH56MEWjHdK1G0RH3n0KPkjC8XJEqwCJV6INUlq9mEVBFY12SD94No/UkIQJ4mYSIPFKtEH6mJoHVzTaIfUVv5BO6EK6In4lbQIkTokySIfLJW4Ormi0Q7qPqLlvaVdCEp4jegOQOCXKIL1KfYIsGu2QKhINMID0GdH9gMQp0QVpW8EyKUEWjXZIRYgUV5T/fvvtt3vfzCaqCkicEl2QWtGoYItGO6QCRPMMFv35RIUAiVOiCtKK2BpGj4o1TLRDKks0wWDRn0hUApA4Jaog/Z/RQ5ACJNoh1SB63WDR70dUG5A4JZogTaJHgy8a7ZBaGWupQ9QEkDgliiCduCn+l+CLRjukMUS0UXfJ3yKO+RCQOCWKIA2iziEUjXZIB2KJ7s3RW/KfECHtByROiR5Ie4oVN3kHLr9EOyRBumJiiM6CP0Fsf4CfI0CyoWi+QOpEQ0IpGvWQNsaLYt7XLPczxOZYjmcIAZINRfMD0tr4G0+EUjTqIQlvS5f1NTnit9SL/+2I6cXRESDZUDQ/ID1MU0MqGv2QsrtKaBLarfbcXfXq7y8WlJrYVUDilmiBNJMaBHUZkjfRD0n+TiJKuv+Ffu+9/kLjIu63TS/ydARINhTNe0inasUuC62oEyAJs6uSOoljdXflAVKIiRJII6hNiEUdAUm48mkFP0aFehzgywiQogLSobJJW0Is6gxIgpDzxzv/KuRWdEP7L0/zZgRIUQGpF70WalGnQHLlfMq6jYcz+SMCpOiAtCmh/JFQizoKkp0BJBuSx5AYjQ+5KCABEo9EA6TFMbelhVwUkACJR6IAUvqdtCD0ooAESDwSBZA+pSctFAUkQOKRyId0rGLCegtFAQmQeCRyIZ2Y9V73PsOXpr9OPawUBSRA4pFIhbSvZzH30cXrE0vss1IUkACJRyIU0uxSvtNdyv1hpSggARKPRCakyQnKE8dK/WyhKCABEo9EJKRlBf3PZK4UyjXmcgAJkHgkEiGl3aa+JqBj6EUBCZB4JBIhTdZcXFNgQ8hFHQNpz56jqpZsbRMghZpIhNRUA4kGhVzUMZCIGqtarsbSbYDEKREIKb2wFlLDkIs6GJJQgooBEqdEIKS9WkdUOeSiDoZ0JIYKABKnRCCk33UgFQ25qAMgvfaYFKLSj/nloRLif0CAxCkRCGmfDqQbQy7qAEgP6cwwOa0BiVMiEFJ6Ue0C8UDIRR0NqeR2QOKUCISUkaxdIt4JuagDII3sLIWoQmf/vDjiiNYDIIWWSIQ0XeMoIchHmSviAEhyi3avHdcAkg2xF1L6nWpIoTzPRQ4gARKPRCKkjNWF/B3dtD/0oo6B9NFHswBJCiB5IZ1q4OeozFoLRR0Dye4Akg2xedWuA9W8zueo7p9WigISIPFIJELqQ7X2H+pbWl6tG3PKUlEHQbq4/vMPVAEkXolASCOo8g7x5dSiD/q+NmaN1aLOgfRtMe1hA0DilciDNDG21DqORR0DaZjeEVlA4pWIgzQ7ochynkWdAulokgSncI2afgEkXok0SEsLJ8ziWtQpkF4WGT2ynfPDxZS5cJ57LmTzr3n+/FU7JvX8pcs2FD0v2DIHrl08v61s7Fd8i14SbJkDNi0CFwN3uGAM6UGihjYyEiFlcs/5bP41MzOvCudtqHrpsg1FM4VrdlS9dmFvZRrBuWiWYMscsGkRyGVpPW8MqTzRr3Y6wqqdDUXtWrU7cmvoDxQzilNW7UpTzAV1GyDxSyRB+vv/6DnuRZ0C6XEqZKsjQLKhqD2QTj1FT1g7+KoXp0AaQ3QCkKQ4HVJ6W7r/BP+yToGUWZKGAJIUp0PqSbcd5l/VMZCEaRT3IyBlOB7S21TlqOFTzS3EMZCEQVTsk4uA5HBIn8WUWW/4MGYrcQ4k4esiVK7fhKU79/gCSLwSIZC+iS+60vip5lbiGEhVqlTRnrUKSLwSGZCWFEpcYPxUc0txDCS9c1YBiVsiAtIvJeOmZgASIHGJcyFtrRgzRnoFJCuQNuoFkHglAiDtrUYDXQOAhEvNOcRRkHaPalq7Qq1Hh23JyDjagJ53NwISIHGIgyAd6eN5tmXCCzsfphZp7mZA4gxp1qRlgMQp4QhpSx3F5nBhesBzYhAgcYZ0I9UCJE4JQ0h7KvvtWIpf7RkBSFwh5cwlSgIkTgk/SOkNVbtoa+AbSY41SNmfNalZRZGy4rwtA0icEn6QpmgOdrwrjwEkK5DOPalzHIkBEqeEH6R6mr92WfkqJECyAqm5jqPa+wCJU8IO0p86f++F7lGAZAHSvliiIs+OGFSN6N6PPvpoSPNYivk+m58jQLKhqBVIE3UgDXKPAiQLkF4hKrpJfE0rSFVcDUtLUF2ed3EAJBtiAdJ7OpC6uEcBkgVIjxL1dg20JDrkGphK9Aog8UrYQXpLB1JH9yhAsgCpNpH76Ov7RPPcTY9QwkFA4pSwgzROB5J8Dy5AsgCpGNFu18AcohGCZ4jjbRwAyYZYgLRUB9JU9yhAsgApiSjFNbCF6AV3UwrRvwCJU8IOUnp5jaOEQ+5RgGQB0g1EP7sGMonquZtyClM5QOKUsIOUMUADqb08BpAsQLqfaJh7qBLF/uMauBKHU4S4JfwgHSmrclR4uzwGkCxAeouoXIZr6AmiSa6BjYSTVrkl/CD9VEJ/CwmQLEHaHkNUedZfgmu33Y1nxdecJ4maARKnhB2kaQVjmhVSbiCN9I4CJCunCLWQ5mZZceBAHNEt0zfMaSy+HwtInBJukIbFJk7KWHmb19HNc3zjAMkKpNO3kHy290veuVv1MiBxSmNJuxIAACAASURBVHhBOtWZSv0gvqZNbJIo/p3jHxh9UjEWkCxdRnGmd7wb0j/3yI5KrOLnCJBsKBoqpCP/phqb5eETW5ZvOu4/GpAsXth3ZulM1+vFftJ2aIEnjnF0BEg2FA0R0o56dP+BAOMBidsVshe3/LCF42odIIUTpDUVqU3A57YAEu4ixCHRDun7YjH90wP2ACRukHheiARIUsIG0scFEj7NpQsgWYR04uMuXdxDy4ve/c4hQOKYMIGU3p9KzMutEyBZgnS+ayLRXTIk6YyR8TmAxC3hAelES6q8NtdegGTp5if3SXu8FZCIHuH53DFAsiFBQtp3D925K/dugGQFUg+JTsW33G/SBjWQ3j4PSLwSDpA2VKWmR030AyQLkHbGEl2/UNGw/CaimB2AxCn5BOnI7z/v9OyiW1yauqSZqQpI1s7+LnzYr+V4MaJXAYlT8gPS/jfrxYjrFaXaLJLeTUmKG26uKiBZgNRAe135CKL6gMQp+QDpo5LekyYf25UxLLbQDJNVAckCpOuJVqmafsUti/klzyGldVRea1TxKSq33GxVQLIAKcFzFy5fjhEVACROyXNIL5F/qv1puiogWbtnw3pV0x9E5QGJU/Ia0iz1DRnuN18VkKzds+EtVdMgonsAiVPyGFLarWpINMt0VUCyAOkjoqL7/VoOFyMaCkickseQFmkc0ROmqwKSBUipRcQt0rWKhj9vJko8BEickseQemkhFTpu0FcTQLJyZsNYaWbf8/VZ15sLPz0mvR3IzxEg2VDUGFJjLST61WxVQLJ00mpP9+wuXOOhO0q7B5tdBSReyWNIt+lAmm22KiBZu4xiVCG/+R7zEtdrZAHJhhhCuksH0iKzVQHJ4vVIB3v57hkY3+I3nowAKW8hNdWBtMlsVUCyfIVs1uoRnZ566LG2r89J58sIkIKEtHfsk/Uq13lsxPaAvQwhDdE6uj7w9eWKABLu2cAh4QDpaL/C8uKf0HV/gH6GkP7QQvqP6d8OSIDEIWEAaWtdBYCbAlzSanxmQ7LaUcJG078ekACJQ/If0t7KfgRKGW/cGEPaUEQFqbf53w9IoUAaLyZdftUEkHglCEjpDVUGahnehS7ASauv+NdodMr8tAJSKJCkubxRftUEkHglCEhTNH+G9426GkP6PjEhSVGBHQliWgEJkDgk3yHV1fwZyhldIW4IaXHhhG/W/dv786NN77GTAkihQKooZqv8qgkg8Yp5SJt1/kNbbNDXCNLqEnGTxJdVfe+7pVy91pPM3PFEEUDCzgYOyW9IE3QgDTboawBpXdmYsaFMpBxAAiQOyW9IOsdSqatBX31IG66nESFNpBxAAiQOyW9Ib+pA6mTQVxfS1ko0KLSJlANIoUA6e/q06/X06X8ASUp+Q/pYB9LrBn31IO2uSq+EOJFyACkUSLdQCXcLNQYkKfkN6UcdSFMN+upASqlLnUOdSDmAFAqkgpSUA0iK5DektOs0jhIPG/TVQjr6L3rG1O1UAwSQQoF0A9EyQFIkvyFl9NNAam/UVQPpRCN6MjXUafQEkEKB9AhRwae7detGdEM3TQCJV4KAdKisGlLDHQZd1ZBONqFGAZ9qaSqAFAqkuTqr5N4AEq8Ec9LqggT/v0J1KvNf/Z4qSKea093BnAtkEEAKafe3+pacgJTfkDImxSr+BgkjTg1KoKd1N5P8IaU/S/UPWZpKdwAptONIP7/GGjduTFSqsSaAxCtBQRpOvttn3Pi92LCmNlXWu+WCP6QeVGufpYmUA0hWDshiZ4Oc/If0a1LxzZ81ktbv4u4b6d7kOf5ybPzL2q0fP0iv0M07LU6lO4AESByS75BO1KGJ4svxjT/9rjjb9PsKdMc6dVclpKF0w2YrU+gLIFmBNGDAJECSku+QelAbveaU1pQ0SHWISAFpBJXROAsxgIRz7TgkvyEtiK18UH/M5JL00Da/Fh+kz2JL/RLipGkCSIDEIfkM6UDF2PlG47Y0pOKfKxu8kKbHF10W6qRpAki2Q5rEfNkrt6VO7t6q/YAfFLc31rYAkmlIzaif8cj0UQUpWb4/1+FxTW8qVKJGuxmnxA2ohIILLE+iN4AUCqR4MRvlV03UkN7TQlrS0v225ylPJ20LIJmG9AndfjLQ+F/rUsW54mv6SO/5D7X+t7hQwkwO0+gJINl+z4YerPUrnrifgb46mbEBUz7tyFiX84JBCyCZhrSxaKHfA/c42T8upsuJY82Vf6akuMkcJtEbQLIbUk4LNty/JbM9e2qN+Hp5CGMTDVoAKcMspNQG9HGunRbfRDVUd+xilqdPGUAKBdI0MdKlfd/rRQUpnbHp/i0zGZviGrjQkTU7q98CSBlmIfWnpiZ6Heyo+R9vrrWp8w8g2b2zYStjy/xberJmctFvGFus3wJIGSYhLS1w/V4zxXYlqiHVsXoNkjKAZDekJYzt9GsQv6L6y4MpjA3RbQEkKWYgHbk5xtw+g/7alXDzj1rOPYBkN6SpjJ3N2r5k3hpPoQ2Mec6HyG7JXtBtASQpZiC1oe7mimkfWW54e5RQAkh2Q3qftVzQQdqznTzqnKthDmOLPCO7seRLei2AJMUEpClU29xFeSditJDqWp1ARQDJGqSL6z//QBVVj16+o0jPbpcapjPmfRL6m4xl6LUAkpTcIW0tmbjGXC29e7FeZ3UCFQEkS5C+LZbb7u+cVoy1W5Rx+eiyTox1uii2fMHYJs/YIYwd02sRhLRlYo6f459rNtQ8d1XItKFq1qVcOvz9EH1ostZWHUilrU6gIlfPcyzmyUUhtzkQUrLtKHpFyGUOZAaANEznr6PqdZqxLu6vmPP9GPtCfP2YsW2ese8ztl+vRRBW1hezXksXUWQkPZpjsusFnb9UXVsnDvFPtndIA+mo6yEghWvU9Ivqx8+d8zzoPCWZNc8ShM8Z2+wZK37/HNZrEYQTs8UcyeSe89n8a2ZmXhXO21D10uXA49cnld5vutgtWkhtLU6fMlcvcCzmSZaQyxwIKXYtArnMAd8pOxpIL4t/jEe2m/0/URAGMLbVtRtvnadF3CJK02uRg22kQDlei6aYL9ZDC8noFpKhBNtIFraRHiRqaJ6RIIxnbJVrH533mGsPxi7otQCSlFwgdaaOQRTblKB2VNn6Tbh8ASQLkMoT/RqEI2GGC9J6xqbJDTmt2XOCXgsgSQkMaVZMlaBu//OiGtIka1PnH0CyAKk0xVxQt6mzf9du7/BIxrYIQhpjA+WGo4wNFvRaAElKQEh7y8UvCara8bv8Hb1gbeJUASQLkB6nQrk5Egb7rubL6cqaSfu/u7OW8obXXMYWCLotgJSRC6THaGCQ5fY0UDpqF/ASpqADSBYgjSE6kRukBYwNlQcXM/aR9Cqu4c12NVzrxppl6LcAUkZgSCPp7iCeOe7OiZ7e7aRSo6xMmE4AyQKkzJI0RN2mzoUOjE2SdoDnLGnN2roukTj3NGstHSvKmcDYeEG/BZAyAkL6o3CxUO6jtblPrViihP97NyX0qdIPIFk5s2Eaxf2Ym6R1zRjrNOarUd0ZY8vdTcsYazVhxcL+jHX+x6gFkAJBSr2TPjcal0tSjx8P8ScDBpAsnSI0iIp9cjEXSes7yqfaPbPS0zSnmbul+zHjFkAKAKm3hetbjZ5qbi2AZO2k1a+LULl+E5bu3OOLps/FRW91at6u38xzvqaD47u0bNd3/qVALYBkCGlRXIX9IVcFpPCDVKVKFe1Zq3rfSiEGkHSTUinWwmXigBR+kHROhAQkflFBSl35zdRFu8SBltTHQlVAAiQOiVhIa58pIc3d2AaffEp1rZzcA0jhB2mjXgCJVxSQUrsW8P5PFZf0m5WqgBR+kOwOIMk5+oDyOz9pjpWqgARIHBKRkNKT/deei1n5SgIkQOKQiIQ0Tr0dent66FUBCZA4JBIhnaio2aPzZehVASn8IP2qyaZdh7IAiVM8kL7V7hptFHpVQAo/SLq7v2OrNJ18RtMVkIKPB9J/tDM54XDIVQEpQiC5tobHZ2s6A1Kw8UB6QGcOm7ydnU4AKfwg1axZzYhSZ0CyHA+kWjrzd3bIVQEp/CAJhxoSFUx+d9LijcumDXsylui+9et++LhlIfEvPQGQrMYDqYEOpJ9CrgpI4Qdpa3GKe8P3PKPD/eOovtQ/tSlRBQ47HQDJlWQdSNtDrgpIYQfp9A0UM9evZX4MPSI9T/lKPaJcL/kDpFzigTRC6+jm0KsCUthBeo/oBVVTZ6IZ0utkor6AZDEeSNsKaCBZOP0bkMIOUl2iVaqmn4kekl5/J2oGSBbjPbOhk9pRMVMP6dMPIIUdpOJEh1VNx4iul17PEt0LSBbjhbTrehWkERaqAlLYQSpG9KeqaRtRQen1DFFjQLIY39nfPxX0cxTMrYo1AaSwg3Qb0XBV00ii6tLrn0TdAMliFNcjLS+rcPSqpacoA1LYQXqdqNQRv5bjZYhekgZeJRoFSBajvEL2LoqXGdVfaK0qIIUdpAOJRJVXKRp+uYkoXnq+5dyCFLtT3R2QgowC0mRqvP/zjk0atnpjtdWqgBR2kISPpP8h64zbejpHEP7a9lk96e1AQVjSMIaog3VHgOQZOnFT3C+8qgJS+EGSVu5cSSifKA91Ek29Jb6WTQEkq/FBGszx2RGAFIaQhK/9d8wWGic9eEyEdJt6vzggBR8vpD3Fils4cKQKIIUjJOGfsbd7GZV+7airbUz7r87r9QWk4OKF9B96h19VQApLSGJSF4x4vVufwRO3XuOiB5A88UBaG38jx6dUAlK4QrIvgOTKw1wfmwxIgMQhEQhpNjWwcNMgTQAJkDgk8iCdqhWzlGdVQAIkDok8SB/SM1yrAlL4QIoXs1F+1QSQeMUF6dB1SX9yrQpI4QNJ2te9UcDTKLyxEdJL1J9vVUACJA6JNEibEq8/wrcqIIUPpGliTouv3+sFkHhFgpRM4zhXBaTwgZRXAaTFMXUsXXykE0ACJA6JLEjpd1q4E6RBAAmQOCSyIH1GTblXBSRA4pCIgnS6UsJ67lUBKYwhXZ7U46GG7b78C5B4JnMQdedfFZDCC9K21x+8/aA8PP9m957vxMEcH44ESClFS+zjXxWQwglSZhcJzh73m7mx3oNI9Tk9GQmQxDxLw22oCkhhBGnfTeSDtLWw601J1793XQUkTlkdV+0k/6qAFEaQLrmuio2vd8z1roU4XPDTs8KZj6XbGI4BJE55iP+ubymAFD6QXhHBlF8kbw8djSOKWeEa3HCD2M7xOllHQ/qKHryQe6/gA0hhA+liEaIGaZ53Q8j3fL75xOVxLoCUkXHylrh1gBTdkL4TuazzNt8jvtstD2eL2079AYlHhtKzmYAU3ZDaEj3ubT0rrtnd4X0nrvQ9Akgcsr9U4R2AFOWQxO+gsd7W2eIX0lDvu1FENwISh7xIAzMAKcohVSGa723tKkL6w/vuG6ISgGQ9GxIqHAWkaIeURLTF2ypuFZXK9r5bTMTxWnPnQnqMJmYAUrRDKkS02TO8X/xCau1b9L8lKgNIljOX6qcDUtRDuoHIex3seBHSRN+iP5aoNiBZTVq9mEUZgBT1kO4jes8z/KAI6ZBv0X+GqBEgWc1YaiG9AFKUQ+pLVPmye/B4LFFV35J/sQzRYECymMPlEjdLr4AU5ZCWi99CH7sHO5P8rEt3Rohv1wOSxbxKfVyvgBTlkHKqE8WOkIamxpDyyebLEoluyRa4xZmQthYsc9A1AEhRDklYIJ38XemZntLZQXS3p/XcG9Jj++YL/OJMSK1otHsAkKIdkvCy736QSe6jsfv7NXc9wT6ZoyNnQloRWyPVPQRIUQ9JeDNOdlTgW3fDPPfbJ7hea+5ISP+iWfIQIEU/JGFrx1Kim2JtDwgKSEXfuMLTkSMhTaImnkFAcgAkQcg+tT3N924eFfrXu1xv2OBMSCeqxP/iGQYkR0Dyz+ULnBE5DFL6snED3/1iU8Yg6uxtAyQHQrIljoF0dGAF9zZmjYLF93pbAQmQACmYLKng2wtaeae3GZAACZCCyDdJyudL3bDR0w5IgARI5rOmiP+T2moekkcAEiABkumk36l+5mEPeQwgARIgmc40tSNK3OYeA0iABEimwzSQaJh7DCABEiCZTVoRLaRG7lGABEiAZDa7tY6oqnsUIAESIJnNrzqQSrpHARIgAZLZ7NSBVMU9CpAACZDM5lSiFtL97lGAFO2QVs5b6HqdN28tIEmxstfuUS2kt91jACnaIVWiku4WagxIUqxA+lTjqMAG9xhAinZIiRR3BZAUsQLpVA01pGflMYAU7ZCqEn2UF5AuZHLP+Wz+NTMzrwrnQ//hpUn+jiodlEdcusxl4lQRrtlR9ZoNf6vMLMGWOWDTIpDLHDivA6ml+OeuUKdOHaLCdTThCOki/2Rl21D04jUhy8JPf+HnqPjvnvYrVzhMmiaCPXPgkg1FLwu2zIEcO4rmvgjoQNoaR8bhCMkRq3Zi/q2YfVXXeZuxahftq3bCzNKApIw1SItiK3WQV+/KvHPc1w5IUQ9JOLN41NChQ4luGaoJIAWZI1ViF2Qc/er1F158e+Ep5QhAin5Icgv22rljCdKz1Et/BCABEiCZzvcx1Y7pjwEkp0BasmQDIEmxACnlhvifDEYBklMg2R0nQHqa+hqNAiSHQTq8Ys60Wcv25ABS8JlBt54wGgdIDoJ0bWELz67wkk8tvAZIwWXvdQWWG44EJOdAWlvL7xhS7T8AKag8RW8ajwQkx0CaHKM6Ghv3DSAFkS+p7knjsYDkFEiLYiU8j3z03W+HNs4e9YSkKn4FIJnOnjIJvwQYDUgOgZRVRZTTfpv3/Y724vsaHB+RFO2QnqQhgUYDkkMg/Vd0M9yvZZjYMhOQTGYcNTgVaDwgOQRSa6KHVE0PEbUFJHPZVqLg+oAdAMkhkKoSfadqmiWu2wGSqaQ/QiMC9wAkh0AqSrRH1bSPqDggmcpIuj89cA9AcgikJKIDqqbDRIUAyUw2Fym6OZcugOQQSFWIFqualhDdDEgmknYffZxbH0ByCKTHiNqomtoRPQFIJjKUHsplxQ6QHAPpMyKa69cyX2z5HJByz4ZCxbfm2gmQHALpTGmiAiMyve/Pf1SAqMwZdTdA0iTtX/RF7r0AySGQhEnSGUIlXlt54JJwOWXVgJLS2y/5OYpeSG/R4yZ6AZJTIAkD5XNVY0p4zl4dwNFR1EJam1hql4lugOQYSMK0kn4nfxefzNNRtEJKvYOmmOkHSM6BJGR8UM3L6JZhaVwdRSuk/tTSVD9AchAkMSfnjXmn7+Axc47zVRS1kFYVKLfPVEdAchYk+xKVkE7UounmegISIAGSYXpTO5M9AQmQAMkoP8aV32+yKyABEiAZ5HiNmFlmqwISIAGSQbrS86arAhIgAZJ+foitfMh0VUACJEBS5vi3nZvUf7TTtCNHboqdb74qIAESIPmS+uF18hHrEndR9yCqAhIgAZI3+x5QnEOVtDGIqoAESIDkySH/mzpX2m2+KiABEiB58oTqps7/l2a6KiABEiDJ+Y7UGW+6KiA5BFKXOnVu5/9QpOiC1EADqYrpqoDkEEgFxcXiFCBlBIC0PVYDiVaZrQpIDoFUW1wq1gNSRgBIE7WOaJjZqoDkEEjDxaViECBlBID0rg6knmarApJDIGXdSZS4F5ACQHpdB9J/zFYFJIdAEs7cRVSb55PFog7SKB1Ir5utCkhOgSRc7i9uTP/7y9/3HzvuDSApMk8H0pdmqwKSUyDVqVOniGY5ASRFTpbQzJ8CB8xWBSSnQNL57xaQ/POiZv60MF0VkAAJkOTsKaqaPQnmT1sFJKdAWq0XQPLLNBWk0earApJTINmdaICUMdzPUf8gqgISIAGSNxuTYryMSk0MpiogARIgeZJ2D419u7aLUdX+KUFVBSRnQTrz05SRQ6cJAv8TwaMB0mB6TPx3+6JpC3N7ZKwmgOQgSGdfq+lac2ksCMsbjD0JSKr8llRqZ6hVAck5kD4tJa/9S5CICkwCJL+k3kmTQq4KSI6B5Nsj5YZE1DcbkBQZQM1DrwpIToH0lUSnytDZO1yQUp+WrmKbCki+rE4oHcTNTtQBJIdAulKRKHH0ZWmUBEkQNl1PVCkLkDw5WY+mWqgKSA6BNJU8z16WIQn7kojGAZInfaiNlaqA5BBIzYhaC36QhN5EHQBJzooCpp/gohtAcgik24iWqCAtJ7obkNw5UTNmpqWqgOQQSIWJDqogHSIqDkjudKdO1qoCkkMglSRaq4K0E5A8WRxXyfwTXHQDSA6BVJtojArSN0TVAUnK0Ztj51msCkgOgfQiUb1L/pAaE7UBJCnPU1erVQHJIZBWEFFXP0iTxZZvAUnMnJgqR6xWBSSHQJK+fyh5vxdS9pg4oqqXASkj42Cl2EWWqwKSUyDtLyPdhKDVhyuJHji6fvTN4ru4JfwcRTCkdtTbelVAcgokYet15J8Yrqd/Ryyk/8VUP269KiA5BpKQ0dbP0U08b30SuZAOVIhfyqEqIDkHkiDsfqWarKhgoxkct48iGVKroO5xYhhAchIkMX/9PGfK9AUbr/BVFLmQvqI6J3hUBSSHQbItkQlpb9mENVyqAhIgORlSMr3FpyogOQlS9qqhzRpUq3TbA10nHwEkMV/QXaf4VAUk50C6OupG3z672MfXANKOkom/cqoKSI6BlFJPdRjphUtOh9SUhvKqCkhOgXS4knu/d83HXkiuU8g13ETnlg1bPu72TNt+Y3Z63k9ivnienJk6uXur9gN+uBrpkD6muzmt2AGScyA9ItF5YKF8i9VljaS3r6g7nX/fg2b0OXfLe1pIS1q63/Y8FdmQtpUo9Ae3qoDkEEg/im5Kr1Q0rCwlbijt8O907S3GkgdOmzhAZDLYTa4Ha/2KJ4ddLauTGRsw5dOOjHU5H8mQ0h+mkfyqApJDIHUiiv/Dr+X3eKLX/TstYaz5Bmlgi8hkkTSQ04IN9++T2Z49Je2nuDyEsYmRDGkEPZDOryogOQRSNaIuqqYuRLf7t/RiTL4efXMyayfdhzWdsen+fWYyNsU1cKEja3Y2ciFtKlLsT45VAckhkIp67yLkzVJxZc+v4ThjL3mGX2Vsn/iylbFl/j/VkzWTf803jC2OQEj//CNBSruXxvGsCkgOgVSYaLeqaR9RQb+G9Yx94Rkez9jvgmtlb6dfH/Erqr88mMLYkAiDdGJ8U3HTMPG2fhvfpSZcKwOSQyBVJ1qnavqDqJpfw8qWLed6hkczJsGbytjZrO1L5q3xlN7AmOcypuyW7IXIgjSjiucgWoH4Ytu4lgYkh0DqSDRM1TScqKW6myfX/sOSpR3g77OWCzpI+7qTR7n3h8+R90JI6caSfcd0IwDSAOXx6Bu3c60NSA6BtJLoujS/lrRyRAuMIP2PuffW9fIdRXp2u9Qw3bs/QhDeZCwjgiC9439mR+3DPIsDkkMgCR2I7kpVvE+9i+hhowckLUpmLQ6JrzmtGGu3KOPy0WWdGOt0UWz5grFNnl5DGDsmvf61XszJv7nnn2s8qy1RXWlPbXhWv3iJZzVPBK5zwJOrmTYUvSBk2VD1n2wbiv59RTgXuMO5AJAyGxMV6Z0iv0vpXYSoaqqml9uYKCT5V2noNGNd3F865/u590R8zNg2T7/3GZNuSySsrC9mvYHJsMm9aki0Mb8nCQnT+L5gfJA+8GRoWemk7zpNe7zTo2kd6TFjt67bo1cka0YL8WvoN3fBc+c8V6SnJLPmWYLwOWObPT1Fb66THVI+EXPgIv9kc6z1s8YRtedY/vIVjsW8EXjOAW+uXbKh6GXBljmQY0fRa0JWLj10IGmXH0V0HP32H3F7aNgZ7YgBjG117cbz7v0Tt5F8m13hvo3UR/vpS6XyK49tpGjfRgoO0oURIqM+2zTtguvI0irXXjvvUdgejF2IGEiP6Hz8jfzKA1K0Q6oZKGos53uJa3UrcvQcCTNckNYzNk1uyGnNnvONDndIdXQgWb/BqjeAFO2QgsllcfVtsHKtbv8u3+kQIxnbIghpjA2UG46KnSMH0m06kH7gVx6QAMkXcQPoo2vKhsG+q/lyurJm0jZYd9ZSvnpiLmOK41DhDqmxDqRN/MoDEiB5c6kt6+J/28gFjA2VBxeLyKRXcQ1vtqvhWjfWzHc8Nuwh9dM6KsPt+lhAAiRFljE2x7/lQgfGJkm2cpa0Zm1dF02ce5q1lo4e5UxgbLyia7hDWqGF1J5jeUByDqTTP0/4QBX/Dp8x1utNX6TDteuaMdZpzFejujPGlnu5tZqwYmF/xjr/E0GQMhqqHRXgc2tIdwDJKZCy3iqg/T/Zv8tg5hfXw5vXd5TfPeO9Tn1OM3dL92PKHw57SMvUn/85ntUBySmQntDZ2Fb16qoDSbi46K1Ozdv1m+k7/Ug4OL5Ly3Z95/vfzSvsIS0v7P/Z7+TwMBdfAMkhkOa5Fp74ChX9ouEWesId0vyiMX7/l9yzl2NxQHIMpGbistNq61V1s2MgfZ0U90nG/Fs9jAr15/IMCl8AySGQanN9hHnEQfq8QMJU8SXtq2cqJ1LZe4bs5FZZDiA5BFISke653s6A9EFs4e89w5euqJ9qziOA5BBIpSnR6Cq+6Ic0iEou8b7RPIyZSwDJIZDuIDqlbnMIpPRuVHGd7y0gAZIFSIOJvnUmpFNtqdoWxXtAAiQLkE6UoAq2fiWFK6QTTaneHmUDIAGSlQOy38bQ7RucB+nwg3TvQb8WQAIkS+faTY+jWDbku3nKRD2k3bdRk2P+TYAESFYg5YxOyO0UoeiD9GdVanVS1QZIgGQFUq/cz7WLOki/3UAvpKkbAQmQLD5ozGmQlpeml7WtgARIFs+1u/N/e86d90tUQ5pfNOY9nWZAAiQLkCoRPaJ/d6CogrRrzVrPProZ0mmqOgEkQLJ2rt0WOx2FA6QVz14nra/W6LMjw3Wa6jTdXoAEc3gFpwAAHa5JREFUSBYg1aUCNl5DEQ6QUlp5N/0KDUz/ILbwbP1+gARIFiD1oti/oxrS5urKvSi1qMxKg46ABEgWIG0n+iGaIaXU8N8fWWidUU9AAiRrz0e6NTOKIT2t3rNv+KhlQAIkK5DO16UmR6MW0irNIbIKxwy6AhIgWTrX7mwTKthr+vo9ykQNpP9oDzZPMugKSIBkBVKVKhWj+MyG8trP1sqgKyABkhVIUX2K0AGdz1bHoC8gARIgGWStzmcra9AXkADJCqSNeokWSBt1IFU06AtIgGT385EiFtKxGC2kuwz6AhIgAZJR6moh9TToCkiABEhGeUMLyegBsYAESIBklL3F1I7uMeoKSIBkBdJyvUQNpIxhKkeJy416AhIgYfe3cVr7fzLda/pcASRAAiTjnFSetppoeMoqIAESb0j1GzeOIkgZ/6bSnk/W0OhaJCmABEiWzv72JfPgss/vIaq+j6OjfIf0Gd17/OsODSrcfE/vJQE7AhIg8dxr92NlKn0oeiDtLZOw1lxPQAIkrru//yxEd3O8jUM+Q2pNA032BCRA4nsc6RMijjf/zl9I31Ft9a2JjQJIgMQX0mmiFlEC6eiNsYvN9gUkQOJ8ZkMxqhElkLpSD9NVAQmQ+EK6GEsFowPS0rhKh01XBSRA4gvpV6JKUQEp9TaaZb4qIAESV0jZ9xE9EhWQ3qA2QVQFJEDiCOnyH/cTUf9ogPR7Yqk9uffyBpAAyQqk6/1SroB0Kk3h1CiAlH4/TQimKiABEu+TVsfwc5R/kEbSI0FVBSRA4gupzAKOjvIN0o7ihTYFVRWQAMkKpC9U+XpXNk9H+QapKQ0PriogARIuNddkCt15KriqgARIgKROSvmENUFWBSRAAiR1OlC/YKsCEiABkirzY6oeD7YqIAESIPnnRLVYo7vXGQeQACkUSHUCJcIh9abng68KSIAUCiTduwdFx12Efi5Q7kDwVQEJkABJmVO30/QQqgISIIUCqZl+qkU+pHfoqVCqAhIgcdvZkPVGvOSIRTKkzYWLbQ+lKiABEi9Iq13fR+W+4+go7yE1oo9DqgpIgMQH0t8vuh7L1elMgD7hD2kc3ZceUlVAAiQukOa4ngB+M88nUeQDpL2lk/4IrSogARIHSKktJUZx/S5wdpTXkJrT2yFWBSRAsg5pUgnJ0e08n8KcL5Bm0q1mbwipDiABklVI+xtJjJKGc7xTcf5AOlI53vBBYrkFkADJGqSrIwpKjh7i+hCK/IHUmV4KuSogAZIlSJvvkBiVmJhji6M8hbQkrtKRkKsCEiBZgHTxddch2BYn7WGUp5BO1Iz5X+hVAQmQQoe0sqrEqPxsuxjlKaT+1M5CVUACpFAhne0sMYrp8rd9jvIA0rax/Z57ddTmjHWJZfdZqApIgBQipNnXS46qrbaRkf2QFt8X4z5j/a5aNMlKVUACpJAgnWguLX7xA7JsdWQzpNQuios/KpywUhWQACkUSBOLS8te/S32MrIZ0qnH/C6jus+KJEACpJAv7CvTp59uIgVSD9UFiW0tVAUkQHLqFbIrNBM+J/SqgARIToXURDPhDUKvCkiAFAqkpECJDEgHCmj/CwjuxvnKABIgOfS+dlN1vktHh1wVkADJoZCG6kB6JeSqgARI4Qwp82/u+eea+/V1HUgvhFz1inCO0/QpczHLhqJ/C9fsqHrVhr/V3xcEO+bAP9k2FM19ETiXj5Cy+OdSjvt1lA6kN0Ouek24xGn6lLly1YaiWUK2HVWzbZkBgi1zIMeOorkvAvkIycZVuxk6kMaFXBWrdli1C+dVOxshHU7UOIrdEXJVQAIkh0LKeEoD6YHQqwISIDkV0jrNgaSfQq8KSIDkVEgZA1WOelqoCkiA5FhI6e38HD0Z5POX/QJIgORYSBnpb8X5HPVJs1IVkADJuZAydiQluffdJTz+s7WqgARIDobUmT448s3wfsO+SrFaFZAAybmQtifdYOn6ckUACZCcC6kTjeJVFZAAybGQNidU4vWFBEiA5FxI7UN8Op9eAAmQnAppQ4GbUrlVBSRAciqkZ+hTflUBCZAcCml9/C38vpAACZCcCqkFTeRYFZAAyZmQfomtaemcIFUACZCcCYnRVJ5VAQmQHAlpTWwtnl9IgARIzoT0OP2Xa1VAAiQnQloeUy+da1VAAiQnQmpMM/lWBSRAciCkpTEW7pevG0ACJAdCeoi+51wVkADJeZB+oLt5VwUkQHIepPtoHu+qgARIjoM0x8qdIA0CSIDkOEj30ELuVQEJkJwG6Qd6mH9VQAIkp0H6Fy3hXxWQAMlhkL6hf/MvCkiA5DBI6XfELOVeFJAAyWmQplFz7jUzAAmQHAYp/daYTbxrSgEkQHIUpEnU3JbFCJAAyUmQ0mrF/QZIgGRDUWdB+pxanwYkQLKhqKMgnaoWtw6QAAmQLGYctc0AJEACJGs5WaXARkACJECymNHUMQOQAAmQrOVk5YRNgARIgGQxI+iFDEACJECyluMVErdmABIgAZK1DKOu0gsgARIgWcjx8onbpVdAAiRAspAh1NP1CkiABEih50iZQrtdA4AESIAUet6iPu4BQAIkQAo5h0sX3uMeAiRAAqSQM4D6yUOABEiAFGpSShTbLw8CEiABUqjpS294BgEJkAApxOwtWuqgZxiQAAmQgs7eNT9tTc3oTYO8LYAESIAUXHa/UoPEFH+yYKlD3kZAAiRACibp7xYhT6rv9DYDEiABUhBJbUWK3LDG0w5IgARIQeR58ssNnu8kQAIkQDKfaaTKI/IIQAIkQDKd1KpqSDTbPQaQAAmQTOd7jSNi7jGABEiAZDqdtZCKnHCNASRAAiTTeUALif5wjQEkQAIk06mpA8n9CGZAAiRAMp16OpDcT+oDJEACJNNpogPJde8TQAIkQDKfQVpHldxjAAmQAMls1jyuhdTdPQqQAAmQzGVRkxgqpHaUtNU9EpAACZD8c/SX2QvXp6ka02Y0IKo7/vciKkieS2QBCZAASZlZjydJQMo+u07ReGJ8daK7Z4hDM/wdPeEBB0iABEi+7HjQaySh20m58dCw8lQgebn7zX+V30ltTnh+EJAACZC8WVNB+XXT0HUB7K7+JahQlz+9fdY/4d1h94XvJwEJkADJk10V/VfcHk/L2NAliUr13+vXbc0bjWpWubvDjBOKNkACJEDyRLN3++Wn46jSsKMmPi8gARIgyflBe5SIao9PNfV5AQmQohpS6uTurdoP+OGqGUhttI56pJv8vIAESNEMaUlL5krPU7lDSi+thdTc7OcFJECKYkirkxkbMOXTjox1OZ8rpD06a3Z1zH5eQAKk6IWU2Z49tUZ8vTyEsYm5QlqjA6ms2c8LSIAUvZBmMjbFNXChI2t2NjdIm3Qg3Wj28wISIEUvpJ6smfxrvmFscW6QjsVqIf2f2c8LSIAUtZDSGesvD6YwNiQ3SBn1tZD6mv28gARIUQtpA2OT5MHsluyFXCG9rYW00uznBSRAilpIcxhb5BnuxpIv5QbpYCm1oyamPy8gAVLUQprO2FrP8JuMZeQGKeNjlaPCv5n+vIAESFEL6QvGNnmGhzB2THq9eFxMxhmDdPGH9F+jftqcvWa+r/lcFv62oeqFizYUPSNctaPqlX9sKJop2DEH8mkR+Nt+SB8zts0z/D5j+6XXlfXFrDf6iRzl3U2KLbRrwhCEW7K9Q7ZB+pyxzZ5h8RvpsPS6d5iYfVmGWXKHx1Gr3ca9dJIdVG+TuSZcsqHqlas2FM0SbJkD2ZdtKHpFsGUO5NhRNPdFwH5IUxlb5xkWt5HSvCMCXWqe/mPvJvUaPPnuxuBWZbGNhG2kqN1GmqM4CtuDsQumIIUYQAKkqIW0nrFp8mBOa/acbwQg2RBAilpIaYwNlAePMjYYkOQAEiAFl+6spXz1xFzGFgCSHEACpOAyg7HZroFr3Vgz3/FYQLKhKCBFMaRzT7PW0tGjnAmMjVe0A5INAaTohSQsY6zVhBUL+zPW+R9A8gSQACnYzGnmvmdD92PKVkCyIYAUzZCEg+O7tGzXd/4lv0ZAsiGAFNWQdANINgSQAIlDAAmQAIlDAAmQAIlDAAmQAIlDAAmQAIlDAAmQAIlDAAmQAIlDAAmQAIlDAAmQAIlDAAmQAIlDAAmQAIlDAAmQAIlDAAmQAIlDAAmQnAcpYjJ72Ln8ngTTGTY1v6fAdPYM+y2/J8F05g/7y2xXQDJM//ppuXcKk9TvlN9TYDor6k/P70kwnUH1j5rtCkiGASRbAkhOCyDZEkByWgDJlgASgiBGASQE4RBAQhAOAaSIzpm0nPyeBNOJpGkNPoCkl0nMl735PTEBcq0NO6t4mzq5e6v2A364mm/TEyh+0xrGM3jLx92eadtvzE5fi6nZCkh6eS98/85++YMpIS1p6Z7inqfyb4qM4zetYTuDz7/vmazRnvNazM1WQNJLD9b6FU8O5/fEGCerh3LhXJ3M2IApn3ZkrMv5fJwog/hPa7jO4GtvMZY8cNrEASKcwe41UZOzFZB0ktOCDc/vaTCRbdKf27twZrZnT60RXy8PYWxiPk6VfvynNWxn8BLGmm+QBraIcBZJA2ZnKyDpJJ2xsD9ouPDd51wrHN6FcyZjU1wDFzqyZmcNfy4/opnWsJ3BvRhb6x7anMzaZQvmZysg6WQrY8vyexpyy5vyqrz3b9uTNZNP6v9G8RzssIhmWsN1Bh9n7CXP8KuM7RPMz1ZA0on4Bb8z9175myXTxXT2LZzif/L95cEUxobk12TpRj2tYTuD1zP2hWd4PGO/BzFbAUknU8W/edb2JfPWnMm9b75msG/h3MDYJHkwuyV7Ib+myDiKaQ3bGbyyZcu5nuHRjO0OYrYCkk7eZy0XdJDWRZJHhfe1fYqFc468cSylG0u+ZPAT+RclpAiYwdf+w5LPBTFbAUknvXwHOZ7dnt8TEyiKhXO6dzPZtU2SYfAT+RclpAiYwf9jrh2LpmcrIGmT04qxdosyLh9d1omxThfze3ICRLFwfsHYJk/zEMaOGfxE/kUxrREwgxclsxaHhCBmKyBpc5qxLu7/es73U2x9hmEUC+fHjG3zNL/P2P58miLjKKY17Gdwqmgm+VdpyPRsBSRtss+duywPpiSz5ln5OjEBo1g4P2dss6dZXAzC6XQBdxTTGuYzOGtGC/Eb032LFtOzFZACZwBjW/N7GoyjWDinMrbO0yyuzIffxb2D/c4L9Cb8ZvBv/xE33YbJ+xNNz1ZACpzxjK3K72kwjv9eO+/hwh6MXcinKTKOAaRwm8EXRoiM+nhX50zPVkAKnBlh9nf2j2LhXM/YNHkwpzV7Lp8mKEAMIIXZDD7fS1yrW+G7cMr0bAUkbfbv2u0dHsnYlnyclFyiWDjTGBsoDx5lbHA+TVCAKKY1fGfwZXFNc7DyKLHp2QpI2gz2XSOT05U1C8vds+4o/5fvzlrKp/nPZWxBfk2RcRTTGr4zWNwk+uiaX4vZ2QpI2ixgbKg8uFicsfk6LYGjhCSuI812DVzrxpqF3/FY5bSG7Qy+1JZ1uezfZHa2ApI2FzowNkmanzlLWrO24XVFgn+UkM49zVpLhzlyJjA2Ph+nySiKaQ3bGbyMsTmqJrOzFZB0sq4ZY53GfDWqO2NseX5PTKD4bcCLS0GrCSsW9mes8z/5OE1GUU5ruM7gzxjr9aYvqVKbydkKSHpZ31E+E+yZlfk9KQHjvydsTjP3RHcPv/ODBNW0hukMHsz8ctDVaG62ApJuLi56q1Pzdv1mhuu5yXJUu5QPju/Ssl3f+eF35rcU/2kNzxncVQ+SudkKSAjCIYCEIBwCSAjCIYCEIBwCSAjCIYCEIBwCSAjCIYCEIBwCSAjCIYCEIBwCSAjCIYCEIBwCSE5I9m+DH6xWokDZW5+blJnf0xKlASTjJBGR6tlSzYjqWClZk+gDKz8fMNtaV0+qvEpnxILbyJtig67Y8bunEcXbUTdiAkjGkSAVP+nXFM6QvouToMzTtF96mvxyR7oNvxyQ8nsCwjgSJGru1xTGkM6UFqc2rsRCdXtmQ+ljlH5x/JzJrzcv65Jkw+pdbpBGPvTQa/x/axgFkIzjgkSzlU1hDGmxOK0jdVbbOonthcbIIy6NLyS+7c3/t+cGqTPRY/x/axgFkIzjhlReeQ1qGEMaRVRBp3mm+BHK7PC9Xxsvfm/t4v7bT2/cuDHQeEBycERIZcTFsLOiKYwhDdWdtKzriBI2KFveFz/SyzZNg3EAycERIfWpLy52ivtzRBykieL0v+HXckXcTiqVbdNEGAaQHBwRUr8/xTWhqr5bgZqDtOOl2wve2PxHafDEkDvLFLy1uecWnW5IqzrVKFKsxvM/+f3Qym63lipSr8V01WZO1ohbC3XR/IazXzxxc8Hi1Z+e4bsjhy6knFpEJc77t70g2vozQCH9bH/7gUoFi9/671dVjwjSmz6dvipIeh/23IdP31m87L/afZXnynkEkIwjQRIGiIvd694mL6QviK73tp4X+7huz7aRqIpw4dkY937ml3KEmSXkfc5N3QuqBCmzo2dH9KNHvCUOPOpprPiV3LRa+l1npK9EDZApZTy9b3Y/K2GJd+e2/+7vbWKLejFf0aFDh1VGhcaKg4p7JA4W30qPfvyrVYynY2zzczrT971vZ4O270feqdtj9GGF7JGlPY11wuoWXSYDSMZxQcqqJi4i3mdNmYD0dz2imAKuJWLMdGmRSnQND3L1FCEN9S5FRJVT5ArbyvoaPSt/0oJ6uQHpQBqk6Bw7RWoxgjTOM2X60RZKjSVq5+tQnaiG+HLmFkVHejJbO30+SDp9VZD0PqwwUNFWNEWIuACScVyQhFUihjuuyk25Q6r4b6q96OyVzQ+LbQlxVP6bU9m7pSOiCaelDiIkcSkq+95vu398QVxppHruwhnlxOG6o35NWdxaWpA+czVKC2p/14J1u/90TZfanl50YNe3D0oApEfLbWjcWFx+Czdu3HitX9dWImrj+4PqFBIai8uxdzVvk9g8QnztLb5eN3T1gV1LXyqqP30+SDp9ZzVuXIGotDh1R40+7AJxLhcfsPLA1m/vFtv+ZervE1YBJOO4IQld5KVJSu6QiJq4jnfm3OlaxE65OrQUB9dIAzWlxvruxvWlxOHRrsF24tCL7idAThS/wEq59riLC2qRWEoYuO6k73E9UjLEnysgr36JG0ZU65pnULuNJH6dVjX8fLqFJosD3ocu9COKOyl+yYhfsJXl8yE2FRE/oaCZPi8k/b7KbSTdDyvyKeVe78t+RBwffk/uzC2AZBwZ0tnyRAXlzWYTkJLkDZ//icMxf7iHt4rDE6QBCVIRz1lHK8U3VSQke8SBZE+x0eIb10kAqyV0tfZoJusDBWzXUrnINaQLqSRRI8PPp1vorLhsd5TbcioRMfF1qThulqfjs0Q1Bc30eSHp91VA0v2wl2LkDy3mF9I70SncA0jGkSEJs8m7OJqA1EtulBYYJg9fjZc3BiRIvvMKHhDfrRJfXxXX/Lw7uC5XkZc+14Kqs4EjbrcU957kIxFt4RrSg3RNXD6bGX4+/ULNiUrIjzZZI7bNFV9nlilT9i9PxxHSDhXN9Hkh6fdVQNL9sKme/2nEXOjTp88Kw4kO1wCScTyQhBbin3mSa8gEpOly4zlxeIinRwUfpFjflvRCsYv0oCBxBayV79eKImKkh8ZJC+q/tVN1knz/eYsRt8XKeH5MA+kvsW8Ho49nUGiW2Cw/NrW7uDmnOelokBKSd/q+1z1FaJAOJN0Pmy3O66Y52gKRE0AyjhfSyeLi/9KuZ3yYgLRe0fidp0dFHyTF0p5T0PWFIS3uH/papS9A6RCTtKB+qZ2qeeR3iPgjzxaFHqRL3u8rnRgUuliU6HlXy9WyRH3VP5VeTQnJO326kHx9fZAMPuzt4murg0aTGgEBJON4IbnOD2gpDZiAtEfR6F3VV0BqqvgN4tsGgvCz/xK9Q3wr7YqWFtTftVMlHek54Hv7nfjWtSKku41UNMA2klGhjuJ2v2tn4o9ii+IsPeHCtnnDW0knvfogeadPDUnV1wfJ4MMuktYTY+5+5+fwfJRG7gEk4/gg5Twkby1wgNRD8RseI6omCHNIk5GCe0FVnUYg5R1xgVMsbb+RfARVF5K4CVLd6OMZFZKOSS2VBjoq9kOnfND0Ru9xVh8k7/QpIen09UEy+LDC6/KbpMYTvZtYkRRAMo4PkrBPHK7wNxdI7yh+wwtElQRhinbZkracpAVV59lWfcXVTMXbQ2Iv1wPsdSE1FrH8rW7s9uSTydnGhaQ1uhfF16xiRF+4x/3zcpwXxj1KSN7p80HS7euDZPBhBWHpvZ73BV4Oo8czmw0gGUcBSRhOrqVLH1JmqN9IyUS3uY+LPtDYL9ISLS2ox7VTNVi0oXhgsPQrXbubdSFJh4d+UrVdjHd/TRkVEnoSlb3mslHQjfCiaxmPrfv88Dkp0raUD5J3+ryQ9Pv6IBl8WCm7P2jkvnKFapzWfu4wDyAZRwnpqnTiz2oDSAeDgfSk4jfUERcqQVgg9jyq/fUGkMaIzYpTaKSt9WXSgC6kX8l3VMgT6ThNiwCFhLXk2oxp6d3j109sqPyl/M2WCyT9vj5IBh9WzqXVb9eRJD1n2CNcA0jGUUISNoorLNWy9CEtCwbSbYrfUJioq/++PkUMIEkbGat8b0d5fqUuJOmaiUJn/NteJfcJFUaFhJwbpe/Nc0mevQIXSxDd4r28MTAkg74+SAYfVpGl14ldzgTsEoYBJOP4QXL9V/uGElI576jPg4EU59vJu1zsMllc2pM8pwq5krp8+XLpkIoBpGPkd4XRE+KWjusAjP71SO+5V0kVkWy59nMbFRKkM96vz/6K6Cb3+y3kOzomCEMCQjLo64Ok/2G3L1myytv2X7HGOu1HCe8AknH8IV24WVxWqisgFfBuYNwbDCTq4y35sLgxIZ1Q1IjoZt9FOE+79okbQhLE6SjpvcRoR4znhBt9SH+JX3oxi5Utb4tl7wtUyH3xxZrHid51v/1BfOu7yLZ5QEgGfRUHZHU/7DvK2Sn9+vk6HzysA0jG8YfkWoEjz9L6tTgkn0jnOt4SBKSip+TGVSQvXTPEgTGerptj5TdGkKQdCB953rQnzzmm+pBce0kKKk40+lHap7YiUCHBte3WNt6NXHB/y3ivUVoWm+s3kk5fBSTdDzuXPKcMCu4dewFvABGOASTjqCC57sfjWVp3STuf5DWfG4ODRA3cJ0f/IV3Jtloaks44i5/q7rnvevE/Z9dprUaQThUnSpDvuiUxqR3g7G9xg+ffYpe4V+WL8S68Jzl6KmAhMcNcn1Q+cVu4XIDoUXmtb7p0pWLZHM30eSAZ9JUgPSRX0/uw0r73mvKej4Plia7zTkqkBJCMo4b0VzkfpGzp0tKGa7OEv8YXomp3BgHpVnEDZMSG/ctelK5Hcp+L47rmiR6dufvwylekHcDuE2+MIAmTxBExHZYePvC/RhISedPdAJKQUVWa6oJPf/z9vHFdikvDNf8JWEhw74Ykmul5K24+Edt0JefQvEfFHyD3sWmDvXb6fYUXxS2wo8LVbKMPK12ylfT6/D0HVrxaVPlFGTEBJOOoIbnOovEsrfNdy1qstGSW3tc4CEjrqpI3j2bJ47+M9TXSq+42Q0jCa4rO8V/LjUaQhL8akV/uOR64kJR7xIaS3tMejrkumI8rLP1b93vp3xtnGUHS7yudB06xJWP3GH3YvyooJ7FZxH0hAVKAaCAJzAdJGF1Y/qs3OCQEA2nPSc+CHdvbdwT/R+8F2hUny5vixpCECd77G1T3bv4YQhKuflHJt4yWeU95PrdeISnSJeo9fW9XevAX6H0+u5Y0MMkIkn5f4XBJV9seow8rbK7pncQKY68KERdACjmpfRtVLFit+Zyg/+rZ81rcklSqXv/tysbLXz9TtWixu9p/kmX0Y8qc+fSxKgWLVn3mW3N3xL+8sPddlQolVaj34tcXQip0eUxyzaSyDQZLmzF7m5ZKrLUs2L7bWlWML1ZHPhSr92GvzO5Qq0R8hYbPfh6R560CEoJwCCAhCIcAEoJwCCAhCIcAEoJwCCAhCIcAEoJwCCAhCIcAEoJwCCAhCIcAEoJwCCAhCIcAEoJwCCAhCIcAEoJwCCAhCIf8P/9r177EB9KXAAAAAElFTkSuQmCC",
      "text/plain": [
       "plot without title"
      ]
     },
     "metadata": {
      "image/png": {
       "height": 420,
       "width": 420
      }
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "colnames(deg_counts) = c(\"formula\", \"sig_genes\")\n",
    "#plot the number of significant genes vs the number of covariates\n",
    "ggplot(deg_counts, aes(x=formula, y=sig_genes)) +\n",
    "    geom_point(size=5) +\n",
    "    geom_line() +\n",
    "    xlab(\"Number of Covariates\") +\n",
    "    ylab(\"Number of Significant Genes\") +\n",
    "    theme(text = element_text(size = 24))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "~sex + W_1 + W_2 + W_3 + W_4 + W_5 + W_6 + W_7 + W_8 + W_9 + \n",
       "    W_10 + W_11 + W_12 + W_13 + W_14 + +W_15 + W_16 + W_17 + \n",
       "    W_18 + recur"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#use 18 covariates\n",
    "design_formula = formula(formula_list[[18]])\n",
    "design_formula"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "converting counts to integer mode\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "keep\n",
       "FALSE  TRUE \n",
       "28254 29965 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "estimating size factors\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "48 rows did not converge in beta, labelled in mcols(object)$betaConv. Use larger maxit argument with nbinomWaldTest\n",
      "\n"
     ]
    }
   ],
   "source": [
    "dds = DESeqDataSetFromMatrix(countData=cts, colData=pdat, design = design_formula)\n",
    "#filter genes for low expression with edgeR\n",
    "y <- DGEList(counts=counts(dds), group=recur)\n",
    "keep <- filterByExpr(y)\n",
    "table(keep)\n",
    "y <- y[keep,]\n",
    "dds <- dds[keep,]\n",
    "#run DESeq\n",
    "dds <- DESeq(dds)\n",
    "res <- results(dds)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Attaching package: ‘genefilter’\n",
      "\n",
      "\n",
      "The following object is masked _by_ ‘.GlobalEnv’:\n",
      "\n",
      "    varFilter\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:readr’:\n",
      "\n",
      "    spec\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:MatrixGenerics’:\n",
      "\n",
      "    rowSds, rowVars\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    rowSds, rowVars\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library('pheatmap')\n",
    "library('genefilter')\n",
    "#use RemoveBatchEffect to adjust for the identified covariates\n",
    "#first create variables for each of the RUV factors\n",
    "ruv1 = pdat$W_1\n",
    "ruv2 = pdat$W_2\n",
    "ruv3 = pdat$W_3\n",
    "ruv4 = pdat$W_4\n",
    "ruv5 = pdat$W_5\n",
    "ruv6 = pdat$W_6\n",
    "ruv7 = pdat$W_7\n",
    "ruv8 = pdat$W_8\n",
    "ruv9 = pdat$W_9\n",
    "ruv10 = pdat$W_10\n",
    "ruv11 = pdat$W_11\n",
    "ruv12 = pdat$W_12\n",
    "ruv13 = pdat$W_13\n",
    "ruv14 = pdat$W_14\n",
    "ruv15 = pdat$W_15\n",
    "ruv16 = pdat$W_16\n",
    "ruv17 = pdat$W_17\n",
    "ruv18 = pdat$W_18\n",
    "\n",
    "#run RemoveBatchEffect using the RUV factors and sex\n",
    "vsd = vst(dds)\n",
    "mat= assay(vsd)\n",
    "assay(vsd) <- removeBatchEffect(assay(vsd), batch=sex, covariates=cbind(ruv1, ruv2, ruv3, ruv4, ruv5, ruv6, ruv7, ruv8, ruv9, ruv10, ruv11, ruv12, ruv13, ruv14, ruv15, ruv16, ruv17, ruv18))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong>png:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{png:} 2"
      ],
      "text/markdown": [
       "**png:** 2"
      ],
      "text/plain": [
       "png \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#plot a PCA of the adjusted data\n",
    "pcaData <- plotPCA(vsd, intgroup=c(\"recur\"), ntop = 500, returnData=TRUE)\n",
    "percentVar <- round(100 * attr(pcaData, \"percentVar\"))\n",
    "\n",
    "#save the plot\n",
    "png(\"results/i/pca_ruvg_corrected.png\", width=1024, height=1024)\n",
    "ggplot(pcaData, aes(PC1, PC2, color=inflammation_status, shape=recur)) +\n",
    "    geom_point(size=5) +\n",
    "    #geom_text(aes(label=sample),hjust=0, vjust=0) +\n",
    "    xlab(paste0(\"PC1: \",percentVar[1],\"% variance\")) +\n",
    "    ylab(paste0(\"PC2: \",percentVar[2],\"% variance\")) + \n",
    "    coord_fixed() +\n",
    "    \n",
    "    theme(text = element_text(size = 24))\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "log2 fold change (MLE): recur Recurrent vs Non.recurrent \n",
       "Wald test p-value: recur Recurrent vs Non.recurrent \n",
       "DataFrame with 6 rows and 7 columns\n",
       "                                 gene  baseMean log2FoldChange     lfcSE\n",
       "                          <character> <numeric>      <numeric> <numeric>\n",
       "ENSG00000100368.15 ENSG00000100368.15  1408.336       0.539294 0.0768059\n",
       "ENSG00000137462.9   ENSG00000137462.9  1049.409       0.588285 0.0868936\n",
       "ENSG00000249437.8   ENSG00000249437.8  3875.876       0.353731 0.0533505\n",
       "ENSG00000185215.11 ENSG00000185215.11  4540.630       0.639195 0.0977397\n",
       "ENSG00000148926.10 ENSG00000148926.10   559.951       0.758926 0.1183117\n",
       "ENSG00000162551.14 ENSG00000162551.14   190.099       0.939503 0.1523298\n",
       "                        stat      pvalue        padj\n",
       "                   <numeric>   <numeric>   <numeric>\n",
       "ENSG00000100368.15   7.02152 2.19471e-12 5.68387e-08\n",
       "ENSG00000137462.9    6.77018 1.28627e-11 1.66558e-07\n",
       "ENSG00000249437.8    6.63033 3.34947e-11 2.89149e-07\n",
       "ENSG00000185215.11   6.53977 6.16133e-11 3.98915e-07\n",
       "ENSG00000148926.10   6.41463 1.41162e-10 7.31164e-07\n",
       "ENSG00000162551.14   6.16756 6.93520e-10 2.64211e-06"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "resOrdered <- res[order(res$padj),]\n",
    "resOrdered$gene <- rownames(resOrdered)\n",
    "resOrdered = resOrdered[,c(7, 1, 2, 3, 4, 5, 6)]\n",
    "head(resOrdered)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "#make a list of the top 15 genes\n",
    "top15 <- resOrdered[1:15,]\n",
    "recur <- data.frame(recur=coldata$recur)\n",
    "rownames(recur) <- colnames(mat)\n",
    "mat15 <- mat[top15$gene,]\n",
    "#center the data\n",
    "mat15 <- mat15 - rowMeans(mat15)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "'select()' returned 1:1 mapping between keys and columns\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A matrix: 6 × 42 of type dbl</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>PB12_770_A9</th><th scope=col>PB138_A3</th><th scope=col>PB140_ILEUM</th><th scope=col>PB146_A1</th><th scope=col>PB17_236_122_A1</th><th scope=col>PB17_236_177_A1</th><th scope=col>PB17_236_292_B1</th><th scope=col>PB17_236_387_A2</th><th scope=col>PB17_236_398_A3</th><th scope=col>PB17_236_520_A4</th><th scope=col>⋯</th><th scope=col>PB600444_A1</th><th scope=col>PB601102_A2</th><th scope=col>PB601990_A4</th><th scope=col>PB602717_A1</th><th scope=col>PB603387_A1</th><th scope=col>PB603967_A1</th><th scope=col>PB604520_A1</th><th scope=col>PB605312_A1</th><th scope=col>PB75_A2</th><th scope=col>PBCHAMP7_A2</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>CSF2RB</th><td> 0.52512344</td><td>2.8377624</td><td>-0.6021220</td><td>-0.1770484</td><td>-0.25796557</td><td>-0.06425362</td><td>0.2868785</td><td>-0.14921148</td><td> 0.2845260</td><td>-0.08435063</td><td>⋯</td><td>-0.4064393</td><td>-0.5568335</td><td>-1.3378116</td><td>-1.16610877</td><td>-0.67243669</td><td> 0.35995732</td><td>-0.83342252</td><td> 0.40257766</td><td>-0.4665220</td><td>0.71925940</td></tr>\n",
       "\t<tr><th scope=row>TLR2</th><td> 0.89179597</td><td>2.2636921</td><td>-1.0031877</td><td> 1.0939214</td><td>-0.55847127</td><td>-0.45627120</td><td>0.5294881</td><td> 0.11124168</td><td> 0.1302254</td><td> 0.11723288</td><td>⋯</td><td> 0.1591417</td><td> 0.7432911</td><td>-0.8464827</td><td>-1.60960695</td><td>-0.02181970</td><td>-0.62367929</td><td> 0.44506311</td><td>-0.37800864</td><td>-0.2652700</td><td>0.74323054</td></tr>\n",
       "\t<tr><th scope=row>NAIP</th><td>-0.02336788</td><td>0.6527014</td><td>-0.2458138</td><td> 0.6773074</td><td> 0.35933408</td><td>-0.26220355</td><td>0.1700708</td><td>-0.03627255</td><td>-0.2029344</td><td> 0.10937801</td><td>⋯</td><td> 0.2534548</td><td> 0.3474533</td><td>-0.6411854</td><td>-0.01508148</td><td> 0.31191087</td><td>-0.41542597</td><td>-0.03802279</td><td>-0.08339674</td><td> 0.3257464</td><td>0.06336752</td></tr>\n",
       "\t<tr><th scope=row>TNFAIP2</th><td> 0.55724426</td><td>1.8458881</td><td>-0.7173009</td><td> 1.3904475</td><td> 0.05184658</td><td>-0.53357675</td><td>0.2616403</td><td>-0.56162364</td><td> 0.2242787</td><td> 0.17617249</td><td>⋯</td><td> 0.1048737</td><td> 0.2863901</td><td>-0.7485523</td><td>-0.76144035</td><td>-0.33943981</td><td>-0.01836612</td><td> 0.03121163</td><td>-0.44461538</td><td>-0.4992836</td><td>0.05442324</td></tr>\n",
       "\t<tr><th scope=row>ADM</th><td> 1.13481914</td><td>1.7517084</td><td>-1.1143721</td><td> 1.5138080</td><td>-0.21274459</td><td>-0.36875478</td><td>0.9361534</td><td> 0.07809647</td><td> 1.2863198</td><td> 0.28126444</td><td>⋯</td><td>-0.2434752</td><td> 0.3774139</td><td>-1.0619564</td><td>-1.13557939</td><td>-0.06722303</td><td>-0.65886378</td><td> 0.22840164</td><td>-1.59898471</td><td>-1.0683988</td><td>0.79392085</td></tr>\n",
       "\t<tr><th scope=row>ALPL</th><td> 1.80656149</td><td>2.0649235</td><td>-0.4402806</td><td> 0.8126067</td><td>-0.52974084</td><td>-0.59631612</td><td>0.7498372</td><td> 0.33195363</td><td> 0.1525135</td><td>-0.29982614</td><td>⋯</td><td>-1.0131805</td><td> 0.8683495</td><td>-1.0037457</td><td>-0.99147772</td><td>-1.05779024</td><td>-0.58845978</td><td>-0.48711737</td><td>-0.29368516</td><td>-0.5780172</td><td>1.21870844</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A matrix: 6 × 42 of type dbl\n",
       "\\begin{tabular}{r|lllllllllllllllllllll}\n",
       "  & PB12\\_770\\_A9 & PB138\\_A3 & PB140\\_ILEUM & PB146\\_A1 & PB17\\_236\\_122\\_A1 & PB17\\_236\\_177\\_A1 & PB17\\_236\\_292\\_B1 & PB17\\_236\\_387\\_A2 & PB17\\_236\\_398\\_A3 & PB17\\_236\\_520\\_A4 & ⋯ & PB600444\\_A1 & PB601102\\_A2 & PB601990\\_A4 & PB602717\\_A1 & PB603387\\_A1 & PB603967\\_A1 & PB604520\\_A1 & PB605312\\_A1 & PB75\\_A2 & PBCHAMP7\\_A2\\\\\n",
       "\\hline\n",
       "\tCSF2RB &  0.52512344 & 2.8377624 & -0.6021220 & -0.1770484 & -0.25796557 & -0.06425362 & 0.2868785 & -0.14921148 &  0.2845260 & -0.08435063 & ⋯ & -0.4064393 & -0.5568335 & -1.3378116 & -1.16610877 & -0.67243669 &  0.35995732 & -0.83342252 &  0.40257766 & -0.4665220 & 0.71925940\\\\\n",
       "\tTLR2 &  0.89179597 & 2.2636921 & -1.0031877 &  1.0939214 & -0.55847127 & -0.45627120 & 0.5294881 &  0.11124168 &  0.1302254 &  0.11723288 & ⋯ &  0.1591417 &  0.7432911 & -0.8464827 & -1.60960695 & -0.02181970 & -0.62367929 &  0.44506311 & -0.37800864 & -0.2652700 & 0.74323054\\\\\n",
       "\tNAIP & -0.02336788 & 0.6527014 & -0.2458138 &  0.6773074 &  0.35933408 & -0.26220355 & 0.1700708 & -0.03627255 & -0.2029344 &  0.10937801 & ⋯ &  0.2534548 &  0.3474533 & -0.6411854 & -0.01508148 &  0.31191087 & -0.41542597 & -0.03802279 & -0.08339674 &  0.3257464 & 0.06336752\\\\\n",
       "\tTNFAIP2 &  0.55724426 & 1.8458881 & -0.7173009 &  1.3904475 &  0.05184658 & -0.53357675 & 0.2616403 & -0.56162364 &  0.2242787 &  0.17617249 & ⋯ &  0.1048737 &  0.2863901 & -0.7485523 & -0.76144035 & -0.33943981 & -0.01836612 &  0.03121163 & -0.44461538 & -0.4992836 & 0.05442324\\\\\n",
       "\tADM &  1.13481914 & 1.7517084 & -1.1143721 &  1.5138080 & -0.21274459 & -0.36875478 & 0.9361534 &  0.07809647 &  1.2863198 &  0.28126444 & ⋯ & -0.2434752 &  0.3774139 & -1.0619564 & -1.13557939 & -0.06722303 & -0.65886378 &  0.22840164 & -1.59898471 & -1.0683988 & 0.79392085\\\\\n",
       "\tALPL &  1.80656149 & 2.0649235 & -0.4402806 &  0.8126067 & -0.52974084 & -0.59631612 & 0.7498372 &  0.33195363 &  0.1525135 & -0.29982614 & ⋯ & -1.0131805 &  0.8683495 & -1.0037457 & -0.99147772 & -1.05779024 & -0.58845978 & -0.48711737 & -0.29368516 & -0.5780172 & 1.21870844\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A matrix: 6 × 42 of type dbl\n",
       "\n",
       "| <!--/--> | PB12_770_A9 | PB138_A3 | PB140_ILEUM | PB146_A1 | PB17_236_122_A1 | PB17_236_177_A1 | PB17_236_292_B1 | PB17_236_387_A2 | PB17_236_398_A3 | PB17_236_520_A4 | ⋯ | PB600444_A1 | PB601102_A2 | PB601990_A4 | PB602717_A1 | PB603387_A1 | PB603967_A1 | PB604520_A1 | PB605312_A1 | PB75_A2 | PBCHAMP7_A2 |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| CSF2RB |  0.52512344 | 2.8377624 | -0.6021220 | -0.1770484 | -0.25796557 | -0.06425362 | 0.2868785 | -0.14921148 |  0.2845260 | -0.08435063 | ⋯ | -0.4064393 | -0.5568335 | -1.3378116 | -1.16610877 | -0.67243669 |  0.35995732 | -0.83342252 |  0.40257766 | -0.4665220 | 0.71925940 |\n",
       "| TLR2 |  0.89179597 | 2.2636921 | -1.0031877 |  1.0939214 | -0.55847127 | -0.45627120 | 0.5294881 |  0.11124168 |  0.1302254 |  0.11723288 | ⋯ |  0.1591417 |  0.7432911 | -0.8464827 | -1.60960695 | -0.02181970 | -0.62367929 |  0.44506311 | -0.37800864 | -0.2652700 | 0.74323054 |\n",
       "| NAIP | -0.02336788 | 0.6527014 | -0.2458138 |  0.6773074 |  0.35933408 | -0.26220355 | 0.1700708 | -0.03627255 | -0.2029344 |  0.10937801 | ⋯ |  0.2534548 |  0.3474533 | -0.6411854 | -0.01508148 |  0.31191087 | -0.41542597 | -0.03802279 | -0.08339674 |  0.3257464 | 0.06336752 |\n",
       "| TNFAIP2 |  0.55724426 | 1.8458881 | -0.7173009 |  1.3904475 |  0.05184658 | -0.53357675 | 0.2616403 | -0.56162364 |  0.2242787 |  0.17617249 | ⋯ |  0.1048737 |  0.2863901 | -0.7485523 | -0.76144035 | -0.33943981 | -0.01836612 |  0.03121163 | -0.44461538 | -0.4992836 | 0.05442324 |\n",
       "| ADM |  1.13481914 | 1.7517084 | -1.1143721 |  1.5138080 | -0.21274459 | -0.36875478 | 0.9361534 |  0.07809647 |  1.2863198 |  0.28126444 | ⋯ | -0.2434752 |  0.3774139 | -1.0619564 | -1.13557939 | -0.06722303 | -0.65886378 |  0.22840164 | -1.59898471 | -1.0683988 | 0.79392085 |\n",
       "| ALPL |  1.80656149 | 2.0649235 | -0.4402806 |  0.8126067 | -0.52974084 | -0.59631612 | 0.7498372 |  0.33195363 |  0.1525135 | -0.29982614 | ⋯ | -1.0131805 |  0.8683495 | -1.0037457 | -0.99147772 | -1.05779024 | -0.58845978 | -0.48711737 | -0.29368516 | -0.5780172 | 1.21870844 |\n",
       "\n"
      ],
      "text/plain": [
       "        PB12_770_A9 PB138_A3  PB140_ILEUM PB146_A1   PB17_236_122_A1\n",
       "CSF2RB   0.52512344 2.8377624 -0.6021220  -0.1770484 -0.25796557    \n",
       "TLR2     0.89179597 2.2636921 -1.0031877   1.0939214 -0.55847127    \n",
       "NAIP    -0.02336788 0.6527014 -0.2458138   0.6773074  0.35933408    \n",
       "TNFAIP2  0.55724426 1.8458881 -0.7173009   1.3904475  0.05184658    \n",
       "ADM      1.13481914 1.7517084 -1.1143721   1.5138080 -0.21274459    \n",
       "ALPL     1.80656149 2.0649235 -0.4402806   0.8126067 -0.52974084    \n",
       "        PB17_236_177_A1 PB17_236_292_B1 PB17_236_387_A2 PB17_236_398_A3\n",
       "CSF2RB  -0.06425362     0.2868785       -0.14921148      0.2845260     \n",
       "TLR2    -0.45627120     0.5294881        0.11124168      0.1302254     \n",
       "NAIP    -0.26220355     0.1700708       -0.03627255     -0.2029344     \n",
       "TNFAIP2 -0.53357675     0.2616403       -0.56162364      0.2242787     \n",
       "ADM     -0.36875478     0.9361534        0.07809647      1.2863198     \n",
       "ALPL    -0.59631612     0.7498372        0.33195363      0.1525135     \n",
       "        PB17_236_520_A4 ⋯ PB600444_A1 PB601102_A2 PB601990_A4 PB602717_A1\n",
       "CSF2RB  -0.08435063     ⋯ -0.4064393  -0.5568335  -1.3378116  -1.16610877\n",
       "TLR2     0.11723288     ⋯  0.1591417   0.7432911  -0.8464827  -1.60960695\n",
       "NAIP     0.10937801     ⋯  0.2534548   0.3474533  -0.6411854  -0.01508148\n",
       "TNFAIP2  0.17617249     ⋯  0.1048737   0.2863901  -0.7485523  -0.76144035\n",
       "ADM      0.28126444     ⋯ -0.2434752   0.3774139  -1.0619564  -1.13557939\n",
       "ALPL    -0.29982614     ⋯ -1.0131805   0.8683495  -1.0037457  -0.99147772\n",
       "        PB603387_A1 PB603967_A1 PB604520_A1 PB605312_A1 PB75_A2    PBCHAMP7_A2\n",
       "CSF2RB  -0.67243669  0.35995732 -0.83342252  0.40257766 -0.4665220 0.71925940 \n",
       "TLR2    -0.02181970 -0.62367929  0.44506311 -0.37800864 -0.2652700 0.74323054 \n",
       "NAIP     0.31191087 -0.41542597 -0.03802279 -0.08339674  0.3257464 0.06336752 \n",
       "TNFAIP2 -0.33943981 -0.01836612  0.03121163 -0.44461538 -0.4992836 0.05442324 \n",
       "ADM     -0.06722303 -0.65886378  0.22840164 -1.59898471 -1.0683988 0.79392085 \n",
       "ALPL    -1.05779024 -0.58845978 -0.48711737 -0.29368516 -0.5780172 1.21870844 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#rename the rownames of mat with gene symbols\n",
    "library(\"AnnotationDbi\")\n",
    "library(\"org.Hs.eg.db\")\n",
    "\n",
    "#remove the ENSEMBL version number\n",
    "rownames(mat15) <- gsub(\"\\\\..*\", \"\", rownames(mat15))\n",
    "rownames(mat15) <- mapIds(org.Hs.eg.db, keys=rownames(mat15), column=\"SYMBOL\", keytype=\"ENSEMBL\", multiVals=\"first\")\n",
    "head(mat15)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "#sort each dataframe by recurrence\n",
    "#first sort the annotation data by recurrence while keeping it as a dataframe\n",
    "recur <- recur[order(recur$recur), , drop=FALSE]\n",
    "#then sort the matrix by the order of the annotation data by getting the rownames of the annotation data and reordering the columns of the matrix\n",
    "col_order = rownames(recur)\n",
    "mat15 <- mat15[,col_order]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "#remove the sample names from the heatmap\n",
    "pheatmap(mat15, annotation_col=recur, cluster_cols=F, show_colnames=F, filename='../../results/recur/i/ruvg_runs/heatmap.png', height=10, width=10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "converting counts to integer mode\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "keep\n",
       "FALSE  TRUE \n",
       "30087 31027 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "estimating size factors\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "fitting model and testing\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "\n",
       "FALSE  TRUE \n",
       "17667  7946 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#adjust for sex and RUV factors\n",
    "design <- as.formula(~ sex + W_1 + W_2 + W_3 + W_4 + W_5 + recur)\n",
    "#create the dds object\n",
    "dds = DESeqDataSetFromMatrix(countData = cts, colData = pdat, design = design)\n",
    "\n",
    "#filter genes for low expression with edgeR\n",
    "y <- DGEList(counts=counts(dds), group=recur)\n",
    "keep <- filterByExpr(y)\n",
    "table(keep)\n",
    "y <- y[keep,]\n",
    "dds <- dds[keep,]\n",
    "\n",
    "#run DESeq\n",
    "dds <- DESeq(dds)\n",
    "res <- results(dds)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\n",
       "FALSE  TRUE \n",
       "25006   607 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "table(res$padj < 0.05)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "log2 fold change (MLE): recur Recurrent vs Non.recurrent \n",
       "Wald test p-value: recur Recurrent vs Non.recurrent \n",
       "DataFrame with 6 rows and 6 columns\n",
       "                    baseMean log2FoldChange     lfcSE      stat      pvalue\n",
       "                   <numeric>      <numeric> <numeric> <numeric>   <numeric>\n",
       "ENSG00000075618.18   476.349       0.572330 0.0916091   6.24753 4.17003e-10\n",
       "ENSG00000185215.11  4541.105       0.686794 0.1109233   6.19161 5.95514e-10\n",
       "ENSG00000160255.18  1823.288       0.534797 0.0920382   5.81060 6.22486e-09\n",
       "ENSG00000090402.8  19287.383      -0.723901 0.1312680  -5.51468 3.49421e-08\n",
       "ENSG00000100368.15  1408.487       0.615353 0.1188780   5.17634 2.26282e-07\n",
       "ENSG00000149781.13   739.985       0.358620 0.0692847   5.17603 2.26656e-07\n",
       "                          padj\n",
       "                     <numeric>\n",
       "ENSG00000075618.18 7.62644e-06\n",
       "ENSG00000185215.11 7.62644e-06\n",
       "ENSG00000160255.18 5.31458e-05\n",
       "ENSG00000090402.8  2.23743e-04\n",
       "ENSG00000100368.15 8.29333e-04\n",
       "ENSG00000149781.13 8.29333e-04"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "resOrdered <- res[order(res$padj),]\n",
    "head(resOrdered)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "#save the results to csv\n",
    "write.csv(resOrdered, file=\"../../results/recur/i/ruvg_runs/res_ruvg9_5.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "# Calculate the average gene expression of each genes and take the top 5000 highest expressed\n",
    "nc <- as.data.frame(assay(vsd))\n",
    "nc$avg_exp <- rowSums(nc,na.rm=TRUE) / ncol(nc) # x should be your normalized matrix\n",
    "nc <- arrange(nc, -avg_exp)%>%dplyr::select(-avg_exp)\n",
    "nc <- nc[1:5000,]\n",
    "\n",
    "# Calculate the variance of each genes, and choose the lowest 1000 genes as the negative control gene\n",
    "nc$row_stv <- rowSds(as.matrix(nc))/(rowSums(nc,na.rm=TRUE) / ncol(nc))\n",
    "nc <- arrange(nc, row_stv) %>% dplyr::select(-row_stv)\n",
    "nc <- nc[1:1000,]\n",
    "##The newSeqExpressionSet is given the raw count matrix, phenotypic data, and then the names of the samples \n",
    "# Create a new Expression Set and perform an upper quartile normalization\n",
    "nc <- round(nc*1000)\n",
    "set <- newSeqExpressionSet(as.matrix(nc),phenoData = data.frame(colData(dds),row.names=colnames(nc)))\n",
    "\n",
    "\n",
    "##Run RUV\n",
    "set_ruv<- RUVg(set, rownames(nc), k=10)\n",
    "# Save the phenotype data\n",
    "ruv_data <- pData(set_ruv)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "\n",
    "plotRLE(set_ruv, outline=FALSE, ylim=c(-0.1, 0.1), col=as.numeric(as.factor(coldata$disease)))\n",
    "# dev.off()\n",
    "#plot RLE plots for ruv factors 1-10\n",
    "for (x in 1:10) {\n",
    "    set_ruv <- RUVg(set, rownames(nc), k=x)\n",
    "    png(sprintf('results/i/rle_%s.png', x))\n",
    "    plotRLE(set_ruv, outline=FALSE, ylim=c(-0.1, 0.1), col=as.numeric(as.factor(coldata$disease)))\n",
    "    dev.off()\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "converting counts to integer mode\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "estimating size factors\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "converting counts to integer mode\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "estimating size factors\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "converting counts to integer mode\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "estimating size factors\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "converting counts to integer mode\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "estimating size factors\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "1 rows did not converge in beta, labelled in mcols(object)$betaConv. Use larger maxit argument with nbinomWaldTest\n",
      "\n",
      "converting counts to integer mode\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "estimating size factors\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "1 rows did not converge in beta, labelled in mcols(object)$betaConv. Use larger maxit argument with nbinomWaldTest\n",
      "\n",
      "converting counts to integer mode\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "estimating size factors\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "2 rows did not converge in beta, labelled in mcols(object)$betaConv. Use larger maxit argument with nbinomWaldTest\n",
      "\n",
      "converting counts to integer mode\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "estimating size factors\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "5 rows did not converge in beta, labelled in mcols(object)$betaConv. Use larger maxit argument with nbinomWaldTest\n",
      "\n",
      "converting counts to integer mode\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "estimating size factors\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "8 rows did not converge in beta, labelled in mcols(object)$betaConv. Use larger maxit argument with nbinomWaldTest\n",
      "\n",
      "converting counts to integer mode\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "estimating size factors\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "8 rows did not converge in beta, labelled in mcols(object)$betaConv. Use larger maxit argument with nbinomWaldTest\n",
      "\n",
      "converting counts to integer mode\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "estimating size factors\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "8 rows did not converge in beta, labelled in mcols(object)$betaConv. Use larger maxit argument with nbinomWaldTest\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 10 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>X1L</th><th scope=col>X12L</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td> 1</td><td>12</td></tr>\n",
       "\t<tr><th scope=row>2</th><td> 2</td><td>11</td></tr>\n",
       "\t<tr><th scope=row>3</th><td> 3</td><td> 4</td></tr>\n",
       "\t<tr><th scope=row>4</th><td> 4</td><td> 4</td></tr>\n",
       "\t<tr><th scope=row>5</th><td> 5</td><td> 7</td></tr>\n",
       "\t<tr><th scope=row>6</th><td> 6</td><td> 1</td></tr>\n",
       "\t<tr><th scope=row>7</th><td> 7</td><td> 1</td></tr>\n",
       "\t<tr><th scope=row>8</th><td> 8</td><td> 1</td></tr>\n",
       "\t<tr><th scope=row>9</th><td> 9</td><td> 1</td></tr>\n",
       "\t<tr><th scope=row>10</th><td>10</td><td> 1</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 10 × 2\n",
       "\\begin{tabular}{r|ll}\n",
       "  & X1L & X12L\\\\\n",
       "  & <int> & <int>\\\\\n",
       "\\hline\n",
       "\t1 &  1 & 12\\\\\n",
       "\t2 &  2 & 11\\\\\n",
       "\t3 &  3 &  4\\\\\n",
       "\t4 &  4 &  4\\\\\n",
       "\t5 &  5 &  7\\\\\n",
       "\t6 &  6 &  1\\\\\n",
       "\t7 &  7 &  1\\\\\n",
       "\t8 &  8 &  1\\\\\n",
       "\t9 &  9 &  1\\\\\n",
       "\t10 & 10 &  1\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 10 × 2\n",
       "\n",
       "| <!--/--> | X1L &lt;int&gt; | X12L &lt;int&gt; |\n",
       "|---|---|---|\n",
       "| 1 |  1 | 12 |\n",
       "| 2 |  2 | 11 |\n",
       "| 3 |  3 |  4 |\n",
       "| 4 |  4 |  4 |\n",
       "| 5 |  5 |  7 |\n",
       "| 6 |  6 |  1 |\n",
       "| 7 |  7 |  1 |\n",
       "| 8 |  8 |  1 |\n",
       "| 9 |  9 |  1 |\n",
       "| 10 | 10 |  1 |\n",
       "\n"
      ],
      "text/plain": [
       "   X1L X12L\n",
       "1   1  12  \n",
       "2   2  11  \n",
       "3   3   4  \n",
       "4   4   4  \n",
       "5   5   7  \n",
       "6   6   1  \n",
       "7   7   1  \n",
       "8   8   1  \n",
       "9   9   1  \n",
       "10 10   1  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "factor_cols = c('sex', 'recur')\n",
    "for (col in factor_cols) {\n",
    "  ruv_data[,col] = as.factor(ruv_data[,col])\n",
    "}\n",
    "sex = ruv_data$sex\n",
    "recur = ruv_data$recur\n",
    "tin = ruv_data$TIN\n",
    "inflammation_status = ruv_data$inflammation_status\n",
    "\n",
    "formula_list = list(\n",
    "~ sex + inflammation_status + W_1 + recur, \n",
    "~ sex + inflammation_status + W_1 + W_2 + recur, \n",
    "~ sex + inflammation_status + W_1 + W_2 + W_3 + recur, \n",
    "~ sex + inflammation_status + W_1 + W_2 + W_3 + W_4 + recur, \n",
    "~ sex + inflammation_status + W_1 + W_2 + W_3 + W_4 + W_5 + recur, \n",
    "~ sex + inflammation_status + W_1 + W_2 + W_3 + W_4 + W_5 + W_6 + recur, \n",
    "~ sex + inflammation_status + W_1 + W_2 + W_3 + W_4 + W_5 + W_6 + W_7 + recur, \n",
    "~ sex + inflammation_status + W_1 + W_2 + W_3 + W_4 + W_5 + W_6 + W_7 + W_8 + recur, \n",
    "~ sex + inflammation_status + W_1 + W_2 + W_3 + W_4 + W_5 + W_6 + W_7 + W_8 + W_9 + recur, \n",
    "~ sex + inflammation_status + W_1 + W_2 + W_3 + W_4 + W_5 + W_6 + W_7 + W_8 + W_9 + W_10 + recur)\n",
    "\n",
    "\n",
    "#create an empty dataframe to store the number of significant genes\n",
    "deg_counts = data.frame(matrix(ncol=10, nrow=0))\n",
    "colnames(deg_counts) = c(\"formula\", \"sig_genes\")\n",
    "\n",
    "for (x in seq_along(formula_list)) {\n",
    "    design_formula = formula(formula_list[[x]])\n",
    "    dds1 <- DESeqDataSetFromMatrix(countData=cts, colData=ruv_data, design = design_formula)\n",
    "    y <- DGEList(counts=counts(dds1), group=coldata$recur)\n",
    "    keep <- filterByExpr(y)\n",
    "    table(keep)\n",
    "    y <- y[keep,]\n",
    "    dds1 <- dds1[keep,]\n",
    "    dds1 <- DESeq(dds1)\n",
    "    res <- results(dds1)\n",
    "    res <- res[order(res$padj),]\n",
    "    #get the number of significant genes\n",
    "    sig_genes <- res[which(res$padj < 0.05),]\n",
    "    #save the number of significant genes to the dataframe\n",
    "    deg_counts = rbind(deg_counts, c(x, nrow(sig_genes)))\n",
    "}\n",
    "\n",
    "head(deg_counts, 10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "converting counts to integer mode\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<strong>png:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{png:} 2"
      ],
      "text/markdown": [
       "**png:** 2"
      ],
      "text/plain": [
       "png \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "\n",
    "#RLE plot\n",
    "# png(\"results/i/rle_1.png\")\n",
    "# plotRLE(set_ruv, outline=FALSE, ylim=c(-0.1, 0.1), col=as.numeric(as.factor(coldata$status)))\n",
    "# dev.off()\n",
    "ruv_data<-pData(set_ruv)\n",
    "\n",
    "factor_cols = c('sex', 'inflammation_status', 'recur')\n",
    "for (col in factor_cols) {\n",
    "  ruv_data[,col] = as.factor(ruv_data[,col])\n",
    "}\n",
    "sex = ruv_data$sex\n",
    "tissue = ruv_data$tissue\n",
    "inflammation_status = ruv_data$inflammation_status\n",
    "recur = ruv_data$recur\n",
    "tin = ruv_data$TIN\n",
    "\n",
    "w_1 = ruv_data$W_1\n",
    "\n",
    "dds1 <- DESeqDataSetFromMatrix(countData=cts, colData=ruv_data, design = ~sex + inflammation_status + W_1 + recur)\n",
    "vsd = vst(dds1)\n",
    "mat= assay(vsd)\n",
    "assay(vsd) <- removeBatchEffect(assay(vsd), batch=sex, batch2=inflammation_status, covariates=cbind(w_1))\n",
    "\n",
    "#save the PCA, accounting for sex, tin score, and the RUV factor (W_1)\n",
    "pcaData <- plotPCA(vsd, intgroup=c(\"recur\"), ntop = 500, returnData=TRUE)\n",
    "percentVar <- round(100 * attr(pcaData, \"percentVar\"))\n",
    "\n",
    "png(\"results/i/pca_w1_infl_corrected.png\", width=1024, height=512)\n",
    "ggplot(pcaData, aes(PC1, PC2, color=recur)) +\n",
    "    geom_point(size=5) +\n",
    "    # geom_text(aes(label=sample),hjust=0, vjust=0) +\n",
    "    xlab(paste0(\"PC1: \",percentVar[1],\"% variance\")) +\n",
    "    ylab(paste0(\"PC2: \",percentVar[2],\"% variance\")) + \n",
    "    coord_fixed() +\n",
    "    # scale_color_manual(values = c(\"Non-recurrent\" = \"#0066cc\", \"Recurrent\" = \"#990000\")) + \n",
    "    theme(text = element_text(size = 24))\n",
    "dev.off()\n",
    "\n",
    "#write out the matrix to csv\n",
    "write.csv(mat, file=\"results/i/data_ruv1.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong>png:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{png:} 2"
      ],
      "text/markdown": [
       "**png:** 2"
      ],
      "text/plain": [
       "png \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "png(\"results/i/pca_w1_source_corrected.png\", width=1024, height=512)\n",
    "ggplot(pcaData, aes(PC1, PC2, color=source)) +\n",
    "    geom_point(size=5) +\n",
    "    # geom_text(aes(label=sample),hjust=0, vjust=0) +\n",
    "    xlab(paste0(\"PC1: \",percentVar[1],\"% variance\")) +\n",
    "    ylab(paste0(\"PC2: \",percentVar[2],\"% variance\")) + \n",
    "    coord_fixed() +\n",
    "    # scale_color_manual(values = c(\"Non-recurrent\" = \"#0066cc\", \"Recurrent\" = \"#990000\")) + \n",
    "    theme(text = element_text(size = 24))\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "converting counts to integer mode\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "keep\n",
       "FALSE  TRUE \n",
       "30087 31027 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "estimating size factors\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "fitting model and testing\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "log2 fold change (MLE): recur Recurrent vs Non-recurrent \n",
       "Wald test p-value: recur Recurrent vs Non.recurrent \n",
       "DataFrame with 6 rows and 7 columns\n",
       "                                 gene  baseMean log2FoldChange     lfcSE\n",
       "                          <character> <numeric>      <numeric> <numeric>\n",
       "ENSG00000198019.13 ENSG00000198019.13  129.4259       1.341187  0.258688\n",
       "ENSG00000255398.3   ENSG00000255398.3  312.7256       2.940584  0.559086\n",
       "ENSG00000276085.1   ENSG00000276085.1   78.5554       1.858028  0.355389\n",
       "ENSG00000185215.11 ENSG00000185215.11 4541.1053       0.682521  0.133846\n",
       "ENSG00000091106.19 ENSG00000091106.19  127.1549       0.899212  0.184106\n",
       "ENSG00000115956.10 ENSG00000115956.10 1560.4577       1.306197  0.272382\n",
       "                        stat      pvalue       padj\n",
       "                   <numeric>   <numeric>  <numeric>\n",
       "ENSG00000198019.13   5.18458 2.16507e-07 0.00197866\n",
       "ENSG00000255398.3    5.25963 1.44349e-07 0.00197866\n",
       "ENSG00000276085.1    5.22816 1.71205e-07 0.00197866\n",
       "ENSG00000185215.11   5.09929 3.40935e-07 0.00233686\n",
       "ENSG00000091106.19   4.88421 1.03843e-06 0.00569411\n",
       "ENSG00000115956.10   4.79545 1.62308e-06 0.00635715"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#now the DESeq analysis\n",
    "dds1 <- DESeqDataSetFromMatrix(countData=cts,colData=ruv_data,design = ~sex + inflammation_status + W_1 + recur)\n",
    "#filter out lowly expressed genes\n",
    "y <- DGEList(counts=counts(dds1), group=coldata$recur)\n",
    "keep <- filterByExpr(y)\n",
    "table(keep)\n",
    "y <- y[keep,]\n",
    "dds1 <- dds1[keep,]\n",
    "#run DESeq\n",
    "dds1 <- DESeq(dds1)\n",
    "#get the results, with nonIBD as the reference\n",
    "res <- results(dds1, contrast=c('recur', 'Recurrent', 'Non-recurrent'))\n",
    "#order the results by adjusted p-value\n",
    "resOrdered <- res[order(res$padj),]\n",
    "#create a column named gene which is the gene name\n",
    "resOrdered$gene = rownames(resOrdered)\n",
    "#set gene as the first column by making the last column the first column\n",
    "resOrdered = resOrdered[,c(7, 1, 2, 3, 4, 5, 6)]\n",
    "head(resOrdered)\n",
    "\n",
    "write.csv(resOrdered, file=\"results/i/res_ruv1_infl.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.0"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "916dbcbb3f70747c44a77c7bcd40155683ae19c65e1c03b4aa3499c5328201f1"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
